<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>EEB313: Quantitative Methods in R for Biology - Project description</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./projects_databases.html" rel="next">
<link href="./assignments/assignment-04.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./projects.html">Project</a></li><li class="breadcrumb-item"><a href="./projects.html">Project description</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">EEB313: Quantitative Methods in R for Biology</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/EEB313/eeb313.github.io" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Syllabus</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./about-us.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024 teaching team</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./downloadingR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Installing R &amp; Ubuntu</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Lectures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec01-r_and_rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to the course</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec02-basic-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to R: assignment, vectors, functions, strings, loops</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec03-command_git.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Command line &amp; Git(Hub)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec04-data-wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data wrangling!</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec05-data-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data visualization with ggplot2</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec06-exploratory-data-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploratory data analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec07-self-directed-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Review using Farrell and Davies (2019) dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec08-intro-inference-1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to inference I</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec09-intro-inference-2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to inference II</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec10-linear-models-1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Linear models I</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/lec11-linear-models-2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Linear models II

## Lesson preamble

&gt; ### Learning objectives
&gt;
&gt; - Learn how dependent data can be modeled, and the sitations in which one may encounter such data.
&gt; - Understand the structure, assumptions, and implementation of GLMs.
&gt; - Understand the structure, assumptions, and implementation of LLMs.
&gt; - Learn how to use simulation to inform what data that should be collected, and how.
&gt; 
&gt; ### Lesson outline
&gt; -   Generalized linear models
&gt;     - Structure and assumptions, including interpretation of the effects
&gt;     - Implement logistic regression using dataset of Farrell and Davies (2019)
&gt;     - Other types of GLMs (Poisson, negative binomial, etc.)
&gt; -   Power analysis!
&gt; -   Dealing with dependent data
&gt;     - Splitting the data into groups...
&gt;     - Controlling for phylogeny
&gt;     - Linear mixed models: structure and assumptions
&gt;       - Difference between fixed, random effects; examples
&gt;       - Implement mixed models using sexual size dimorphism data

```{r, message=F, warning=F}
library(tidyverse)
library(broom)
library(lme4)
```

## Generalized linear models: theory and examples

So far we have seen how linear models can be used to describe relationships between a response variable and predictors when the data is normally distributed or can be transformed so that this assumption is not violated. What if we were, say, interested in a *binary* response (e.g., infection status of a host with a particular parasite) and how it changes with a continuous predictor (e.g., age of the host)? In this case, one can use a special kind of linear model called *logistic regression* to estimate the additive effect of predictor(s) on the binary response. Logistic regression is a special kind of **generalized linear model**.

### GLMs: structure and assumptions

Generalized linear models describe how the mean of a response variable changes as a function of the predictors when important assumptions of the linear modeling framework (normality, constant error variance, etc.) are violated. In particular, **GLMs allow us to work with data that are not normal, whose range is restricted, or whose variance depends on the mean.** The latter might be important if, say, larger values of the response were also more variable.

A GLM models the transformed mean of a response variable $Y_i$ as a linear combination of the predictors $x_1,\dots,x_p$. The goal of using a GLM is often to estimate how the predictors (e.g., sex and previous exposure to the disease) affect the response (e.g., infection status). sThe transformation of the response is done using a "link" function $g(\cdot)$ that is specific to the distribution which used to model the data. Written out, a GLM is of the form

$$g(\mu_i) = \beta_0 + \beta_1 x_{1i} + \cdots \beta_p x_{pi}.$$

The link functions for the Normal, Gamma, Exponential, Poisson, and Multinomial distributions are known. In general, GLMs apply when the data are modeled using a member of the [exponential family](https://en.wikipedia.org/wiki/Exponential_family). The distributions will have a mean parameter $\mu$ and, sometimes, a parameter $\tau$ which characterizes the dispersion around the mean. _GLMs are fitted using by maximizing the likelihood function that results from assuming the data arise from a distribution in the exponential family (with a mean and dispersion that depend on the predictors), using using numerical optimization methods._

### Interpretation of the effects

Notice that, in a GLM, because the mean of the response been transformed in a particular way, the coefficients must be interpreted carefully. In particular, $\beta_{j}$ is how a per-unit change in $x_{j}$ increases or decreases the _transformed_ mean $g(\mu_i)$.

### Example: logistic regression

In a previous class, we estimated the probability of death given infection for the wild boar when infected with viruses in the family _Coronaviridae_. In your current homework, you have been tasked with extending that model to all host and parasite family combinations. Excitingly, the dataset which have used includes information about the mean evolutionary isolation of all hosts that are infected with a parasite from all other hosts which are known to be infected.

Farrell and Davies tested if the mean evolutionary isolation affected the probability of death. They used a complex model to control for sampling bias and other confounding aspects of the data. We will ignore those complexities and see if, using a GLM, we can recapitulate their findings. To get started, load the `disease_distance.csv` dataset.

```{r}
disease_distance &lt;- read_csv("data/disease_distance.csv")

disease_distance %&gt;% 
  mutate(AnyDeaths = case_when(Deaths &gt; 0 ~ 1,
                               Deaths == 0 ~ 0)) -&gt; DataBern

DataBern |&gt;
  ggplot(aes(x = EvoIso, y = AnyDeaths)) + 
  geom_point()
```

These are binary data. We CANNOT use a linear model, as this assumes the data are Normal. In fact, these data are distributed according to a Bernoulli($p$) distribution. This is a member of the exponential family and the type of generalized linear model when the data are distributed in this way (i.e., are binary) is called **logistic regression**.

The mean of Bernoulli($p$) data is just $p$, and the link function for $p$ is

$$ \text{logit}(p) = \log\frac{p}{1-p}.$$

$\text{logit}(p)$ is called the *log-odds*, which can be thought of as a likelihood the response takes on the value one. In logistic regression, the log-odds ratio is modeled as a linear combination in the predictors:

$$ \text{logit}(p_i) =  \beta_0 + \beta_1 x_{1i} + \cdots + \beta_p x_{pi}.$$

Notice that increasing $x_j$ by one unit results in change $\beta_j$ to the link-transformed response.This is how the effect sizes are interpreted for GLMs such as this one.


```{r}
model &lt;- glm(AnyDeaths ~ EvoIso, 
             family = "binomial", 
             data = DataBern)
summary(model)
```

#### Challenge

How do we interpret the regression coefficient above?

#### Challenge

What are the log-odds of death if the evolutionary isolation of hosts is $EI = 200$? How much does this quantity change if the evolutionary isolation were to increase by 20 million years?

#### Visualizing the fitted model

```{r}
DataBern |&gt;
  ggplot(aes(x = EvoIso, y = AnyDeaths)) + 
  geom_point() +
  geom_smooth(method = "glm", 
              method.args = list(family = "binomial")
              )

### base R implementation

EvoIso &lt;- seq(0, 200, 0.1)
predicted_prob &lt;- predict(model, list(EvoIso = EvoIso), type = "response")

plot(DataBern$EvoIso, DataBern$AnyDeaths)
lines(EvoIso, predicted_prob)
```

### Other GLMs

Here are some common types of response variables and their corresponding distributions:

-   **Count data**: the **Poisson** distribution
-   **Over-dispersed count data** (when the count data is more spread out than "expected"): the **negative binomial** distribution
-   **Binary data** (two discrete categories): the **binomial** distribution
-   Counts of occurrences of $K$ different types: the **multinomial** distribution
-   **Times** between $r$ events: the **gamma** distribution, which is equivilent to the exponential when $r=1$

You will have the opportunity to implement such models in your homework and on the challenge assignment!

## Power analysis

When designing experiments or how best to collect data, it is best to think about the model you will fit and the kind of experiment you need to design in order to have sufficient power to detect an effect of interest. For example, if we thought that age affected the likelihood that a host dies of a disease, then we would likely fit a logistic regression-type model. By _simulating_ data from such a model, we can determine

- the minimum sample size required to reliably estimate an effect of a certain size
- the minimum sample size required to detect an effect of a certain size at a fixed significance level
- the minimum effect size that can be detected at a fixed significance level and sample size

### Example

Below is an example of a simulation in which binary disease data (death/no death; 0/1) are simulated hosts of various ages, assuming the the log-odds of disease is a linear function of age. (Notice that we assume that host lifetimes are exponentially distributed with rate parameter $\lambda = 1/10$ years. This means that hosts, in this simulation, live for an average of 10 years.) We then fit a logistic regression to this data to determine the effect of age on disease risk.

```{r}
data_generator &lt;- function(sample_size = 100, effect = 0.1){
  age &lt;- rexp(n = sample_size, rate = 1/10)
  linear_predictor &lt;- effect*age
  prob &lt;- 1/(1+exp(-linear_predictor))
  
  disease_status &lt;- c()
  
  for (i in 1:length(prob)){
  disease_status[i] &lt;- rbinom(n = 1, size = 1, prob = prob[i])
  }
  
  return(data.frame(age = age, disease_status = disease_status))
}

data &lt;- data_generator()

data %&gt;% pivot_longer(! age) %&gt;% 
  ggplot(aes(x = age, y = value)) + geom_point() + 
  geom_smooth(method = "glm", method.args = list(family = "binomial")
              ) + labs(y = "prob. of disease (i.e., disease status =1)")

model &lt;- glm(disease_status~scale(age), family = binomial, data = data)
summary(model)
```

Next, we will write a function that performs a **power analysis**. We will use this function determine the sample size that is required so that simulating the age-disease data repeatedly we identify a significant effect of age on disease status (i.e., reject the null hypothesis) at level $\alpha = 0.01$ *at least $95\%$ of the time*.

```{r}
power_analysis_function &lt;- function(sample_size){
  
  sims &lt;- 1000
  pvalues &lt;- c()
  for (i in 1:sims){
  data &lt;- data_generator(sample_size)
  model &lt;- glm(disease_status~scale(age), family = binomial, data = data)
  pvalues[i] &lt;- summary(model)$coefficients[2,4]
  }
  
  power_estimate &lt;- length(which(pvalues &lt; 0.01))/length(pvalues)
  
  return(power_estimate)
}

sample_sizes &lt;- seq(150,200,10); power_estimates &lt;- c()

for (i in 1:length(sample_sizes)){
  power_estimates[i] &lt;- power_analysis_function(sample_size = sample_sizes[i])
}

knitr::kable(cbind(n = sample_sizes, power = power_estimates))
```

## Dealing with dependent data!

To illustrate how mixed models work and what kinds of questions we can answer using them, we will use the sexual size dimorphism data which we analyzed last class. Recall that we did NOT find a significant effect of sex on the average body size. There was no effect of the interaction between Order and sex on body size. Indeed, this matches what you saw in the homework questions where you had to visualize the data --- most of the variation in body size was _between_ orders.

In the models we built, we ignored a pretty important fact about the data: species have a common history (i.e., phylogeny). Thus, observations are not independent! This can make drawing inferences from comparative data difficult. **We will address how to deal with the non-independence of data (due to a common history of species, replication in blocks, etc.) using three approaches.**

But, first, let's download the data we will use in this section!

```{r}
SSDdata &lt;- read_csv("data/SSDinMammals.csv")
```

```{r}
mammal_length &lt;- SSDdata %&gt;%
  select(c("Order", "Scientific_Name", "lengthM", "lengthF")) %&gt;%
  pivot_longer(c(lengthM, lengthF), names_to = "sex", values_to = "length",
               names_pattern = "length(.)")

mammal_mass &lt;- SSDdata %&gt;%
  select(c("Order", "Scientific_Name", "massM", "massF")) %&gt;%
  pivot_longer(c(massM, massF), names_to = "sex", values_to = "mass",
               names_pattern = "mass(.)")

mass_nodup &lt;- mammal_mass %&gt;% 
  group_by(Scientific_Name, sex) %&gt;%
  distinct(Scientific_Name, sex, .keep_all = TRUE)

length_nodup &lt;- mammal_length %&gt;% 
  group_by(Scientific_Name, sex) %&gt;%
  distinct(Scientific_Name, sex, .keep_all = TRUE)

mammal_long &lt;- full_join(mass_nodup, length_nodup, 
                         by = join_by("Scientific_Name", "sex", "Order")) |&gt;
  drop_na()

glimpse(mammal_long)
```

### Group-by-group analyses

**One first way we can handle dependent data is to split observations into groups such that, within each group, observations are independent (or approximately so).** This is what we did last class when we fit order-specific regression coefficients of sex on body size. We saw that ALL order-specific effects had confidence intervals which overlapped zero!

```{r, warning=F, message=F}
# run linear model of size on sex for EACH Order

Order_specific_models &lt;- 
  mammal_long |&gt; 
  group_by(Order) |&gt;
  do(model = tidy(lm(log(mass) ~ sex, data = .), conf.int = T))
  
# get coefficients for each Order

Order_specific_models |&gt;
  unnest() |&gt;
  select(Order, estimate)

# visualize effects, CIs, and p values

Order_specific_models |&gt;
  unnest() |&gt;
  subset(term == "sexM") |&gt;
  group_by(Order) |&gt;
  ggplot(aes(y = Order, x = estimate, 
             xmin = conf.low,
             xmax = conf.high
             )
         ) +
  geom_pointrange() +
  geom_vline(xintercept = 0, lty = 2)
```

No effect of sex on body size, for any of the orders!

#### Challenge

How would you adjust the previous plot to show the estimated intercepts AND the effects of sex?

```{r}
Order_specific_models |&gt;
  unnest() |&gt;
  # subset(term == "sexM") |&gt;
  group_by(Order) |&gt;
  ggplot(aes(y = Order, x = estimate, 
             xmin = conf.low,
             xmax = conf.high,
             color = term
             )
         ) +
  geom_pointrange() +
  geom_vline(xintercept = 0, lty = 2)
```

#### Challenge

How would you retrieve model fits for each Order in base R?

```{r, include=F}
for (i in 1:nrow(Order_specific_models)){
  print(summary(Order_specific_models$mod[[i]]))
}
```

#### Challenge

Adjust the plot above so that it the size of the estimates depend on the number of observations in an Order? _Hint: determine the number of observations per Order and then use the `merge()` function with the `Order_specific_models` tibble. To adjust the range of point sizes, use `scale_size()`._

```{r, include=F}
mammal_long |&gt; 
  group_by(Order) |&gt;
  summarise(number_obs = n()) |&gt;
  merge(Order_specific_models) |&gt;
  unnest() |&gt;
  ggplot(aes(y = Order, x = estimate, 
             xmin = conf.low,
             xmax = conf.high,
             color = term,
             size = number_obs
             )
         ) +
  geom_pointrange() +
  geom_vline(xintercept = 0, lty = 2) +
  scale_size(range = c(0, 2))
```

#### Pros and cons

There are several advantages to preforming a group-by-group analysis:

- **Fitting more complex models (e.g., those with random effects) can be difficult.** There may be convergence issues, and interpretation of effects and $p$-values can be tricky.
- The group-by-group analysis is **robust in the face of unbalanced data** (i.e., when there are more observations for some groups than others).
- **The conclusions from a group-by-group analysis are _conservative_.**
- The group-by-group analysis is **fairly easy to implement**.

Among the disparages to this approach are the following:

- With fewer samples per group, the analysis **may be under-powered**.
- Splitting the data into groups means there are more coefficients to estimate, and thus confidence intervals to be estimated and hypothesis tests to be performed. This means there is a greater chance that we obtain a **spurious association**.*
- It is **difficult to draw conclusions about the variance between groups.** In some applications, this is of interest. In others, it is not.

*One solution is to adjust the significant level based on the number of tests conducted. If $k$ hypotheses are tested, an [common adjustment](https://en.wikipedia.org/wiki/Bonferroni_correction) is to set $\alpha = 0.05/k$.

### Controlling for phylogeny

A common reason data are dependent in biology is that **species share a common history of descent with modification.** When we assume that the observations are independent, we make implicit assumptions about the degree to which the characters at the tips of a phylogeny have underwent independent evolution. Sometimes, when species are diverged by many millions of years and traits evolve quickly, it is reasonable to ignore the phylogenetic constraints on the data. In other cases, it is essential to consider the role of phylogeny.

Several methods can control for phylogeny (if known). In fact, such methods can _use_ information in the branching pattern of species to draw inferences about the evolution of ecologically-important traits (such as body and brain size, dispersal rate, etc.). We will not discuss how to implement phylogenetic comparative methods, but it is good to know they exist and how, at a surface level, they work.

Intuition for how PCMs work is easiest to grasp when we consider a tree with $n$ species at the tips. If we know the pairwise divergence times for all species, then we can _transform_ the data so that observations are independent by looking at all _differences_ of traits. Not all differences are equally informative; if species have diverged a long time ago, there has been more time for differences to build up. PCMs account for this by specifying how the distribution of trait differences between species $i$ and $j$ depends on the time that has elapsed since these species diverged, In particular, the larger the divergence time, the greater the variance in $Y_i-Y_j$, the difference in trait values between species $i$ and $j$. A simple way to do this is to write $\sigma_{ij} = \sigma^2/T_{ij}$.^[Note: this gives rise to a likelihood function that is functionally VERY similar to the one for the linear model with constant error variance.]

### Random effects!

Another way one can account for dependent data is by including _random effects_. Random effects are often used to control for the fact that observations are clustered (e.g., trait data for species belonging to a higher taxonomic unit, measurements of plant biomass from plots of land).

#### Structure and assumptions

A common way models with random effects are formulated is as follows:

$$Y_{ij} \sim \text{Normal}(\beta_0 + \beta_1 x_{1ij} + \cdots + \beta_{p} x_{pij} + b_{0i} + b_{1i} z_{1ij} + \cdots + b_{qi} z_{qij}, \sigma_{ij}^2),$$

where $ij$ is the $i$th observation from the $j$th cluster and

$$b_{ki} \sim \text{Normal}(0,\tau_k^2).$$

This gives rise to a distribution for $Y$ which depends on the values of the random effects $b_0,\dots,b_q$. Under the hood, methods that fit models of this form numerically maximize a version of the likelihood function that results from these assumptions.

#### Challenge

In each of the following examples, which effects might be reasonably treated as fixed vs. random? Justify your answer.

1. Suppose we are working with rodents that have been infected with an evolved strain of the parasite that causes malaria. Some rodents have been treated with a prospective vaccine and others sham-vaccinated. We are interested if a proxy for virulence (e.g., density of infected red blood cells) depends on vaccination status.

2. Suppose we conduct the same experiment, except rodents are caged are sets of four.

3. Suppose we measure the time between sightings of a raccoon in a Toronto neighborhood using six camera traps (strategically placed in the neighborhood). We do this for year, and want to ask if mean daily temperature predicts the frequency of raccoon occurrence.

4. Suppose that, every year, we go to the Amazon and measure the number of bird calls that come from $n=30$ trees. Assume that the trees are far enough apart we can treat them as independent. We measure the number of birds in a tree, characteristics of the tree (cover, height, width, age), the year in which a measurement was done, and the mean temperature that year. We want to know if tree cover affects the frequency of bird calls, and if it interacts with temperature.

For more on the difference between fixed and random effects, check out the following resources

- [this Cross Validated post](https://stats.stackexchange.com/questions/4700/what-is-the-difference-between-fixed-effect-random-effect-in-mixed-effect-model)
- [this Dynamic Ecology post](https://dynamicecology.wordpress.com/2015/11/04/is-it-a-fixed-or-random-effect/)
- [this book chapter](https://bookdown.org/steve_midway/DAR/random-effects.html)

#### Using `lme4` to fit random and mixed effect models

We will start by fitting a model of log body size on log length where the intercept is random depending on the Order. Based on a quick visualization of the data, such a model may be appropriate.

```{r}
ggplot(data = mammal_long, aes(y = log(mass), 
                               x = log(length)
                               )
       ) + 
    geom_point(aes(color = Order), size = 3) -&gt; p

p 
```

```{r, include=F}
# no Order-dependence:
p + geom_smooth(method = "lm")

# intercept AND slope fixed and specific to the Order:
p + geom_smooth(method = "lm") + facet_wrap(~Order) 
```

To fit regress the response on a set of fixed effects with random _intercepts_ that depends the values assumed by a random effect `x`, we call `lmer` and write a linear model with `(1|x)`.

```{r}
## random intercept for Order

model &lt;- lmer(formula = log(mass) ~ log(length) + (1|Order), data = mammal_long)
summary(model)

fixef(model)
ranef(model)
```

From top to bottom, the output is telling us that the model was fit using a method called REML (restricted maximum likelihood). It returns information about the residuals, random effects, and fixed effects. The output also tells us about the **estimated variance for the random effects** in the model. Here, the variance associated with Order is 4.256. The variance explained by Order is 4.256/(4.256+0.847) _after controlling for the fixed effects_. Note the denominator here is the total variance, including from the residuals. As usual, we also have information about the fixed effects.

To visualize the model, we can predict the overall and Order-specific relationship of log mass on log length for all of the Orders represented in the data.

```{r}
mammal_long |&gt; ungroup() |&gt; select(mass, length, Order) |&gt; 
  mutate(fit.m = predict(model, re.form = NA), # does not include random effects
         fit.c = predict(model, re.form = NULL) # includes random effects
         ) -&gt;
  predicted_values

predicted_values |&gt;
  ggplot(aes(x = log(length), y = log(mass), color = Order)) +
  geom_point(size = 3) +
  geom_line(inherit.aes = F, aes(x = log(length), y = fit.c, color = Order), size = 1) +
  geom_line(inherit.aes = F, aes(x = log(length), y = fit.m), color = "black", size = 2)
```

The thick black line corresponds to the fitted values associated with the fixed-effect component of the model. The colored lines correspond to the fitted values estimated for each Order. 

Perhaps a model with Order-specific random intercepts AND slopes would be better. We fit this model using the code below. The key difference in syntax is that we write `(1+fixed effect|random effect)` to indicate that the random effect has an effect on both the intercept _and_ slope of the response on the fixed effect.

```{r}
## random intercept and slope for Order

model2 &lt;- lmer(formula = log(mass) ~ log(length) + (1+log(length)|Order), data = mammal_long)
summary(model2)

fixef(model2)
ranef(model2)
```

```{r}
mammal_long |&gt; ungroup() |&gt; select(mass, length, Order) |&gt; 
  mutate(fit.m = predict(model2, re.form = NA),
         fit.c = predict(model2, re.form = NULL)
         ) -&gt;
  predicted_values

predicted_values |&gt;
  ggplot(aes(x = log(length), y = log(mass), color = Order)) +
  geom_point(size = 3) +
  geom_line(inherit.aes = F, aes(x = log(length), y = fit.c, color = Order), size = 1) +
  geom_line(inherit.aes = F, aes(x = log(length), y = fit.m), color = "black", size = 2)
```

Why does this look like a mess? As stated above, **mixed models do not work well when datasets are unbalanced.** Indeed, the number of observations in the different Orders are quite variable.

#### Challenge

Regress log body size on sex with Order as a random effect that affects the intercept AND slope of the sex-size relationship.

```{r, include=F}
Model &lt;- lmer(log(mass) ~ sex + (1+sex|Order), mammal_long)

mammal_long |&gt; ungroup() |&gt; select(Scientific_Name, mass, sex, Order) |&gt; 
  mutate(fit.m = predict(Model, re.form = NA),
         fit.c = predict(Model, re.form = NULL)
         ) -&gt;
  predicted_values

predicted_values |&gt;
  ggplot(aes(x = sex, y = log(mass))) +
  geom_line(aes(group = Scientific_Name), color = "lightgrey") +
  geom_point(color = "lightgrey", size = 2) +
  geom_point(inherit.aes = F, aes(x = sex, y = fit.c, color = Order), size = 3) +
  geom_line(inherit.aes = F, aes(x = sex, y = fit.c, color = Order, group = Order), size = 2) +
  geom_point(inherit.aes = F, aes(x = sex, y = fit.m), color = "black", size = 3) +
  geom_line(inherit.aes = F, aes(x = sex, y = fit.m, group = ""), color = "black", size = 3) +
  theme_classic()
```</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Assignments</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignments/assignment-00.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment00: individual interest description</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignments/assignment-01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 01: base R, command line, &amp; Git(Hub)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignments/assignment-02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 02: data wrangling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignments/assignment-03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 03: data visualisation and exploration</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignments/assignment-04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 04: inference!</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./projects.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Project description</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./projects_databases.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Some open-access databases</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Datasets</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/data/portal_data.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">portal_data.csv</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/data/pseudo.ara.busco" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pseudo.ara.busco</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/data/pseudo.LTRs" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pseudo.LTRs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/data/disease_distance.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">disease_distance.csv</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/data/SSDinMammals.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SSDinMammals.csv</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lectures/data/meas.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">meas.csv</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text">Miscellaneous notes</span></span>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#option-1-hypothesis-driven-project" id="toc-option-1-hypothesis-driven-project" class="nav-link active" data-scroll-target="#option-1-hypothesis-driven-project">Option 1: Hypothesis-driven project</a>
  <ul class="collapse">
  <li><a href="#what-is-a-hypotheses-what-is-a-prediction" id="toc-what-is-a-hypotheses-what-is-a-prediction" class="nav-link" data-scroll-target="#what-is-a-hypotheses-what-is-a-prediction">What is a hypotheses? What is a prediction?</a></li>
  </ul></li>
  <li><a href="#option-2-modeling" id="toc-option-2-modeling" class="nav-link" data-scroll-target="#option-2-modeling">Option 2: Modeling</a></li>
  <li><a href="#option-3-simulation-study" id="toc-option-3-simulation-study" class="nav-link" data-scroll-target="#option-3-simulation-study">Option 3: Simulation study</a></li>
  <li><a href="#project-timeline-and-deliverables" id="toc-project-timeline-and-deliverables" class="nav-link" data-scroll-target="#project-timeline-and-deliverables">Project timeline and deliverables</a>
  <ul class="collapse">
  <li><a href="#individual-interest-description" id="toc-individual-interest-description" class="nav-link" data-scroll-target="#individual-interest-description">Individual interest description</a></li>
  <li><a href="#project-proposal" id="toc-project-proposal" class="nav-link" data-scroll-target="#project-proposal">Project proposal</a></li>
  <li><a href="#mid-project-update" id="toc-mid-project-update" class="nav-link" data-scroll-target="#mid-project-update">Mid-project update</a></li>
  <li><a href="#presentation" id="toc-presentation" class="nav-link" data-scroll-target="#presentation">Presentation</a></li>
  <li><a href="#report" id="toc-report" class="nav-link" data-scroll-target="#report">Report</a></li>
  </ul></li>
  <li><a href="#project-rubric" id="toc-project-rubric" class="nav-link" data-scroll-target="#project-rubric">Project rubric</a>
  <ul class="collapse">
  <li><a href="#individual-interest-description-1" id="toc-individual-interest-description-1" class="nav-link" data-scroll-target="#individual-interest-description-1">Individual interest description</a></li>
  <li><a href="#project-proposal-1" id="toc-project-proposal-1" class="nav-link" data-scroll-target="#project-proposal-1">Project proposal</a></li>
  <li><a href="#mid-project-update-1" id="toc-mid-project-update-1" class="nav-link" data-scroll-target="#mid-project-update-1">Mid-project update</a></li>
  <li><a href="#the-presentation" id="toc-the-presentation" class="nav-link" data-scroll-target="#the-presentation">The presentation</a></li>
  <li><a href="#the-report" id="toc-the-report" class="nav-link" data-scroll-target="#the-report">The report</a></li>
  </ul></li>
  <li><a href="#tips-on-writingpresenting-a-research-project" id="toc-tips-on-writingpresenting-a-research-project" class="nav-link" data-scroll-target="#tips-on-writingpresenting-a-research-project">Tips on writing/presenting a research project</a></li>
  <li><a href="#miscellaneous" id="toc-miscellaneous" class="nav-link" data-scroll-target="#miscellaneous">Miscellaneous</a>
  <ul class="collapse">
  <li><a href="#issues-working-in-groups" id="toc-issues-working-in-groups" class="nav-link" data-scroll-target="#issues-working-in-groups">Issues working in groups</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./projects.html">Project</a></li><li class="breadcrumb-item"><a href="./projects.html">Project description</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Project description</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<hr>
<p>The course project is a self-directed group project that may take a couple forms (data analysis, modelling, or simulation study), but will make use of technical and scientific tools that we have learned in class. Aside from a couple constraints, groups are free to do their projects on any subject within ecology and evolution that they find interesting. The forms the project may take are:</p>
<ol type="1">
<li>A hypothesis-driven analysis of existing ecological or evolutionary data.</li>
<li>Development and analysis of a mathematical model.</li>
<li>Simulate data and analyze it using appropriate tools.</li>
</ol>
<p>In order to utilize conventional data science collaboration tools, students are required to use <a href="https://github.com">Github</a> for code-sharing, project management, and individual contributions. All sections of the project will be submitted via Github, not Quercus. Everyone in your group will need to make a GitHub account prior to the first project deadline. We will cover the basics of GitHub in advance so that you are prepared to use it.</p>
<section id="option-1-hypothesis-driven-project" class="level2">
<h2 class="anchored" data-anchor-id="option-1-hypothesis-driven-project">Option 1: Hypothesis-driven project</h2>
<p>Groups will formulate their own hypotheses based on their interests within ecology and evolution. Groups will test predictions borne out of their hypotheses with reproducible and quantitative analysis techniques (e.g., ANOVA). If your group has an idea for statistical analyses that are beyond the scope of the course, please let us know. We are happy to support any groups who want to learn new tools, but expect that these groups are ready to learn how these tools work on their own; we hope to equip you with enough understanding to learn new things independently. Finally, the work must be original â€“ while we may be repurposing data, we will not be simply redoing analyses. Keep in mind also that any work you do as part of this course may not be submitted for credit in another course (such as a fourth-year research project) and vice versa. While you may not submit your work for this course for credit in another course, you are welcome to publish or present your work in an academic setting. (In fact, if you do so, please let us know!)</p>
<p>A note about community/citizen science websites: since the data is community-controlled, it may not always be research quality. There may be incorrect species IDs, inaccurate geolocations or time of observations, or discrepancies in protocols. When working with community science data, make sure that the data is cleaned and wrangled so that it is reliable. Quality control is a good first step when working with data, as simple errors can exist in any dataset.</p>
<section id="what-is-a-hypotheses-what-is-a-prediction" class="level3">
<h3 class="anchored" data-anchor-id="what-is-a-hypotheses-what-is-a-prediction">What is a hypotheses? What is a prediction?</h3>
<p>A <strong>hypothesis</strong> is a testable and falsifiable statement that offers a possible explanation of a phenomenon based on background knowledge, preliminary observations, or logic.</p>
<p><em>E.g., Primary productivity is an important driver of mammal species richness.</em></p>
<p>A <strong>prediction</strong> is based on a hypothesis. It is meant to describe what will happen in a specific situation, such as during an experiment, if the hypothesis is correct.</p>
<p><em>E.g., If primary productivity is an important driver of mammal species richness, then more mammalian species would be found in sites with more plant biomass (proxy for primary productivity) compared with sites with less plant growth.</em></p>
</section>
</section>
<section id="option-2-modeling" class="level2">
<h2 class="anchored" data-anchor-id="option-2-modeling">Option 2: Modeling</h2>
<p>Groups will develop a mathematical model to answer a question in ecology and/or evolution they find interesting. There are many reasons to develop models: they help clarify assumptions, generate predictions, nullify hypotheses, provide mechanistic explanations for observed data, and help us know <em>what kinds of data</em> to look for. New models almost always build on existing and well-studied ones (e.g., the Lotka-Volterra model). The fact models are simplifying representations of the real world is by design! The goal of building a model is to identify the key features that make a process interesting, represent the process mathematically (and, in doing so, clarify what assumptions are being made!), characterize the behaviour of the model, and from this characterization draw conclusions about how the process being modelled works. Characterization of a model can involve mathematical analysis, simulation, and confrontation with data.</p>
<p>The key steps in this project are to 1) identify an interesting question in ecology or evolution, 2) develop (and likely revise) a model to address that question, 3) characterize the behaviour of the model, and 4) draw biological conclusions from the model and its characterization.</p>
<p>If you are interested in modeling, let Mete and ZoÃ« know as soon possible!</p>
</section>
<section id="option-3-simulation-study" class="level2">
<h2 class="anchored" data-anchor-id="option-3-simulation-study">Option 3: Simulation study</h2>
<p>Similar to Option 1, groups that do a simulation study will formulate hypotheses and use reproducible and quantitative analysis techniques to test predictions borne out of those hypotheses. The difference is that students will <em>simulate</em> their own data, instead of using an existing dataset. One reason to do a simulation study is to see what kind of data would be needed to test a hypothesis in the field, e.g., how much data would be needed to find a significant association between response and predictor variables.</p>
<p>If you are interested in doing a simulation study, let Mete and ZoÃ« know as soon possible!</p>
</section>
<section id="project-timeline-and-deliverables" class="level1">
<h1>Project timeline and deliverables</h1>
<p>As instructors, we are here to help your group work towards a project idea that you are excited about â€“ and to help you execute that idea to the fullest! We have included multiple check points and small assignments throughout the semester for you to find the right group-mates, get feedback on your project ideas, ask us question, and make sure the analysis is on-track. By September 17, you will need to provide us with a short description of your interests in ecology and evolution. September 26 will be devoted to identifying a group and begin the project proposal. All classes after October 24 will be devoted to working on projects with your groups. We (Mete and ZoÃ«) will be around during those times provide support, if needed.</p>
<section id="individual-interest-description" class="level2">
<h2 class="anchored" data-anchor-id="individual-interest-description">Individual interest description</h2>
<p><em>Due Sept 12th</em>, worth 1% of final grade</p>
<p>To make sure you pair up with group mates who share common interests, we ask that you write a short (4-5 sentences) description of your interests in ecology and evolution. Please do so in RStudio by creating an .Rmd file and knitting it into a pdf.</p>
<p>Here are some discussion questions to help you:</p>
<ul>
<li>What is a scientific paper or popular science article you read (or a podcast you listened to) recently that you found interesting (how microbial communities differ among environments, frequency of herbicide resistance alleles in weed populations, bird species richness in regions that have experienced climatic shifts, understanding the relationship between longevity and traits like body size in miceâ€¦)?</li>
<li>What is your favourite EEB course so far? Why did you like it?</li>
<li>Thinking about EEB professors, was there anyone whose work you are particularly interested in?</li>
<li>Browse through some recent issues of broad scope EEB journals such as <em>Trends in Ecology and Evolution</em> and <em>Annual Review of Ecology, Evolution, and Systematics</em>. Any articles catching your eyes?</li>
<li>Check out <a href="https://besjournals.onlinelibrary.wiley.com/doi/10.1111/1365-2745.12025">this paper</a>. Any of those questions spark your interest?</li>
</ul>
</section>
<section id="project-proposal" class="level2">
<h2 class="anchored" data-anchor-id="project-proposal">Project proposal</h2>
<p><em>Due Oct 17rd</em>, worth 3% of final grade</p>
<p>Good research takes time! The purpose of the proposal is to get your group started on this process early on so that you will have sufficient time to do your project justice. This will also serve as official documentation of your project development process. Your projects will likely evolve over time, and there can be many reasons for this. For instance, as you explore your data, you might be inspired to ask different questions, or you may need to refine your hypotheses due to limitations in the data.</p>
<p>Include the following information in your proposal:</p>
<ul>
<li>Option 1: your hypotheses and predictions (point form or short paragraph) and data source (short paragraph). Include a citation, a brief description of how the data was collection, and which section of the dataset you plan to use in your analysis (e.g., which columns).</li>
<li>Option 2: a question you want to answer using a mathematical model (short paragraph describing the problem and the value modeling may add). Be sure to include a description of the variables that you may want to track and the kind of model you envision using.</li>
<li>Option 3: same as 1, except with a description of how to simulate the data.</li>
</ul>
</section>
<section id="mid-project-update" class="level2">
<h2 class="anchored" data-anchor-id="mid-project-update">Mid-project update</h2>
<p><em>Due Nov 21st, worth 6% of final grade</em></p>
<p>The purpose of the mid-project update is to ensure you are on track with your projects. By now, you should have completed your exploratory data analyses, modelling, or simulation. You should have also solidified your hypotheses, predictions, and analyses plan (i.e., the Methods section of your final report!).</p>
<p>Include the following information in your mid-project report:</p>
<ol type="1">
<li>Options 1 and 3:</li>
</ol>
<ul>
<li>Your hypotheses and predictions (point form or short paragraph). If these differ from the ones in your proposal, explain clearly the rationale for the change.</li>
<li>A detailed description of your data (a paragraph), including how the data was collected or simulated, along with any manipulation(s) you performed to get your data ready for the analysis.</li>
<li>Your analysis plan (a paragraph): describe the statistical test(s) that you will use to test <em>each</em> prediction, including how you will validate the assumptions of each test.</li>
</ul>
<ol start="2" type="1">
<li>Option 2:</li>
</ol>
<ul>
<li>A detailed description of the question you want to answer, any previous work (modelling and otherwise), the model you have built to answer this question, and your modelling assumptions.</li>
<li>Detailed descriptions of the model analysis and biological interpretations of the results so far.</li>
<li>Your analysis plan (a paragraph): describe additional analysis that you will do and any assumptions you would like to relax.</li>
</ul>
</section>
<section id="presentation" class="level2">
<h2 class="anchored" data-anchor-id="presentation">Presentation</h2>
<p><em>Due Dec 3rd, worth 10% of final grade</em></p>
<p>The presentations will be held on the last day of class during regular class hours (<em>Dec 3rd</em>, 2-4 pm). Each presentation will be 10 minutes long, followed by 2 minutes of questions from the audience. If you cannot make it to class, please get in touch with us to make alternative arrangements no later than Nov 28th.</p>
</section>
<section id="report" class="level2">
<h2 class="anchored" data-anchor-id="report">Report</h2>
<p><em>Due Dec 10th, worth 20% of final grade</em></p>
<p>This report will be styled as a journal article, with these sections:</p>
<ul>
<li>Abstract</li>
<li>Introduction</li>
<li>Methods (including â€œData Descriptionâ€ and â€œData Analysisâ€ subsections)</li>
<li>Results</li>
<li>Discussion</li>
<li>References</li>
<li>Supplementary material consisting of data and code required to reproduce analysis</li>
</ul>
<p><strong>For your sake (and ours), we are enforcing a two page limit (single spaced, excluding figures, tables, code, references, and appendices). Please use a standard font, size 12, with regular margins. One goal of this assignment is to write clearly and concisely â€“ it is often clarifying to put your analyses in as few words as possible.</strong></p>
<p>For the report, you are expected to:</p>
<ul>
<li>Put your research questions in the context of existing research and literature.</li>
<li>Have clear and explicit objectives, hypotheses, and/or predictions.</li>
<li>Adequately describe and properly cite the data source(s) you will analyze. If your project involves modeling, describe other modeling work that is relevant.</li>
<li>Describe your analysis in sufficient detail for others to understand.</li>
<li>Discuss the interpretation of your results and their implications.</li>
</ul>
<p>The data and code associated with your report is expected to be entirely reproducible. Your supplementary files must include the following:</p>
<ul>
<li>A description of what <em>every</em> column/row in your submitted data file.</li>
<li>A well-annotated R script or R notebook file. We must be able to run your code once you submit the project. <a href="https://swcarpentry.github.io/r-novice-inflammation/06-best-practices-R/">This lesson</a> on best practices for writing R code is a good starting place. Also check out this <a href="https://style.tidyverse.org/index.html">coding style guide</a> and <a href="https://gist.github.com/wojteklu/73c6914cc446146b8b533c0988cf8d29">these simple rules</a> on how to write code that is easy to read.</li>
</ul>
<p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5210691/">Hermann et al.&nbsp;2016</a> is a great example of what we expect your code to look like. Refer to their supplementary materials for examples of <a href="https://figshare.com/articles/LTER_Lampyrid_data/2068098">how to describe your data set</a> and <a href="https://github.com/cbahlai/lampyrid/blob/master/lampyrid_analysis.R">how to annotate your code</a>.</p>
</section>
</section>
<section id="project-rubric" class="level1">
<h1>Project rubric</h1>
<section id="individual-interest-description-1" class="level2">
<h2 class="anchored" data-anchor-id="individual-interest-description-1">Individual interest description</h2>
<p><em>1 marks total</em></p>
<p>This part of the project will be graded based on completion. That said, it will help determine your group-mates. Make sure to spend some time on it and reflect on what questions in EEB you would like to work on.</p>
</section>
<section id="project-proposal-1" class="level2">
<h2 class="anchored" data-anchor-id="project-proposal-1">Project proposal</h2>
<p><em>3 marks total</em></p>
<p>Option 1: Two marks for your hypotheses and associated predictions, and one mark for a description of your data source(s). Students are expected to demonstrate effort in formulating hypotheses and predictions, and identifying a suitable dataset.</p>
<p>Option 2: 1.5 marks each for 1) a clear description of the question or problem in ecology or evolution you would like to address using a model, and 2) a description of the kind of model you envision using, including what variables to track.</p>
<p>Option 3: One mark for your hypotheses and associated predictions, and two marks for describing the appropriate analyses.</p>
<p>These components will be graded mostly on completion. The purpose of this assignment is to ensure you start early and are heading towards the right track.</p>
</section>
<section id="mid-project-update-1" class="level2">
<h2 class="anchored" data-anchor-id="mid-project-update-1">Mid-project update</h2>
<p><em>6 marks total</em></p>
<p>Options 1 and 3: Three marks are given to clearly stating hypotheses and predictions. In the case that these are different from those in the proposal, the rationale for refinement needs to be clearly explained.</p>
<p>Each of the following criteria are scored out of 3: 3 == excellent, 2 == good, 1 == acceptable, but needs improvement.</p>
<ol type="1">
<li>Data description
<ul>
<li>The data source(s) are sufficiently described, specifically, where was the obtained and how it was originally collected.</li>
<li>The data is sufficient described, including any initial observations from your exploratory data analyses.</li>
<li>The suitability of the data is justified.</li>
<li>Any manipulations done to the data are thoroughly explained and well-justified.</li>
</ul></li>
<li>Data analysis plan
<ul>
<li>Clearly lay out the statistical test(s) you will use to test each prediction.</li>
<li>State how you will validating assumptions associated with each statistical test.</li>
</ul></li>
</ol>
<p>Option 2: Each of the following criteria are scored out of 3: 3 == excellent, 2 == good, 1 == acceptable, but needs improvement.</p>
<ol type="1">
<li>Description of question, previous work, the model, modeling assumptions, and any predictions you have ahead of the analysis
<ul>
<li>The question you want to address and previous work in that direction (modeling or otherwise) is described in detail.</li>
<li>The relationship between the question/problem and modeling approach is clear and well-justified.</li>
<li>Modeling assumptions and choices (including limitations) are clear and well-motivated.</li>
<li>Predictions for how the model will behave, what it might have to say about the question/problem, etc. are inclued and well thought out.</li>
</ul></li>
<li>Analysis and analysis plan
<ul>
<li>The details of all analysis (mathematical or computational) are explained clearly.</li>
<li>The biological interpretations of results so far are clearly presented and their validity/applicability is discussed.</li>
<li>Clearly lay out plans for remaining analysis (e.g., relaxing model assumptions) and justify why they are reasonable.</li>
</ul></li>
</ol>
</section>
<section id="the-presentation" class="level2">
<h2 class="anchored" data-anchor-id="the-presentation">The presentation</h2>
<p><em>10 marks total</em></p>
<p>Each of the following criteria are scored out of 3: 3 == excellent, 2 == adequate, 1 == needs improvement.</p>
<ol type="1">
<li>Content â€“ background and methods
<ul>
<li>The context for the study, along with hypotheses and predictions, are clearly set up.</li>
<li>Data source(s), manipulations, and statistical tests used are succinctly and adequately described.</li>
<li>If modeling, the relationship between the question/problem addressed and modeling approach is well-explained, and previous work (modeling or otherwise) is discussed.</li>
</ul></li>
<li>Content â€“ results and conclusions
<ul>
<li>Results are accurately described and interpreted, with particular attention to how they related to the hypotheses and predictions the group set out to test.</li>
<li>The conclusion to the study is succinct and clear.</li>
</ul></li>
<li>Delivery
<ul>
<li>All students participated in presenting the information.</li>
<li>All students spoke clearly and without jargon.</li>
<li>The presentation is well organized and ideas flowed naturally from one to the next.</li>
<li>The presentation is well rehearsed and is an appropriate length.</li>
<li>Figures are easy to read (e.g., axis labels are big enough to read and are informative) and are explained thoroughly (e.g., x and y axis and what each data point is).</li>
</ul></li>
</ol>
<p>The final 1 mark will be assigned to the question period, and students will be assessed on whether they are able to answer questions thoughtfully.</p>
</section>
<section id="the-report" class="level2">
<h2 class="anchored" data-anchor-id="the-report">The report</h2>
<p><em>20 marks total</em></p>
<p>Each of the following criteria are scored out of 4: 4 == excellent, 3 == good, 2 == acceptable, 1 == needs improvement.</p>
<ol type="1">
<li>Content and concepts
<ul>
<li>Authors demonstrate a full understanding of the existing literature on the topic, and these concepts are critically integrated into their own insights.</li>
<li>Options 1 and 3: Hypotheses and predictions are clearly defined, and rational for choosing/simulating this data is justified.</li>
<li>Option 2: The question, modeling approach, and relevant work are thoughtfully explained; the rationale for using the model (and its assumptions) is justified.</li>
</ul></li>
<li>Communication
<ul>
<li>Writing is succinct, clear, logical, and free of grammatical and spelling errors.</li>
</ul></li>
<li>Analysis: see below.</li>
<li>Results
<ul>
<li>Results are accurately and sufficiently described.</li>
<li>Conclusions are supported by evidence.</li>
<li>Figures and tables are clearly presented and are informative.</li>
</ul></li>
<li>Coding style and reproducibility
<ul>
<li>Data and code are well-organized and well-documented.</li>
<li>The analysis is easily reproducible.</li>
<li>All team members have pushed to a common GitHub repo.</li>
</ul></li>
</ol>
<p><em>Note:</em> marks for the 3rd criterion (Analysis) depend on if groups did a modeling or data-driven project:</p>
<p>Options 1 and 3: Statistical analysis</p>
<ul>
<li>Statistical tests chosen or modeling choices made are appropriate.</li>
<li>Assumptions for each statistical test is validated.</li>
<li>Limitations in the data and analysis are discussed.</li>
</ul>
<p>Option 2: Analysis of model</p>
<ul>
<li>Characterization of the model is appropriate and explained in detail.</li>
<li>Importantly, biological conclusions explained in detail and in terms of the processes described (or not described) by the model.</li>
<li>Limitations of modeling assumptions are discussed, and extensions are proposed.</li>
</ul>
<p>Please note that we are only going to be marking the two pages of your report. <strong>Please do not go over the page limit (with the exception of tables, figures, references, and appendices).</strong></p>
</section>
</section>
<section id="tips-on-writingpresenting-a-research-project" class="level1">
<h1>Tips on writing/presenting a research project</h1>
<p>We know that students have very unique research interests and ideas, and we hope that your project encapsulates that! As instructors, we do not know everything, but we are excited to learn from you and your projects. Below are some tips that we have gathered that you may find helpful when preparing for the project presentation and writing your report.</p>
<ol type="1">
<li>Use a title that summarizes your project/results clearly.</li>
<li>Define everything! Do not assume that we know about your question, study system, etc. For your presentations, adding some pictures will help when you are defining something.</li>
<li>After introducing your study system, tell us clearly your hypothesis and prediction: â€œI hypothesize that there are more mosquitoes in the boreal forest because it is warmer. I predict this because insects have a thermal toleranceâ€. Then, after your methods, results, etc., remind us of your hypothesis again! For your presentation, you can even show the same slide you used for your hypothesis with a big red X or a big green checkmark. Assume we forgot and that we know nothing about the system.</li>
<li>NEVER USE THE WORD â€œproveâ€. Science cannot prove or disprove anything â€” the evidence can only support (or fail to support) how we think the world works.</li>
<li>Use an appropriate font and font size. Also, use colours wisely (e.g., avoid red and blue together because of folks that are colourblind).</li>
<li>A 10-minute presentation is about 10 slides (more or less depending on if you use animations). A note about animations: use â€œAppearâ€, not any of the fancy stuff. And no slide transitions!</li>
<li>We will ask questions after your presentation, but we are not trying to trick you â€” we just want more information. Give us your best answer, and remember that itâ€™s okay to say â€œI donâ€™t know, but I think thatâ€¦â€ or â€œI can test this further by doing thisâ€. At this point, you should know more about your projects than we do. Also, when preparing for the presentation, it useful to think about what questions listeners may have and try to answer them preemptively.</li>
<li>Practice your presentation at least once with your group! Itâ€™ll get rid of any nerves you have if you already know the words you are going to say. Itâ€™ll also help you ensure that you speak louder and slower. We know you all will do great projects, and we are excited to hear about them!</li>
</ol>
<p>Reading widely and often is one of the best ways to learn how to write well. Here are some papers which we think are clear, concise, and free of grammatical and logical flaws.</p>
<ul>
<li><a href="https://www.pnas.org/doi/abs/10.1073/pnas.1919176117?url_ver=Z39.88-2003&amp;rfr_id=ori:rid:crossref.org&amp;rfr_dat=cr_pub%20%200pubmed">Viral zoonotic risk is homogenous among taxonomic orders of mammalian and avian reservoir hosts</a></li>
<li><a href="https://academic.oup.com/evolut/article/76/1/29/6728984">Estimation of the strength of mate preference from mated pairs observed in the wild</a></li>
<li><a href="https://www.nature.com/articles/nature02104">The role of evolution in the emergence of infectious diseases</a></li>
<li><a href="https://www.pnas.org/doi/abs/10.1073/pnas.1508397112">Coevolution of parasite virulence and host mating strategies</a></li>
<li><a href="https://www.pnas.org/doi/abs/10.1073/pnas.1817482116">A rigorous measure of genome-wide genetic shuffling that takes into account crossover positions and Mendelâ€™s second law</a></li>
<li><a href="https://www.science.org/doi/10.1126/science.abo7719">The role of divergent ecological adaptation during allopatric speciation in vertebrates</a></li>
</ul>
</section>
<section id="miscellaneous" class="level1">
<h1>Miscellaneous</h1>
<section id="issues-working-in-groups" class="level2">
<h2 class="anchored" data-anchor-id="issues-working-in-groups">Issues working in groups</h2>
<p>If you are having trouble working with your group, e.g., because you feel like the work is not being equitably divided, please let us know as soon as possible. We will work with you and your group to identify a solution that works for everyone. Do not wait until the last minute to let us know that your group has been having trouble â€“ at that point, there is little we can do to fix the situation. Moreover, we expect group members to communicate with each other and to try to work out their concerns before we become involved.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./assignments/assignment-04.html" class="pagination-link" aria-label="Assignment 04: inference!">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Assignment 04: inference!</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./projects_databases.html" class="pagination-link" aria-label="Some open-access databases">
        <span class="nav-page-text">Some open-access databases</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>