<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>EEB313: Quantitative Methods in R for Biology - 9&nbsp; Introduction to inference II</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../assignments/assignment-00.html" rel="next">
<link href="../lectures/lec08-intro-inference-1.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lectures/lec01-r_and_rstudio.html">Lectures</a></li><li class="breadcrumb-item"><a href="../lectures/lec09-intro-inference-2.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to inference II</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">EEB313: Quantitative Methods in R for Biology</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/EEB313/eeb313.github.io" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Syllabus</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../about-us.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024 teaching team</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../downloadingR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Installing R &amp; Ubuntu</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Lectures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec01-r_and_rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to the course</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec02-basic-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to R: assignment, vectors, functions, strings, loops</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec03-command_git.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Command line &amp; Git(Hub)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec04-data-wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data wrangling!</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec05-data-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data visualization with ggplot2</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec06-exploratory-data-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploratory data analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec07-self-directed-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Review using Farrell and Davies (2019) dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec08-intro-inference-1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to inference I</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec09-intro-inference-2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to inference II</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Assignments</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-00.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment00: individual interest description</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 01: base R, command line, &amp; Git(Hub)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 02: data wrangling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 03: data visualisation and exploration</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 4: Inference!</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../projects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Project description</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../projects_databases.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Some open-access databases</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Datasets</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/portal_data.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">portal_data.csv</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/pseudo.ara.busco" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pseudo.ara.busco</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/pseudo.LTRs" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pseudo.LTRs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/disease_distance.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">disease_distance.csv</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/SSDinMammals.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SSDinMammals.csv</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/meas.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">meas.csv</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text">Miscellaneous notes</span></span>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#lesson-preamble" id="toc-lesson-preamble" class="nav-link active" data-scroll-target="#lesson-preamble"><span class="header-section-number">9.1</span> Lesson preamble</a></li>
  <li><a href="#review-of-the-likelihood-function" id="toc-review-of-the-likelihood-function" class="nav-link" data-scroll-target="#review-of-the-likelihood-function"><span class="header-section-number">9.2</span> Review of the likelihood function</a></li>
  <li><a href="#numerical-maximization-of-the-log-likelihood" id="toc-numerical-maximization-of-the-log-likelihood" class="nav-link" data-scroll-target="#numerical-maximization-of-the-log-likelihood"><span class="header-section-number">9.3</span> Numerical maximization of the (log-)likelihood</a>
  <ul class="collapse">
  <li><a href="#step-1-specify-the-likelihood-i.e.-the-distribution-of-the-data" id="toc-step-1-specify-the-likelihood-i.e.-the-distribution-of-the-data" class="nav-link" data-scroll-target="#step-1-specify-the-likelihood-i.e.-the-distribution-of-the-data"><span class="header-section-number">9.3.1</span> Step 1: specify the likelihood (i.e., the distribution of the data)</a></li>
  <li><a href="#step-2-evaluate-the-likelihood-of-individual-observations" id="toc-step-2-evaluate-the-likelihood-of-individual-observations" class="nav-link" data-scroll-target="#step-2-evaluate-the-likelihood-of-individual-observations"><span class="header-section-number">9.3.2</span> Step 2: evaluate the likelihood of individual observations</a></li>
  <li><a href="#step-3-evaluate-the-likelihood-of-all-observations" id="toc-step-3-evaluate-the-likelihood-of-all-observations" class="nav-link" data-scroll-target="#step-3-evaluate-the-likelihood-of-all-observations"><span class="header-section-number">9.3.3</span> Step 3: evaluate the likelihood of <em>all</em> observations</a></li>
  <li><a href="#step-4-put-the-log-likelihoods-together" id="toc-step-4-put-the-log-likelihoods-together" class="nav-link" data-scroll-target="#step-4-put-the-log-likelihoods-together"><span class="header-section-number">9.3.4</span> Step 4: put the (log-)likelihoods together!</a></li>
  <li><a href="#step-5-identify-the-value-of-the-parameters-most-likely-to-explain-the-data" id="toc-step-5-identify-the-value-of-the-parameters-most-likely-to-explain-the-data" class="nav-link" data-scroll-target="#step-5-identify-the-value-of-the-parameters-most-likely-to-explain-the-data"><span class="header-section-number">9.3.5</span> Step 5: identify the value of the parameter(s) most likely to explain the data</a></li>
  <li><a href="#homework-question" id="toc-homework-question" class="nav-link" data-scroll-target="#homework-question"><span class="header-section-number">9.3.6</span> Homework question…</a></li>
  </ul></li>
  <li><a href="#review-of-hypothesis-testing" id="toc-review-of-hypothesis-testing" class="nav-link" data-scroll-target="#review-of-hypothesis-testing"><span class="header-section-number">9.4</span> Review of hypothesis testing</a>
  <ul class="collapse">
  <li><a href="#null-and-alternative-hypotheses" id="toc-null-and-alternative-hypotheses" class="nav-link" data-scroll-target="#null-and-alternative-hypotheses"><span class="header-section-number">9.4.1</span> Null and alternative hypotheses</a></li>
  <li><a href="#test-statistics" id="toc-test-statistics" class="nav-link" data-scroll-target="#test-statistics"><span class="header-section-number">9.4.2</span> Test statistics</a></li>
  <li><a href="#how-do-we-do-this" id="toc-how-do-we-do-this" class="nav-link" data-scroll-target="#how-do-we-do-this"><span class="header-section-number">9.4.3</span> How do we do this?</a></li>
  </ul></li>
  <li><a href="#hypothesis-testing-using-the-log-likelihood-ratio" id="toc-hypothesis-testing-using-the-log-likelihood-ratio" class="nav-link" data-scroll-target="#hypothesis-testing-using-the-log-likelihood-ratio"><span class="header-section-number">9.5</span> Hypothesis testing using the (log-)likelihood ratio</a>
  <ul class="collapse">
  <li><a href="#challenge" id="toc-challenge" class="nav-link" data-scroll-target="#challenge"><span class="header-section-number">9.5.1</span> Challenge</a></li>
  </ul></li>
  <li><a href="#confidence-intervals" id="toc-confidence-intervals" class="nav-link" data-scroll-target="#confidence-intervals"><span class="header-section-number">9.6</span> Confidence intervals</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lectures/lec01-r_and_rstudio.html">Lectures</a></li><li class="breadcrumb-item"><a href="../lectures/lec09-intro-inference-2.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to inference II</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to inference II</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="lesson-preamble" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="lesson-preamble"><span class="header-section-number">9.1</span> Lesson preamble</h2>
<blockquote class="blockquote">
<h3 id="lesson-objectives" data-number="9.1.1" class="anchored"><span class="header-section-number">9.1.1</span> Lesson objectives</h3>
<ul>
<li>Learn how to estimate parameters by numerically maximizing the (log-)likelihood function.</li>
<li>Implement numerical maximization using dataset of Farrell and Davies (2019).</li>
<li>Review of confidence intervals and hypothesis testing, focusing on their relationship with each other and the likelihood function.</li>
<li>Learn how to preform <em>interval</em> estimation using the log-likelihood.</li>
<li>Practice constructing confidence intervals using dataset of Farrell and Davies.</li>
</ul>
<h3 id="lesson-outline" data-number="9.1.2" class="anchored"><span class="header-section-number">9.1.2</span> Lesson outline</h3>
<ul>
<li>Point estimation
<ul>
<li>Review of the idea of maximum likelihood estimation</li>
<li>Example of likelihood maximizion using Farrell and Davies dataset</li>
</ul></li>
<li>Hypothesis testing and interval estimation
<ul>
<li>Review of hypothesis testing</li>
<li>Hypothesis testing &amp; confidence interval construction using the log-likelihood ratio test statistic</li>
</ul></li>
</ul>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="review-of-the-likelihood-function" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="review-of-the-likelihood-function"><span class="header-section-number">9.2</span> Review of the likelihood function</h2>
<p>Given data <span class="math inline">\(x_1,x_2,\dots,x_n \sim f(x|\theta)\)</span>, likelihood-based statistical methods allow us to estimate parameters of of the model, <span class="math inline">\(\theta\)</span>, which we use to explain the data. (If the data are, say, present-day genome sequences, parameters of the model could include historical mutation and recombination rates and effective population sizes. If the data were occurrence-absence data for a species in a given location, parameters of the model could include immigration and emigration rates.)</p>
<p>To determine what parameters were mostly likely to have given rise to the data <em>under the assumption <span class="math inline">\(f\)</span> models the data generating process</em>, we find what value of <span class="math inline">\(\theta\)</span> maximizes</p>
<p><span class="math display">\[ \ln L(x_1,\dots,x_n|\theta) = \sum_{i=1}^n \ln f(x_i|\theta).\]</span></p>
<p>Remember that the logarithm shows up because <span class="math inline">\(\ln L\)</span> has nice statistical properties, and there may be numerical issues if probabilities (or probability densities) are multiplied.</p>
<p>The intuition for the method is quite simple: the value of a parameter (or, in the multi-parameter setting, the combination of values for the parameters) which gives rise to the highest probability of observing our data was most likely to have given rise to that data.</p>
</section>
<section id="numerical-maximization-of-the-log-likelihood" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="numerical-maximization-of-the-log-likelihood"><span class="header-section-number">9.3</span> Numerical maximization of the (log-)likelihood</h2>
<p>Numerical methods are often used to evaluate and maximize the likelihood. When you use functions in many packages, they use numerical techniques to maximize the likelihood function or something like it. So that you know WHAT they are actually doing, the following code chunk illustrates how to evaluate the likelihood function and identify where it assumes a maximum value — using the Farrell and Davies data-set which you analyzed in the self-directed lecture a couple weeks ago.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">read_csv</span>(<span class="st">"data/disease_distance.csv"</span>) <span class="ot">-&gt;</span> disease_distance</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 4157 Columns: 24
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (7): Country, Disease, Parasite, Host, HostOrder, ParaType, ParaFamily
dbl (17): Year, Cases, Deaths, Destroyed, Slaughtered, SR, EvoIso, Taxonomic...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p><em>Note: we will not do anything fancy to maximize the likelihood. People have thought up clever ways to maximize functions, but we will do a <strong>brute-force search of parameter space</strong> to find the parameter(s) most likely to explain the data. This always works, but if there are a lot of combinations of parameters which need to be tested, this is intractable and more sophisticated methods are needed.</em></p>
<section id="step-1-specify-the-likelihood-i.e.-the-distribution-of-the-data" class="level3" data-number="9.3.1">
<h3 data-number="9.3.1" class="anchored" data-anchor-id="step-1-specify-the-likelihood-i.e.-the-distribution-of-the-data"><span class="header-section-number">9.3.1</span> Step 1: specify the likelihood (i.e., the distribution of the data)</h3>
<p>The likelihood one writes down will depend, in general, on the available data and the question that a researcher wants to answer. For this example, we will ignore very important complexities in the data (e.g., the fact that case and death count data are uneven between countries and years, the fact some hosts and parasites are represented more than others) to make progress.</p>
<p>Our goal will be to estimate, for the host species <em>Sus scrofa</em> (the wild boar) and parasite family <em>Coronavirinae</em> (which includes SARS-CoV-1 and -2), the probability that a case results in a death. We will group the data, regardless of the year, sampling location, etc.</p>
<p>What distribution makes sense to model the number of <em>Sus scrofa</em> deaths due to viruses in the family <em>Coronavirinae</em>? The Binomial! One can think of the number of (confirmed) cases of infection with a member of <em>Coronavirinae</em> as the “number of trials” and probability death as a “success probability”. Written out, <span class="math inline">\(\text{deaths}_{ij} \sim \text{Binomial}(\text{cases}_{ij}, p)\)</span> for each country-year <span class="math inline">\(ij\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> disease_distance <span class="sc">|&gt;</span> <span class="fu">subset</span>(ParaFamily <span class="sc">==</span> <span class="st">"Coronavirinae"</span> <span class="sc">&amp;</span> Host <span class="sc">==</span> <span class="st">"Sus_scrofa"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How many observations are there?</p>
</section>
<section id="step-2-evaluate-the-likelihood-of-individual-observations" class="level3" data-number="9.3.2">
<h3 data-number="9.3.2" class="anchored" data-anchor-id="step-2-evaluate-the-likelihood-of-individual-observations"><span class="header-section-number">9.3.2</span> Step 2: evaluate the likelihood of individual observations</h3>
<p>Since we will need to evaluate the probability of <em>all</em> observations in order to estimate <span class="math inline">\(p\)</span>, the first thing to figure out to evaluate likelihood when we have only one observation (i.e., the first row of the previous data frame). The below function evaluate the probability there are exactly <span class="math inline">\(d\)</span> deaths when there are <span class="math inline">\(N\)</span> cases and a probability <span class="math inline">\(p\)</span> that a boar dies due to infection (somewhere, at some time) with a member of the <em>Coronavirinae</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>binomial_evaluator <span class="ot">&lt;-</span> <span class="cf">function</span>(d, N, p){</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">dbinom</span>(<span class="at">x =</span> d, <span class="at">size =</span> N, <span class="at">prob =</span> p)) <span class="co"># log=T returns the log-likelihood of the observation</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># let's figure out the likelihood of a single observation -- the first row</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>cases <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(data[<span class="dv">1</span>,<span class="st">"Cases"</span>])</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>deaths <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(data[<span class="dv">1</span>,<span class="st">"Deaths"</span>])</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># make sure that your observations are numeric vectors, not data frames!</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="fu">binomial_evaluator</span>(<span class="at">d =</span> deaths, <span class="at">N =</span> cases, <span class="at">p =</span> <span class="fl">0.01</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9227447</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">binomial_evaluator</span>(<span class="at">d =</span> deaths, <span class="at">N =</span> cases, <span class="at">p =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.4304672</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">binomial_evaluator</span>(<span class="at">d =</span> deaths, <span class="at">N =</span> cases, <span class="at">p =</span> <span class="fl">0.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1677722</code></pre>
</div>
</div>
<p>Of these values of <span class="math inline">\(p\)</span>, which is most likely to explain the observation in the first row?</p>
</section>
<section id="step-3-evaluate-the-likelihood-of-all-observations" class="level3" data-number="9.3.3">
<h3 data-number="9.3.3" class="anchored" data-anchor-id="step-3-evaluate-the-likelihood-of-all-observations"><span class="header-section-number">9.3.3</span> Step 3: evaluate the likelihood of <em>all</em> observations</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>dataALL <span class="ot">&lt;-</span> data[,<span class="fu">c</span>(<span class="st">"Cases"</span>, <span class="st">"Deaths"</span>)]</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">binomial_evaluator</span>(<span class="at">d =</span> dataALL<span class="sc">$</span>Deaths, <span class="at">N =</span> dataALL<span class="sc">$</span>Cases, <span class="at">p =</span> <span class="fl">0.01</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 9.227447e-01 0.000000e+00 9.043821e-01 0.000000e+00 9.043821e-01
 [6] 1.000000e-12 2.084925e-01 9.509900e-01 9.900000e-01 6.670506e-21
[11] 0.000000e+00 9.509900e-01 8.179069e-01 9.065279e-55</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">binomial_evaluator</span>(<span class="at">d =</span> dataALL<span class="sc">$</span>Deaths, <span class="at">N =</span> dataALL<span class="sc">$</span>Cases,<span class="at">p =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 4.304672e-01 1.693764e-02 3.486784e-01 0.000000e+00 3.486784e-01
 [6] 1.000000e-06 7.274974e-08 5.904900e-01 9.000000e-01 1.596866e-07
[11] 0.000000e+00 5.904900e-01 1.215767e-01 1.551125e-13</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">binomial_evaluator</span>(<span class="at">d =</span> dataALL<span class="sc">$</span>Deaths, <span class="at">N =</span> dataALL<span class="sc">$</span>Cases,<span class="at">p =</span> <span class="fl">0.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1.677722e-01  1.516705e-96  1.073742e-01 3.860350e-306  1.073742e-01
 [6]  6.400000e-05  7.621456e-16  3.276800e-01  8.000000e-01  4.470958e-04
[11] 8.870964e-220  3.276800e-01  1.152922e-02  1.208139e-04</code></pre>
</div>
</div>
</section>
<section id="step-4-put-the-log-likelihoods-together" class="level3" data-number="9.3.4">
<h3 data-number="9.3.4" class="anchored" data-anchor-id="step-4-put-the-log-likelihoods-together"><span class="header-section-number">9.3.4</span> Step 4: put the (log-)likelihoods together!</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># numerical problems can arise when you take products of small numbers</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># this is one reason why we need to take the log of the likelihood</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># product of individual likelihoods (for each observation) = the sum of log-likelihoods</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">prod</span>(<span class="fu">binomial_evaluator</span>(<span class="at">d =</span> dataALL<span class="sc">$</span>Deaths, <span class="at">N =</span> dataALL<span class="sc">$</span>Cases, <span class="at">p =</span> <span class="fl">0.2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">log</span>(<span class="fu">binomial_evaluator</span>(<span class="at">d =</span> dataALL<span class="sc">$</span>Deaths, <span class="at">N =</span> dataALL<span class="sc">$</span>Cases, <span class="at">p =</span> <span class="fl">0.2</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -1502.624</code></pre>
</div>
</div>
</section>
<section id="step-5-identify-the-value-of-the-parameters-most-likely-to-explain-the-data" class="level3" data-number="9.3.5">
<h3 data-number="9.3.5" class="anchored" data-anchor-id="step-5-identify-the-value-of-the-parameters-most-likely-to-explain-the-data"><span class="header-section-number">9.3.5</span> Step 5: identify the value of the parameter(s) most likely to explain the data</h3>
<p>So far, we have only evaluated the log-likelihood (of one and all of observations) at specific values of the parameter of interest, <span class="math inline">\(p\)</span>. To identify what value of <span class="math inline">\(p\)</span> is most likely to explain the data, we need to evaluate the log-likelihood across a RANGE of <span class="math inline">\(p\)</span> values and identify when the function is maximized.</p>
<p>We can do this in a couple ways, but here is how using a for loop:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>LogLik <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>index <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>p_values_to_test <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.001</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (p <span class="cf">in</span> p_values_to_test){</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  LogLik[index] <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">log</span>(<span class="fu">binomial_evaluator</span>(<span class="at">d =</span> dataALL<span class="sc">$</span>Deaths, <span class="at">N =</span> dataALL<span class="sc">$</span>Cases, <span class="at">p =</span> p)))</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  index <span class="ot">&lt;-</span> index <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>LLs <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(LogLik, <span class="at">p =</span> p_values_to_test)</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>LLs <span class="sc">|&gt;</span> <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> LogLik)) <span class="sc">+</span> <span class="fu">geom_line</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lec09-intro-inference-2_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Values of <span class="math inline">\(p\)</span> where <span class="math inline">\(\ln L(p) - -\infty\)</span> are <em>very</em> unlikely to explain the data. Since we are interested in the value of <span class="math inline">\(p\)</span> which gives rise to the largest <span class="math inline">\(\ln L\)</span>, and these values are infinitely <em>negative</em>, they are not candidates for our estimator of <span class="math inline">\(p\)</span>. As a result, they are suppressed in the plot.</p>
<p>Another thing to notice is the curvature of the likelihood function around the value where it assumes a maximum. This tells us something about how confident we should be about our “best guess” for the parameter, or the maximum likelihood estimate of that parameter. More on this in a couple minutes…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>LLs <span class="sc">|&gt;</span> <span class="fu">subset</span>(LogLik <span class="sc">==</span> <span class="fu">max</span>(LogLik))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       LogLik     p
235 -1476.786 0.234</code></pre>
</div>
</div>
</section>
<section id="homework-question" class="level3" data-number="9.3.6">
<h3 data-number="9.3.6" class="anchored" data-anchor-id="homework-question"><span class="header-section-number">9.3.6</span> Homework question…</h3>
<p>In the next homework, you will be asked to write a function with arguments that specify the host species and the parasite family. Then, using this function on all host and parasite family combinations, you will be asked to estimate the most likely values of <span class="math inline">\(p\)</span>, i.e., the value which maximizes the likelihood function that arises from the assumption that, for a combination of host and parasite family, the number of deaths of the host is Binomial with a probability of death that does not depend on time or sampling location. If there are no observations for a specific host and parasite family combination, the function should return <code>NA</code>; else, the function should return the value of <span class="math inline">\(p\)</span> which maximizes the likelihood function formed from the data specific to <em>that</em> host and parasite family. You may also run into issues where the log-likelihood is negative infinity at all <span class="math inline">\(p\)</span>; in this case, the function should return <code>NA</code>. This is a really challenging exercise, but we think it will be (hopefully) equally as rewarding!</p>
<p><em>Hint: this exercise will involve modifying the code that you have been provided in this lecture. In particular, you will have to wrap the code above in a function and loop over all combinations of hosts and parasite families.</em></p>
</section>
</section>
<section id="review-of-hypothesis-testing" class="level2" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="review-of-hypothesis-testing"><span class="header-section-number">9.4</span> Review of hypothesis testing</h2>
<p>Often, the objective of a study is not to estimate a parameter but to decide which of two (or more) contradictory claims about the parameter is consistent with the data. This part of statistics is called <em>hypothesis testing</em>. Hypothesis testing is intimately connected to the construction and interpretation of confidence intervals which quantify estimation uncertainty.</p>
<p><img src="figures/hypothesistesting.png" class="img-fluid"></p>
<section id="null-and-alternative-hypotheses" class="level3" data-number="9.4.1">
<h3 data-number="9.4.1" class="anchored" data-anchor-id="null-and-alternative-hypotheses"><span class="header-section-number">9.4.1</span> Null and alternative hypotheses</h3>
<p>A statistical hypothesis is an assertion about the values of one or more parameters, or the form of a probability distribution that is used to model the data.</p>
<p>Two contradictory hypotheses of the first kind are</p>
<p><span class="math display">\[H_0: \theta = \theta_0\]</span> <span class="math display">\[H_1: \theta \neq \theta_0\]</span></p>
<p>The first hypothesis is called the <em>null hypothesis</em> and may correspond to an expectation we have about the parameter (from, e.g., prior data). The second hypothesis is called the <em>alternative hypothesis</em>. The data are used to make a principle conclusion about if the null hypothesis is consistent with the data; if so, we reject the alternative hypothesis and, if not, we reject the null hypothesis as an explanation for the data generative process. In the following sections, we will describe the process for conducting such a hypothesis test.</p>
</section>
<section id="test-statistics" class="level3" data-number="9.4.2">
<h3 data-number="9.4.2" class="anchored" data-anchor-id="test-statistics"><span class="header-section-number">9.4.2</span> Test statistics</h3>
<p>Suppose we have data <span class="math inline">\(x_1,\dots,x_n \sim f(x|\theta)\)</span> and wish to test the above hypotheses, i.e., to decide if <span class="math inline">\(\theta \neq \theta_0\)</span>. We do this by constructing a test statistic, i.e., function of the data, and assessing if the realized value of statistic is consistent with its distribution under the null hypothesis. Is the value of the statistic, at some level of significance, different from what we would expect if the null hypothesis were true?</p>
<p>Many choices of test statistic are possible, but the likelihood ratio is one that is commonly used:</p>
<p><span class="math display">\[\lambda_{LR} = -2 (\ln L(\theta_0)-\ln L(\hat{\theta}_{\text{MLE}})).\]</span></p>
<p>The statistic is based on the likelihood function, and its asymptotic distribution (as the sample size becomes large) under the null hypothesis is known. <span class="math inline">\(\lambda_{LR}\)</span> has an approximate <span class="math inline">\(\chi^2\)</span> distribution under <span class="math inline">\(H_0\)</span>. (<em>Recall that one must know the approximate distribution of a test statistic to preform a hypothesis test.</em>) Given the value and distribution of our test statistic under the null hypothesis, we can determine which two competing hypothesis is consistent with the data.</p>
</section>
<section id="how-do-we-do-this" class="level3" data-number="9.4.3">
<h3 data-number="9.4.3" class="anchored" data-anchor-id="how-do-we-do-this"><span class="header-section-number">9.4.3</span> How do we do this?</h3>
<p>To decide between the null and alternative hypothesis, given a test statistic and its distribution under <span class="math inline">\(H_0\)</span>, we must specify a <em>significance level</em> <span class="math inline">\(\alpha\)</span>. The significance level measures how likely we are to reject the null hypothesis, given that it is true: <span class="math inline">\(\alpha = P(\text{reject } H_0 | H_0)\)</span>. The significance level is chosen <em>before data collection</em>, and is typically set to 0.05 or smaller. By a similar token, the <em>power</em> of a statistical test is defined as the probability of rejecting the alternative hypothesis, given it is true: <span class="math inline">\(1-\beta = \Pr(\text{reject } H_0 | H_1).\)</span> Many factors affect the power of a test, but a test based on the likelihood ratio test statistic is the uniformly most powerful among all alternatives to test the above hypothesis. In general, one can preform (and it is best practice to preform!) analyses ahead of data collection to ensure power at a certain level.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
<p><strong>We conduct a hypothesis test at significance level <span class="math inline">\(\alpha\)</span> as follows:</strong></p>
<ol start="0" type="1">
<li>State the null and alternative hypothesis and significance level <span class="math inline">\(\alpha\)</span>.</li>
<li>Collect data, possibly with knowledge of the sample size required to achieve a certain power.</li>
<li>Calculate the realized value <span class="math inline">\(s\)</span> of a test statistic <span class="math inline">\(S\)</span>, e.g., <span class="math inline">\(\lambda_{LR}\)</span>. The test statistic must have a known distribution under the null hypothesis. The likelihood ratio has a <span class="math inline">\(\chi^2\)</span> distribution under <span class="math inline">\(H_0\)</span> above.</li>
<li>Compute the probability of observing the realized value of the test statistic or something more extreme, given the null hypothesis is true, i.e., <span class="math inline">\(p = \Pr(S &gt; s | H_0)\)</span>. This probability is called a <span class="math inline">\(p\)</span> value. If <span class="math inline">\(p &lt; \alpha\)</span>, we reject the null hypothesis at significance level <span class="math inline">\(\alpha\)</span> and, if not, we fail to reject <span class="math inline">\(H_0\)</span>.</li>
</ol>
<p>To illustrate how this works, we will return to the previous example.</p>
</section>
</section>
<section id="hypothesis-testing-using-the-log-likelihood-ratio" class="level2" data-number="9.5">
<h2 data-number="9.5" class="anchored" data-anchor-id="hypothesis-testing-using-the-log-likelihood-ratio"><span class="header-section-number">9.5</span> Hypothesis testing using the (log-)likelihood ratio</h2>
<p>Using data of the number of cases and deaths of the wild boar infected with members of the viral family <em>Coronavirinae</em>, we estimated that the probability of death given infection is <span class="math inline">\(\hat{p}_{\text{MLE}} = 0.234\)</span>. Is this estimate significantly different than say 0.2?</p>
<p>A test of <span class="math inline">\(H_0: p = 0.2\)</span> vs <span class="math inline">\(H_1: p \neq 0.2\)</span> at significance level <span class="math inline">\(\alpha = 0.05\)</span> can be preformed by</p>
<ol type="1">
<li>calculating the log-likelihood ratio test statistic</li>
<li>determining the probability of obtaining a statistic more extreme, under the expected distribution for the statistic. For the log-likelihood statistic, as it has a <span class="math inline">\(\chi^2\)</span> distribution with one degree of freedom, this can be done as follows:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>LL_at_null <span class="ot">&lt;-</span> LLs <span class="sc">|&gt;</span> <span class="fu">subset</span>(p <span class="sc">==</span> <span class="fl">0.2</span>) <span class="sc">|&gt;</span> <span class="fu">select</span>(LogLik)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>LL_at_max <span class="ot">&lt;-</span> LLs <span class="sc">|&gt;</span> <span class="fu">subset</span>(LogLik <span class="sc">==</span> <span class="fu">max</span>(LogLik)) <span class="sc">|&gt;</span> <span class="fu">select</span>(LogLik)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>test_statistic <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="sc">-</span><span class="dv">2</span><span class="sc">*</span>(LL_at_null <span class="sc">-</span> LL_at_max))</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>test_statistic <span class="do">### realized value of the likelihood ratio test statistic</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 51.67639</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pchisq</span>(test_statistic, <span class="at">df =</span> <span class="dv">1</span>, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6.544503e-13</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="do">### if p &lt; 0.05, then we reject H0; otherwise, we fail to reject H0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="challenge" class="level3" data-number="9.5.1">
<h3 data-number="9.5.1" class="anchored" data-anchor-id="challenge"><span class="header-section-number">9.5.1</span> Challenge</h3>
<p>How might we extend this operation – of calculating the <span class="math inline">\(p\)</span>-value, or the probability of a the log-likelihood ratio test statistic more extreme than the one we observed – to determine the set of <span class="math inline">\(p_0\)</span> such that we fail to reject <span class="math inline">\(H_0: p = p_0\)</span>?</p>
</section>
</section>
<section id="confidence-intervals" class="level2" data-number="9.6">
<h2 data-number="9.6" class="anchored" data-anchor-id="confidence-intervals"><span class="header-section-number">9.6</span> Confidence intervals</h2>
<p>Having gotten an estimate for <span class="math inline">\(p\)</span>, the probability a wild boar dies when infected with a member of the <em>Coronavirinae</em>, you might be asking if it is possible to get an estimate of the uncertainty around our “best guess” (i.e., maximum likelihood estimator) for <span class="math inline">\(p\)</span>. The answer is a resounding YES! It turns out the previous calculations enable us to construct a confidence interval for <span class="math inline">\(p\)</span>. This is because <strong>the confidence interval for a parameter is nothing but the set of values of the parameter for which we cannot reject the null hypothesis!</strong><a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> That is, there is a duality between confidence intervals and hypothesis testing. When you do one, you (secretly) do the other. In the previous example, we showed that <span class="math inline">\(p=0.2\)</span> is not in the confidence interval for <span class="math inline">\(p\)</span>!</p>
<p>There are other ways to construct confidence intervals, but the one below involves finding which <span class="math inline">\(\theta_0\)</span> are such that <span class="math inline">\(\lambda_{LR} -2(\ln L(\theta_0)-\ln L(\hat{\theta}_{\text{MLE}})) &lt; \chi^2_c\)</span>, where <span class="math inline">\(\chi^2_c\)</span> a cutoff based on <span class="math inline">\(100(1-\alpha)\%\)</span>-ile for a <span class="math inline">\(\chi^2\)</span> distribution with one degree of freedom.<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>cutoff <span class="ot">&lt;-</span> <span class="fu">qchisq</span>(<span class="fl">0.95</span>,<span class="at">df=</span><span class="dv">1</span>)<span class="sc">/</span><span class="dv">2</span> </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="co"># cutoff for admissible values based on 95%-ile for a chi-squared dist df=1</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>MLE <span class="ot">&lt;-</span> LLs <span class="sc">|&gt;</span> <span class="fu">subset</span>(LogLik <span class="sc">==</span> <span class="fu">max</span>(LogLik)) </span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>LLs <span class="sc">%&gt;%</span> <span class="fu">subset</span>(<span class="fu">abs</span>(LogLik <span class="sc">-</span> MLE<span class="sc">$</span>LogLik) <span class="sc">&lt;</span> cutoff) <span class="ot">-&gt;</span> values_inconfidence_interval</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="fu">min</span>(values_inconfidence_interval<span class="sc">$</span>p),</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">max</span>(values_inconfidence_interval<span class="sc">$</span>p)) <span class="do">### confidence interval!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.225 0.243</code></pre>
</div>
</div>
<p>Now, let’s display our point estimate for <span class="math inline">\(p\)</span> as well as the confidence interval we have just constructed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>LLs <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> LogLik)) <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> MLE<span class="sc">$</span>p, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">min</span>(values_inconfidence_interval<span class="sc">$</span>p), </span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>   <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">max</span>(values_inconfidence_interval<span class="sc">$</span>p), </span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"p"</span>, <span class="at">y =</span> <span class="st">"log-likelihood"</span>) <span class="sc">+</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="fl">0.2</span>,<span class="fl">0.3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lec09-intro-inference-2_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The limits for the <span class="math inline">\(95\%\)</span> confidence interval for <span class="math inline">\(p\)</span> are given by the dashed red lines. The maximum likelihood estimate for <span class="math inline">\(p\)</span> is represented by the solid red line. The key thing to note is that the width confidence region is determined by the curvature of the log-likelihood function about the maximum likelihood estimate for the focal parameter.</p>


</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p>We will learn how to do power analyses in the linear models lectures!<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>There are other ways to define and understand confidence intervals but they are somewhat oblique.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p><em>Note:</em> the code in the previous section assumed <span class="math inline">\(\alpha = 0.05\)</span>. The functions we use in this section, to keep things general, are a bit different.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../lectures/lec08-intro-inference-1.html" class="pagination-link" aria-label="Introduction to inference I">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to inference I</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../assignments/assignment-00.html" class="pagination-link" aria-label="Assignment00: individual interest description">
        <span class="nav-page-text">Assignment00: individual interest description</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>