<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>EEB313: Quantitative Methods in R for Biology - 13&nbsp; Multivariate statistics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./lec12-command-line_git.html" rel="next">
<link href="./lec10-model-selection.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./lec00-rstudio.html">Lectures</a></li><li class="breadcrumb-item"><a href="./lec11-multivariate-stats.html"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Multivariate statistics</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">EEB313: Quantitative Methods in R for Biology</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/EEB313/eeb313.github.io" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Syllabus</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./about-us.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The 2023 teaching team</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./downloadingR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Downloading R</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Lectures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec00-rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Getting started with RStudio and R Notebook</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec01-markdown-workflows.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Markdown, project workflows</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec02-base-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Base R: assignment, vectors, functions, and loops</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec03-dataframes-dplyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data frames, intro to dplyr</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec04-data-wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data wrangling in dplyr</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec05-data-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Data visualization in ggplot</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec06-exploratory-data-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Exploratory data analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec07-intro-to-stats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to inference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec08-linear-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Linear models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec09-mixed-effects-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Linear mixed models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec10-model-selection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Model selection</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec11-multivariate-stats.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Multivariate statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec12-command-line_git.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Command line &amp; Git(Hub)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec14-mathematical-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Mathematical modeling I</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec15-mathematical-models-2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Mathematical modeling II</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Assignments</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment-01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Assignment 01</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment-02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Assignment 02</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment-03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Assignment 03</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment-04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Assignment 04</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment-05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Assignment 05</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./challenge-assignment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Challenge assignment</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./projects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Project description</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./databases.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Some open-access databases</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Miscellaneous notes</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tests-assignment-3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">More on hypothesis testing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./power-analysis-practice.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Power analysis practice</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#multivariate-statistics" id="toc-multivariate-statistics" class="nav-link active" data-scroll-target="#multivariate-statistics"><span class="header-section-number">14</span> Multivariate Statistics</a>
  <ul class="collapse">
  <li><a href="#lesson-intro" id="toc-lesson-intro" class="nav-link" data-scroll-target="#lesson-intro"><span class="header-section-number">14.1</span> Lesson intro</a>
  <ul class="collapse">
  <li><a href="#a-bit-of-theory1" id="toc-a-bit-of-theory1" class="nav-link" data-scroll-target="#a-bit-of-theory1"><span class="header-section-number">14.1.1</span> A bit of theory</a></li>
  </ul></li>
  <li><a href="#lets-practice-toy-example" id="toc-lets-practice-toy-example" class="nav-link" data-scroll-target="#lets-practice-toy-example"><span class="header-section-number">14.2</span> Let’s practice: toy example</a>
  <ul class="collapse">
  <li><a href="#a-note-on-distance-matrices" id="toc-a-note-on-distance-matrices" class="nav-link" data-scroll-target="#a-note-on-distance-matrices"><span class="header-section-number">14.2.1</span> A note on distance matrices</a></li>
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca"><span class="header-section-number">14.2.2</span> PCA</a></li>
  <li><a href="#pcoa" id="toc-pcoa" class="nav-link" data-scroll-target="#pcoa"><span class="header-section-number">14.2.3</span> PCoA</a></li>
  </ul></li>
  <li><a href="#lets-practice-real-life-example" id="toc-lets-practice-real-life-example" class="nav-link" data-scroll-target="#lets-practice-real-life-example"><span class="header-section-number">14.3</span> Let’s practice: real life example</a>
  <ul class="collapse">
  <li><a href="#multicollinearity" id="toc-multicollinearity" class="nav-link" data-scroll-target="#multicollinearity"><span class="header-section-number">14.3.1</span> Multicollinearity</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Multivariate statistics</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<blockquote class="blockquote">
<h3 id="learning-objectives" data-number="13.0.1" class="anchored"><span class="header-section-number">13.0.1</span> Learning objectives</h3>
<ul>
<li>Learn conceptual theory behind common multivariate techniques</li>
<li>Learn how to run and interpret the output from PCA and PCoA</li>
<li>Learn how to assess multicollinearity</li>
<li>Be able to apply the above techniques on real-world data</li>
<li>Your types of data and biological questions drive your types of statistical analysis</li>
<li><strong>Recognize that the choices required in a statistical pipeline can drastically affect your results, but there is no one right way of analyzing your data</strong></li>
</ul>
</blockquote>
<blockquote class="blockquote">
<h3 id="lesson-outline" data-number="13.0.2" class="anchored"><span class="header-section-number">13.0.2</span> Lesson outline</h3>
<ul>
<li>Examples of multivariate statistics using well-behaved data: PCA and PCoA</li>
<li>Discuss caveats and analytically choices embedded in PCA and PCoA</li>
<li>Work through multivariate analysis of soil invasion data!</li>
</ul>
</blockquote>
<blockquote class="blockquote">
<h3 id="required-packages" data-number="13.0.3" class="anchored"><span class="header-section-number">13.0.3</span> Required packages</h3>
<ul>
<li><code>tidyverse</code></li>
<li><code>car</code></li>
<li><code>ggfortify</code></li>
<li><code>wesanderson</code></li>
<li><code>vegan</code></li>
</ul>
</blockquote>
<section id="multivariate-statistics" class="level1" data-number="14">
<h1 data-number="14"><span class="header-section-number">14</span> Multivariate Statistics</h1>
<section id="lesson-intro" class="level2" data-number="14.1">
<h2 data-number="14.1" class="anchored" data-anchor-id="lesson-intro"><span class="header-section-number">14.1</span> Lesson intro</h2>
<p>Most things in the natural world are inter-related and correlated, where many trends and observations tend to change in a predictable way together. This is personally one of the things that I find exciting about biology–the sheer complexity of interactions in natural systems.</p>
<p>But this also makes our lives as scientists challenging, because we then have to distill, what is sometimes called a “causal soup”, into variables of interest and then figure out a meaningful analysis that can tease apart which effects came from where.</p>
<p>And this is where it gets complicated—how do we accurately represent the true complex and multidimensional nature of ecological systems without breaking our brains?</p>
<p>Cue multivariate statistics.</p>
<p>We use multivariate techniques to reduce complex situations into more manageable ones and to deal with the fact that patterns are correlated with other patterns, which are in turn driven by a number of interacting ecological processes.</p>
<p>Up until now we’ve been working with univariate methods – when there is a single response variable is of interest, or when we have multiple predictors but we are mostly interested in their independent effects on the response.</p>
<p>In contrast, multivariate statistics is a family of statistical tools used to to handle situations where the predictor and/or the response variables come in sets, allowing us to distill through some of that correlated (and sometimes redundant) information and to analyze these variables simultaneously, as they should be.</p>
<p>Unfortunately (?), it is not within the scope of this class (or even this course!) to teach all of this. Multivariate statistics area large discipline in their own right. They are a flexible tool with a lot of variations - which can make it challenging to choose the correct technique for an analysis.</p>
<p>Today, we will focus on a couple of the most common techniques that you are likely to come across in the papers you read and/or use in your own work. This is a quick overview and we necessarily have to have to skip over some of the details—if want to know more or have a specific question related to data, we are happy provide readings or to chat during office hours.</p>
<section id="a-bit-of-theory1" class="level3" data-number="14.1.1">
<h3 data-number="14.1.1" class="anchored" data-anchor-id="a-bit-of-theory1"><span class="header-section-number">14.1.1</span> A bit of theory<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></h3>
<p>Multivariate statistics roughly split into unconstrained vs constrained ordination techniques.</p>
<p>Unconstrained ordinations have no a priori groupings or relationships in the variables or observations. Common examples include: PCA, PCoA, cluster analysis, correspondence analysis (CA), factor analysis and NMDS. We will focus on this broad category today.</p>
<p>Constrained ordinations specify a priori groupings. Examples include discriminant analysis, canonical variate analysis, canonical correlation analysis, and redundancy analysis.</p>
<p>Unconstrained ordinations generally work by the extraction of the eigenvectors of an association matrix. For example, imagine your data containing n objects and p variables. The n objects can be represented as a cluster of points in the p-dimensional space. This cluster is generally not spheroid: it is elongated in some directions and flattened in others. These directions are not necessarily aligned with a single dimension (= a single variable) of the multidimensional space.</p>
<p>The direction where the cluster is most elongated corresponds to the direction of largest variance of the cluster and this is the first axis that an ordination will extract. The next axis to be extracted is the second most important in variance, provided that it is orthogonal (linearly independent, uncorrelated) to the first one.</p>
<p>The process continues until all axes have been computed. When there are a few major structures in the data (gradients or groups) and the method has been efficient at extracting them, then the few first axes contain most of the useful information, i.e.&nbsp;they have extracted most of the variance of the data. In that case, the distances among sites in the projection in reduced space are relatively similar to the distances among objects in the multidimensional space.</p>
<p>How the methods do this specifically varies and depends on the types of data and specific methods you are using.</p>
<p>For more details see <a href="https://link.springer.com/content/pdf/10.1007/978-1-4419-7976-6.pdf">this</a> chapter in Borcard, D., Gillet, F. and Legendre, P. 2011. <em>Numerical Ecology with R.</em></p>
</section>
</section>
<section id="lets-practice-toy-example" class="level2" data-number="14.2">
<h2 data-number="14.2" class="anchored" data-anchor-id="lets-practice-toy-example"><span class="header-section-number">14.2</span> Let’s practice: toy example</h2>
<p>For this section we will be using soil chemistry data from the <code>vegan</code> library.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(varechem)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(varechem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      N    P     K    Ca    Mg    S    Al   Fe    Mn   Zn  Mo Baresoil Humdepth
18 19.8 42.1 139.9 519.4  90.0 32.3  39.0 40.9  58.1  4.5 0.3     43.9      2.2
15 13.4 39.1 167.3 356.7  70.7 35.2  88.1 39.0  52.4  5.4 0.3     23.6      2.2
24 20.2 67.7 207.1 973.3 209.1 58.1 138.0 35.4  32.1 16.8 0.8     21.2      2.0
27 20.6 60.8 233.7 834.0 127.2 40.7  15.4  4.4 132.0 10.7 0.2     18.7      2.9
23 23.8 54.5 180.6 777.0 125.8 39.5  24.2  3.0  50.1  6.6 0.3     46.0      3.0
19 22.8 40.9 171.4 691.8 151.4 40.8 104.8 17.6  43.6  9.1 0.4     40.5      3.8
    pH
18 2.7
15 2.8
24 3.0
27 2.8
23 2.7
19 2.7</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(varechem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   24 obs. of  14 variables:
 $ N       : num  19.8 13.4 20.2 20.6 23.8 22.8 26.6 24.2 29.8 28.1 ...
 $ P       : num  42.1 39.1 67.7 60.8 54.5 40.9 36.7 31 73.5 40.5 ...
 $ K       : num  140 167 207 234 181 ...
 $ Ca      : num  519 357 973 834 777 ...
 $ Mg      : num  90 70.7 209.1 127.2 125.8 ...
 $ S       : num  32.3 35.2 58.1 40.7 39.5 40.8 33.8 27.1 42.5 60.2 ...
 $ Al      : num  39 88.1 138 15.4 24.2 ...
 $ Fe      : num  40.9 39 35.4 4.4 3 ...
 $ Mn      : num  58.1 52.4 32.1 132 50.1 ...
 $ Zn      : num  4.5 5.4 16.8 10.7 6.6 9.1 7.4 5.2 9.3 9.1 ...
 $ Mo      : num  0.3 0.3 0.8 0.2 0.3 0.4 0.3 0.3 0.3 0.5 ...
 $ Baresoil: num  43.9 23.6 21.2 18.7 46 40.5 23 29.8 17.6 29.9 ...
 $ Humdepth: num  2.2 2.2 2 2.9 3 3.8 2.8 2 3 2.2 ...
 $ pH      : num  2.7 2.8 3 2.8 2.7 2.7 2.8 2.8 2.8 2.8 ...</code></pre>
</div>
</div>
<p>This data frame is composed of 24 rows that correspond to different sites and 14 columns, which correspond to the different chemicals measured at each site.</p>
<section id="a-note-on-distance-matrices" class="level3" data-number="14.2.1">
<h3 data-number="14.2.1" class="anchored" data-anchor-id="a-note-on-distance-matrices"><span class="header-section-number">14.2.1</span> A note on distance matrices</h3>
<p>Many multivariate methods rely on being able to represent your observations as being more or less similar to each other (similarities or distances). This makes, because if you’re trying to group observations by variance along an axis, you need to know how far apart each observation is from other observations.</p>
<p>This is easy to do (and visualize) with quantitative data—a distance is just the numeric difference between the observations along an axis. To deal with negative numbers and multiple dimensions, rather than subtracting one value the other, we instead compute <strong>euclidean distances</strong>.</p>
<p>For visualization, let’s compute the euclidean distances between the first four soil chemicals. We will use the <code>vegdist</code> function in <code>vegan</code><em>.</em> <code>vegdist</code> is handy because it allows us to specify which type of distance matrix we want to use.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>?vegdist</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># let's just look at the first column</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>euclidean.distance <span class="ot">&lt;-</span> <span class="fu">vegdist</span>(varechem[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">method=</span><span class="st">"euclidean"</span>) </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># lets see it </span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>euclidean.distance</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          18        15        24        27        23        19        22
15 165.14239                                                            
24 459.56128 618.58209                                                  
27 328.81899 482.43858 141.98528                                        
23 261.12068 420.92078 198.55261  78.22110                              
19 175.28391 335.26172 285.02937 156.53428  86.77350                    
22 221.62195 382.15759 239.50061 116.61587  43.40369  47.14149          
16 125.38142  49.65350 583.95530 450.65908 385.46067 299.21572 345.65371
28 260.84978 404.51356 231.11534  90.72365  86.64825 110.39909  96.51135
13 175.40339 235.66141 446.46377 304.81017 271.65128 207.69502 243.84187
14  10.92932 157.07342 465.96998 334.10115 267.45923 181.30130 227.78744
20 231.32127 386.96504 232.30166  98.16649  42.52905  61.98032  39.88233
25 129.45443 292.09529 330.43671 203.54086 132.25891  48.99612  93.83017
7  337.13646 191.42586 796.46802 664.90879 598.04244 512.08316 558.00082
5  296.18195 171.77805 752.47104 624.68037 554.89965 469.70461 514.65909
6  272.41973 144.90680 730.14008 601.23737 532.27095 446.78778 492.18765
3  307.97131 167.57619 766.90027 636.19769 568.62795 482.93683 528.74291
4   87.65546 172.76415 456.04597 316.57096 263.35787 183.75062 229.19383
2   87.45988  80.33941 539.16396 403.86099 341.29395 255.84222 302.76968
9   24.71882 148.66563 480.77366 351.15263 282.65819 197.46881 243.57432
12  50.88271 121.96975 509.56721 379.21342 311.23155 225.19911 271.19956
10 205.77395 357.02368 263.83591 125.84288  76.71760  52.82717  59.88447
11 653.73125 814.06559 196.79954 336.95884 393.54222 479.37130 432.96563
21  51.98279 143.46616 500.87685 373.94623 303.28060 218.05641 262.75243
          16        28        13        14        20        25         7
15                                                                      
24                                                                      
27                                                                      
23                                                                      
19                                                                      
22                                                                      
16                                                                      
28 376.81408                                                            
13 228.66139 217.27572                                                  
14 118.15198 264.60605 169.54852                                        
20 353.81686  59.52411 230.46991 236.84586                              
25 254.51902 149.74198 194.73007 136.46575 106.65711                    
7  214.70470 590.73971 423.76340 331.17554 567.57318 466.05208          
5  181.39942 554.78769 404.46266 291.45274 526.91462 422.83496  62.65477
6  156.05073 530.56984 378.72901 267.48510 503.62712 400.06362  73.67035
3  187.20286 562.92360 401.51605 302.37391 538.61420 436.49603  31.85750
4  154.15687 234.16757  95.94514  82.45684 224.87883 150.10506 362.41522
2   52.55578 327.48820 181.37481  79.01114 307.61853 212.39995 263.62223
9  107.37337 283.60679 191.48940  26.21927 253.52177 150.59459 316.16703
12  78.15120 310.71511 200.32099  46.29903 281.94744 179.42770 287.37355
10 326.07708  62.24942 197.40995 210.59713  36.10637  89.77878 540.37336
11 778.43029 425.20842 638.57443 660.35430 428.65991 524.50748 989.92165
21  96.48300 309.98103 217.30067  51.80463 276.74176 171.62628 297.57023
           5         6         3         4         2         9        12
15                                                                      
24                                                                      
27                                                                      
23                                                                      
19                                                                      
22                                                                      
16                                                                      
28                                                                      
13                                                                      
14                                                                      
20                                                                      
25                                                                      
7                                                                       
5                                                                       
6   29.48525                                                            
3   38.15573  44.10873                                                  
4  334.10919 308.22925 336.35840                                        
2  232.42560 206.79284 236.51364 101.98505                              
9  274.31192 250.47305 286.59810 100.16816  74.62433                    
12 246.28329 222.22664 258.28515 113.13218  55.33643  32.30681          
10 501.74134 477.63498 511.88009 192.37204 278.54732 228.16376 256.08832
11 944.08465 922.41863 960.16075 652.44735 734.72664 674.65552 703.12318
21 252.26522 229.82385 267.82042 130.32041  81.19384  35.54786  26.96294
          10        11
15                    
24                    
27                    
23                    
19                    
22                    
16                    
28                    
13                    
14                    
20                    
25                    
7                     
5                     
6                     
3                     
4                     
2                     
9                     
12                    
10                    
11 460.16602          
21 253.41513 692.92554</code></pre>
</div>
</div>
<p>Note that the output here is an object of class “distance” that shows all the pairwise euclidean distances between all cells in the data.</p>
<p>Thinking about distances are more challenging to do if you have some data that is categorical (what is the similarity along an axis of apples to oranges?). Luckily there are different types of distance matrices that can handle different types of data.</p>
<p>Another really common type of distance matrix is the Gower distance, which can handle categorical and mixed data, and so is very useful. For more detail on distance matrices read <a href="https://chrischizinski.github.io/snr_r_notebooks/distancematrices/">this</a>.</p>
<p>Okay, now we’re ready to do some statistics!!</p>
</section>
<section id="pca" class="level3" data-number="14.2.2">
<h3 data-number="14.2.2" class="anchored" data-anchor-id="pca"><span class="header-section-number">14.2.2</span> PCA</h3>
<p>One of the most common (unconstrained) ordination techniques used in EEB is the <em>Principle components analysis</em> (PCA). A PCA seeks to find a new coordinate system (PC axes) that best accounts for the variation in a multivariate data space, thus reducing dimensionality in the process but still accounting for as much of the original variation as possible. Critically, a PCA assumes a linear relationship between your variables, can only take quantitative data, and requires more rows than columns in the data itself.</p>
<p>The procedure to conduct a PCA is as follows:</p>
<ol type="1">
<li><p>Imagine a 14-dimensional object with each of our soil measurements as an axis (k=14).</p></li>
<li><p>Plot our sites (n=24) in this 14-D space, and then find the Euclidean distance between those points (do not despair, <code>R</code> will do the actual calculations for us).</p></li>
</ol>
<p>Let’s visualize this (but in less dimensions because we are human). One way to do this is with a <code>pairs</code> plot in base <code>R</code>, which plots the relationships between all variables (columns) as a scatter plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(varechem[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]) <span class="co">#only the first 4 columns for simplicity</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec11-multivariate-stats_files/figure-html/visualize the correlations-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><em>Which pairs of chemicals are tightly related to each other? Which are not closely related?</em></p>
<ol start="3" type="1">
<li><p>After you have the distances between points, we can use procedures that are similar (in concept) to least squares to find a “best-fit line” through our data points in multidimensional space (again, <code>R</code> will take care of all this). This will be our first principle component (PC) axis and it explains the most variance in our data.</p></li>
<li><p>Find another best-fit line through the data under the condition that it has to be orthogonal to the first axes (i.e., at a 90° angle, meaning that these two axes are as unrelated to each other as possible). This will be our second PC axis. <a href="https://i.stack.imgur.com/JZF5m.gif">See here for visual PCA axis.</a></p></li>
<li><p>This process continues until we have k PC axes.</p></li>
</ol>
<p>Okay, let’s work through some of the details in an example.</p>
<p>We will use the function <code>prcomp</code>. At its most basic, <code>prcomp</code> takes your data and then you specify if you want to <code>scale = TRUE/FALSE</code>. The scale command specifies whether you want to run the pca on a correlation matrix or a covariance matrix. This sounds complex but it boils down to a simple idea. Choosing <code>scale = TRUE</code> means you are selecting a correlation matrix which standardizes your data to have unit variance ie puts all your variables on the same scale. Choose this option if your data contains multiple types of units.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>varechem.pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(varechem, <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Phew, that wasn’t too bad. Let’s take a look at our varechem.pca object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(varechem.pca) <span class="co"># this is ugly becuase of the linebreaks, look at in console, it's easier</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Importance of components:
                          PC1    PC2    PC3     PC4     PC5     PC6     PC7
Standard deviation     2.2785 1.7869 1.2983 1.03392 0.90332 0.84013 0.66061
Proportion of Variance 0.3708 0.2281 0.1204 0.07636 0.05829 0.05042 0.03117
Cumulative Proportion  0.3708 0.5989 0.7193 0.79564 0.85392 0.90434 0.93551
                           PC8    PC9    PC10    PC11    PC12   PC13    PC14
Standard deviation     0.60732 0.4132 0.38669 0.29199 0.26430 0.1872 0.15367
Proportion of Variance 0.02635 0.0122 0.01068 0.00609 0.00499 0.0025 0.00169
Cumulative Proportion  0.96185 0.9740 0.98473 0.99082 0.99581 0.9983 1.00000</code></pre>
</div>
</div>
<p>This gives us a table <em>summary</em> of variance explained by each axes. <code>R</code> even did the math for us and showed cumulative variance explained as we move to the higher dimensions, so convenient! We see that the first two PC axes already explained close to 60% of the variation of the data. But it doesn’t give us any information about our variables or sites!</p>
<p>Let’s have a look at this object in another way.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(varechem.pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 5
 $ sdev    : num [1:14] 2.278 1.787 1.298 1.034 0.903 ...
 $ rotation: num [1:14, 1:14] 0.0946 -0.3559 -0.3632 -0.3598 -0.3523 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:14] "N" "P" "K" "Ca" ...
  .. ..$ : chr [1:14] "PC1" "PC2" "PC3" "PC4" ...
 $ center  : Named num [1:14] 22.4 45.1 162.9 569.7 87.5 ...
  ..- attr(*, "names")= chr [1:14] "N" "P" "K" "Ca" ...
 $ scale   : Named num [1:14] 5.53 14.95 64.84 243.58 41.01 ...
  ..- attr(*, "names")= chr [1:14] "N" "P" "K" "Ca" ...
 $ x       : num [1:24, 1:14] 0.132 0.528 -3.963 -3.241 -1.837 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:24] "18" "15" "24" "27" ...
  .. ..$ : chr [1:14] "PC1" "PC2" "PC3" "PC4" ...
 - attr(*, "class")= chr "prcomp"</code></pre>
</div>
</div>
<p>We see that this object is structured as a five-item list. We mostly care about two of these items:</p>
<ul>
<li><p>Loadings – appropriately named rotation.</p></li>
<li><p>Site scores – lazily named x.</p></li>
</ul>
<p>To extract items from a list, use the $ notation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>varechem.pca<span class="sc">$</span>x <span class="co"># this gives us the particular location (ie the coordinates) of each of our sites in the multidimensional space formed by the PC axes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           PC1         PC2          PC3         PC4          PC5         PC6
18  0.13222328  1.70101442 -0.279475522 -0.47325693  0.103821307 -0.97013110
15  0.52756644  0.75652573  0.004377408 -1.24113026  1.202410202 -0.87748856
24 -3.96343087 -2.54549515 -0.017887116 -1.01474228 -1.896445495  1.65855241
27 -3.24105345  1.03869050  1.093443151  0.74721229  1.122790249  0.14867036
23 -1.83673667  1.74473867 -0.298946184 -0.06989374 -0.804034232 -0.76167251
19 -1.96510029  1.49265376 -1.119483753 -0.37139226 -1.332867226 -0.59222144
22 -0.78264065  1.74589273  0.277038549  0.64096723 -0.079834902  0.23366653
16  1.61990524  1.54627261 -0.182700904 -0.30489556  0.009249015 -0.02669483
28 -2.93871880  0.90850144  0.336742959  1.89138943  0.861994408  0.57878720
13 -1.61179969 -1.72776197 -1.851617499  1.61763757 -0.161168189 -0.99674051
14 -0.10849087  1.23231788 -1.057894796 -0.85576997 -0.235369165  0.24258363
20 -2.41827770  0.41983066 -1.013756555  0.72805394 -0.288590298  0.31608364
25 -0.88868934  1.82326518  1.044623907  0.75063513  1.242461532  0.45818828
7   3.88231709 -0.44736574 -0.728450291  1.01917052 -0.786719915 -0.24475511
5   3.63433625  1.21449816  1.339689070  0.83264806 -0.499185602  2.33214424
6   3.37347840  0.07290487 -0.358467072 -1.02764414  0.555584417  0.44585427
3   4.25544857 -1.67410775 -0.504746514  1.63909051 -0.558770476 -0.18623480
4  -0.68902667 -3.34189696 -2.563741690 -0.71125297  0.968989199  0.43945828
2   1.53878447 -2.46012329 -0.294009378  0.85232874  0.817340772 -0.72299099
9   0.08438668 -1.40485249 -0.514479851 -1.58312456  0.733411043  0.78911410
12  1.23674421  0.53360601  0.271442076 -1.49529509  0.586950936  0.06898806
10 -0.85106814 -1.93158072  2.455295189 -0.60099298  0.837723132 -0.46303287
11 -0.25592697 -3.08135124  3.665611461 -0.10586883 -1.117079105 -1.05247280
21  1.26576946  2.38382268  0.297393354 -0.86387385 -1.282661605 -0.81765547
            PC7         PC8         PC9        PC10        PC11         PC12
18 -0.412107828  0.10430066  0.89379861 -0.19515377  0.15181987  0.446288151
15  0.140335713  0.35570552  0.18415914 -0.32702920 -0.26463780  0.216772715
24  0.584842783  0.19856484  0.59859756 -0.34214482 -0.03469025  0.310027914
27  0.577453010 -0.34781468 -0.05752049 -0.59425418  0.07314193  0.063378029
23  0.030023854 -0.13218763  0.51649591  0.70693203  0.03233029  0.054332723
19  1.238669139 -0.34158207 -0.31889053 -0.05364217 -0.16861193 -0.476514833
22  0.539332177 -0.05889938 -0.51844860  0.03006888  0.45835909  0.245481007
16 -0.373476941  0.78419427 -0.06115766  0.48740293  0.23745992  0.059594599
28  0.267720379 -0.27061746  0.09308865  0.99318331 -0.21893369  0.097099274
13  0.028618132  1.41746150 -0.78794716 -0.41248879 -0.06593851  0.293911543
14  0.117482816  0.34746586 -0.11334200  0.29864105  0.21053275 -0.319591277
20 -1.808311684 -0.48521666  0.13191310 -0.29763388 -0.11048490 -0.198738243
25 -0.296177146 -0.70116183 -0.08870626 -0.56597254  0.03863706 -0.175523779
7   0.003896015 -0.07322921 -0.07022198  0.03076471 -0.32299944  0.093391136
5  -0.372775208  1.06173977  0.03575899 -0.10222536  0.04588475 -0.150519119
6   0.437849242 -0.66318161 -0.02922027 -0.12788259  0.20774098  0.368308444
3   0.473990937 -1.15101958  0.18069717  0.01462754 -0.33603215  0.155223171
4  -1.092945035 -0.43319856 -0.16366937  0.32388296  0.31092450 -0.016558671
2   0.764129828  0.52251461  0.87005247 -0.21039984  0.45544416 -0.635306833
9   0.313163663 -0.40901417 -0.38867009  0.02167642 -0.40521219 -0.212953448
12  0.473997752  0.13222574 -0.44070463  0.20907903  0.10370214  0.071014379
10 -0.259576328  0.80876134  0.14111906  0.27398425 -0.67108522 -0.072769123
11 -0.473633258 -0.53811517 -0.46308153  0.16295082  0.47248237  0.001778163
21 -0.902502014 -0.12769610 -0.14410013 -0.32436679 -0.19983373 -0.218125923
           PC13        PC14
18  0.180710086 -0.14511706
15 -0.264956749  0.10969811
24 -0.084404758  0.05320728
27 -0.169590445 -0.18408979
23  0.215246883 -0.07599179
19 -0.109832641 -0.17267154
22  0.344342190  0.09944716
16 -0.068810220  0.04176308
28 -0.286368123  0.10527341
13  0.019257591  0.08668395
14 -0.001286276  0.11266438
20  0.218426950  0.21618583
25  0.079312882 -0.09860196
7   0.116566189 -0.41080566
5  -0.068070865 -0.09223078
6  -0.078853433  0.10186206
3  -0.036095816  0.23450059
4  -0.232369489 -0.25165355
2   0.038260431  0.10380174
9   0.379813850  0.05207296
12  0.041195274  0.05293037
10  0.132399959 -0.03854731
11 -0.044201689 -0.01970105
21 -0.320691780  0.11931958</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>varechem.pca<span class="sc">$</span>rotation <span class="co"># this tells us which of our soil chemicals are contributing most strongly to each PC axis (ie which variables are being summarized by each newly created PC axis)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 PC1         PC2         PC3         PC4          PC5
N         0.09462709  0.12435859 -0.16204826  0.79318367 -0.351576656
P        -0.35589258 -0.20731069  0.09086948  0.08242977  0.219919629
K        -0.36318854 -0.16172669 -0.04259682  0.21401371  0.176886144
Ca       -0.35975485 -0.08457237  0.32644437 -0.01826328 -0.236121676
Mg       -0.35228219 -0.09776116  0.08467641 -0.03784222 -0.486392688
S        -0.33246862 -0.29668916 -0.21512714 -0.01410909  0.070039686
Al        0.11552207 -0.48048232 -0.26540100  0.07767669  0.005266422
Fe        0.20299687 -0.38301396 -0.15691339  0.31769008 -0.022858769
Mn       -0.28756460  0.19318776  0.08921960  0.37885343  0.489948050
Zn       -0.34603172 -0.16198093  0.04178581 -0.05229606 -0.150267479
Mo       -0.04154531 -0.22863241 -0.58319905 -0.23286801 -0.029395588
Baresoil -0.16349721  0.31570370 -0.27274616 -0.06312303 -0.407367062
Humdepth -0.23531857  0.33736134 -0.28352933  0.02541797 -0.046678588
pH        0.16613845 -0.32216601  0.45164891  0.06492250 -0.259486172
                 PC6          PC7          PC8          PC9        PC10
N         0.31876043 -0.020445294  0.106079920 -0.029711941  0.25325314
P         0.08346185 -0.202993610 -0.291937970  0.529497898  0.48691417
K        -0.30156737 -0.083967201  0.345810459 -0.347935071  0.19105757
Ca       -0.10026235 -0.056467690 -0.241640431 -0.132902952  0.15581868
Mg       -0.11335320  0.422483826  0.002896128  0.185900569 -0.18858761
S        -0.09229889 -0.001743922  0.271322793 -0.041707742 -0.01394397
Al       -0.20467312 -0.090888045 -0.037652261 -0.325530088 -0.02736515
Fe       -0.33950608  0.197580497 -0.192886023  0.424437201 -0.32652805
Mn        0.07123733 -0.136543653 -0.274959999 -0.080205160 -0.54341919
Zn        0.54306692  0.032492006  0.238361969 -0.005992104 -0.36679941
Mo        0.41761742 -0.120157357 -0.352212047 -0.082184113  0.02092970
Baresoil -0.32655495 -0.662017409 -0.006567573  0.150678212 -0.20591182
Humdepth -0.16741045  0.425702015 -0.465061249 -0.315633320  0.12595591
pH        0.05330910 -0.262135909 -0.374832351 -0.349371603 -0.07947962
                 PC11        PC12        PC13         PC14
N        -0.005103292  0.03375847  0.12866179 -0.023577121
P        -0.212772662 -0.15369608 -0.16014570 -0.146803748
K         0.118589614  0.28135388 -0.50806589  0.191151759
Ca        0.669374213 -0.03834446  0.36929697  0.035843646
Mg       -0.239307771  0.42332112 -0.05980040 -0.345891242
S        -0.378600375 -0.07818118  0.63006839  0.345755661
Al        0.018994101 -0.30010097  0.04257464 -0.654805487
Fe        0.238879497 -0.09798831 -0.12533928  0.354078026
Mn       -0.035024496  0.18554028  0.13661147 -0.181857434
Zn        0.104956310 -0.48566331 -0.29628189  0.070169749
Mo        0.159957373  0.43567172 -0.03761246  0.117855258
Baresoil -0.053374989 -0.10187147 -0.07709019  0.007445089
Humdepth -0.168085863 -0.37001287 -0.13026298  0.156358945
pH       -0.406958380  0.04733921 -0.12013250  0.265236656</code></pre>
</div>
</div>
<p>We can interpret these outputs directly, specifically the rotations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>varechem.pca<span class="sc">$</span>rotation[ ,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>] <span class="co"># let's just look examine the first 4 PC axes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 PC1         PC2         PC3         PC4
N         0.09462709  0.12435859 -0.16204826  0.79318367
P        -0.35589258 -0.20731069  0.09086948  0.08242977
K        -0.36318854 -0.16172669 -0.04259682  0.21401371
Ca       -0.35975485 -0.08457237  0.32644437 -0.01826328
Mg       -0.35228219 -0.09776116  0.08467641 -0.03784222
S        -0.33246862 -0.29668916 -0.21512714 -0.01410909
Al        0.11552207 -0.48048232 -0.26540100  0.07767669
Fe        0.20299687 -0.38301396 -0.15691339  0.31769008
Mn       -0.28756460  0.19318776  0.08921960  0.37885343
Zn       -0.34603172 -0.16198093  0.04178581 -0.05229606
Mo       -0.04154531 -0.22863241 -0.58319905 -0.23286801
Baresoil -0.16349721  0.31570370 -0.27274616 -0.06312303
Humdepth -0.23531857  0.33736134 -0.28352933  0.02541797
pH        0.16613845 -0.32216601  0.45164891  0.06492250</code></pre>
</div>
</div>
<p>The way we interpret this is by looking, within each PC axis, for the variables that have the largest absolute values. These are the variables that are contributing variation to this axis.</p>
<p>For example, PC1 sets up an axis of increasing concentrations of P, K, Ca, Mg, S, Zn. PC2 is driven largely by Al, Baresoil, Humdepth and pH. Note that the positive and negative numbers indicate a positive and negative relationship (eg along PC1 P and Fe are negatively correlated, sites that have high P tend to have low Fe and vice versa).</p>
<p>This is much, much easier to see visually. We can do this with the base <code>R</code> function <code>biplot</code>, which is a bit ugly.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(varechem.pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec11-multivariate-stats_files/figure-html/biplot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>On this figure, there are 2 sets of information. Each site is plotted (these are the numbers). As well, the variable loadings are represented as arrows and tell us how each variable is contributing to each axis. The clustering of sites along the axes (and rotations) tell us which sites are similar in terms of their chemical composition. Eg, we can see that sites 19, 20, 27 and 28 all have a similar chemical composition. We can also see clearly that along PC2 sites with low pH also tend to have low Al, while they have high Humdepth, Baresoil and Mn. Future analyses might relate the soil characteristics of these sites to other environmental data (history of land use, pollution etc)</p>
<p>We can do this with the function <code>autoplot</code> in the <code>ggfortify</code> package, which takes advantage of ggplot style plotting.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(varechem.pca) <span class="sc">+</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec11-multivariate-stats_files/figure-html/autoplot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The simplest version of <code>autoplot</code> displays less information than our <code>biplot</code> function. On this figure we can see that we are plotting PC1 by PC2 and each point on the figure is each of the rows aka sites in our data. <code>autoplot</code> handily also shows us the proportion of variance explained by each of these PC axes. Now lets add the loadings.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remember to tell R where the original data is coming from! </span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Also, `autoplot` will refuse to understand you if you spell "colour" the non-Canadian way</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(varechem.pca, <span class="at">loadings=</span><span class="cn">TRUE</span>, <span class="at">loadings.label=</span><span class="cn">TRUE</span>, </span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">loadings.colour=</span><span class="st">"grey30"</span>, </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">data=</span>varechem) <span class="sc">+</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec11-multivariate-stats_files/figure-html/add the loadings-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>But do we make biplot for each of our PC axes??? Nope, this would take too long and each subsequent axis explains less of the variation in our data… ie it becomes less important. As well, we often perform PCA to simplify our data from many variables to few.</p>
<p>So the next step is to pick how many axes show meaningful information.</p>
<p>We can do this with a scree plot. This shows us the amount of variation explained by each axis. As we know, each axis explains subsequently less variation. We use scree plots to help us determine at which point the axes stop adding much new information. We can do this visually or quantitatively. We will just go over the visual method as determining which axes to keep is pretty subjective and depends on your system and question of interest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(varechem.pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec11-multivariate-stats_files/figure-html/scree plot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Each bar is a PC axis, starting from PC 1. In a scree plot, what you want to look for is a big jump between subsequent axes. Its subjective, but based on this scree plot I would probably be inclined to keep the first 2 or 3 PC axes. As well, we can look at the cumulative proportion of explained variance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(varechem.pca)<span class="sc">$</span>importance[,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            PC1      PC2      PC3      PC4      PC5      PC6
Standard deviation     2.278498 1.786853 1.298271 1.033922 0.903324 0.840125
Proportion of Variance 0.370830 0.228060 0.120390 0.076360 0.058290 0.050420
Cumulative Proportion  0.370830 0.598890 0.719280 0.795640 0.853920 0.904340</code></pre>
</div>
</div>
<p>And we can see that by PC3 we have explained ~72% of the variation in the data, which is not bad (not great either, but not bad). By the time we get to PC4, each individual axis is explaining less than 10% of variation in the data, so adding them, and conversely excluding them, isn’t going to have a large effect.</p>
<p>*Note 1: there are many functions across many libraries for running a pca: <code>prcomp</code>, <code>princomp</code>, <code>rda</code>, <code>dudi.rda</code>. Each method makes a slightly different choice about how the underlying math is performed. <code>prcomp</code> and <code>rda</code> are both very common choices.</p>
<p>*Note 2: for those of you working with phylogenetic data: you may want to perform a phylogenetic PCA (pPCA). The function <code>phyl.pca</code> is available in package <code>phytools</code>.</p>
<p>For further readings see:</p>
<ul>
<li><p><a href="https://towardsdatascience.com/principal-component-analysis-pca-explained-visually-with-zero-math-1cbf392b9e7d">PCA: a visual explanation</a></p></li>
<li><p><a href="https://stats.stackexchange.com/questions/2691/making-sense-of-principal-component-analysis-eigenvectors-eigenvalues">PCA explained using a metaphor with wine, dinner, grandmothers, and animations</a>PCoA</p></li>
</ul>
</section>
<section id="pcoa" class="level3" data-number="14.2.3">
<h3 data-number="14.2.3" class="anchored" data-anchor-id="pcoa"><span class="header-section-number">14.2.3</span> PCoA</h3>
<p>A PCoA (principal coordinates analysis) is a more flexible form of a PCA that can take any type of distance matrix. Remember that distance matrices allow you to express differences between your data points so by specifying the type of distance matrix, so this means that PCoA can handle a mix of data types—which is very handy if you have both categorical and numeric data. Note that your choice of distance matrix can have a large effect on the conclusions you draw in your analysis–make sure to read the literature and learn what best practices there are for you specific data types.</p>
<p>You interpret the output of a PCoA the same way as a PCA. Common functions are <code>cmdscale</code> in base <code>R</code>, <code>wcmdscale</code> in <code>vegan</code>, and <code>pcoa</code> in <code>ape</code>.</p>
<p>In sum, both PCA and PCoA preserve the distances between your data points–for this reason they are commonly called “rotations”–these methods do not change the underlying relationships in your data, they just group variation in data across many dimensions and then make a nice tidy output that our brain’s can interpret.</p>
</section>
</section>
<section id="lets-practice-real-life-example" class="level2" data-number="14.3">
<h2 data-number="14.3" class="anchored" data-anchor-id="lets-practice-real-life-example"><span class="header-section-number">14.3</span> Let’s practice: real life example</h2>
<p>For this section of the class, we will try to apply what we’ve learned above on a real world data set. We’ll be walking through the analysis and the choices we make and then at each section, I will give you a minute or so to try and either a. shout out the answer or b. write the code.</p>
<p>We are working with a cool dataset that was collected by Vicki as part of her PhD work outside of Churchill, Manitoba. Vicki generally studies invasive species in the tundra and the data we are working with today is part of a larger project that investigates the effect of soil on plant invasions.</p>
<p>The key question we are trying to answer today is if soil chemicals, which we expect to vary across a landscape according to human presence and land use, can predict whether or not invasive species are present.</p>
<p>Let’s read in our data and take a look at it!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>soil.data <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"soil.data.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 28 Columns: 11
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): sample_id, invaded_status, location, notes
dbl (7): site_number, latitude, longitude, P, C, N, avg_pH

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(soil.data) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 11
  sample_id site_number invaded_status location latitude longitude     P     C
  &lt;chr&gt;           &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;       &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 1A                  1 invaded        town         58.8     -94.2 95.2   9.92
2 1B                  1 uninvaded      town         58.8     -94.2 42.4   5.47
3 6A                  2 invaded        town         58.8     -94.2  4.63  2.22
4 6B                  2 uninvaded      town         58.8     -94.2  8.31  1.74
5 7A                  3 invaded        town         58.8     -94.2 33.6   5.49
6 7B                  3 uninvaded      town         58.8     -94.2 77.2   4.93
# ℹ 3 more variables: N &lt;dbl&gt;, avg_pH &lt;dbl&gt;, notes &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(soil.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>spc_tbl_ [28 × 11] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ sample_id     : chr [1:28] "1A" "1B" "6A" "6B" ...
 $ site_number   : num [1:28] 1 1 2 2 3 3 4 4 5 5 ...
 $ invaded_status: chr [1:28] "invaded" "uninvaded" "invaded" "uninvaded" ...
 $ location      : chr [1:28] "town" "town" "town" "town" ...
 $ latitude      : num [1:28] 58.8 58.8 58.8 58.8 58.8 ...
 $ longitude     : num [1:28] -94.2 -94.2 -94.2 -94.2 -94.2 ...
 $ P             : num [1:28] 95.2 42.4 4.63 8.31 33.6 77.2 12.1 22.5 7.55 11.4 ...
 $ C             : num [1:28] 9.92 5.47 2.22 1.74 5.49 4.93 4.4 7.7 6.07 6.86 ...
 $ N             : num [1:28] 0.52 0.12 0 0 0.1 0.2 0.11 0.35 0.04 0.07 ...
 $ avg_pH        : num [1:28] 7.57 7.45 7.56 7.5 7.46 ...
 $ notes         : chr [1:28] NA NA "total N &lt; 0.02" "total N &lt; 0.02" ...
 - attr(*, "spec")=
  .. cols(
  ..   sample_id = col_character(),
  ..   site_number = col_double(),
  ..   invaded_status = col_character(),
  ..   location = col_character(),
  ..   latitude = col_double(),
  ..   longitude = col_double(),
  ..   P = col_double(),
  ..   C = col_double(),
  ..   N = col_double(),
  ..   avg_pH = col_double(),
  ..   notes = col_character()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
</div>
<p>Okay, so in this data, we have columns that contain our environmental data: the concentration of various chemicals in the soil, if there are invasive species (categorical, two levels, invaded/uninvaded), and then a bunch of spatial information that relates each site to a broader geographic region ( 3 levels, town, airport, CNSC). We may expect that each of these broader regions will have a particular soil chemical composition, which then should (according to our question) affect the presence of invasive species.</p>
<p>More formally then, we are testing for a relationship between location, P, C, N, and avg_pH (our explanatory variables) and invaded_status (our response variable).</p>
<p>Before we get started, let’s do some data organization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(soil.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "sample_id"      "site_number"    "invaded_status" "location"      
 [5] "latitude"       "longitude"      "P"              "C"             
 [9] "N"              "avg_pH"         "notes"         </code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert response into 1s and 0s</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>soil.data<span class="sc">$</span>binary.invaded_status <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(soil.data<span class="sc">$</span>invaded_status<span class="sc">==</span><span class="st">"invaded"</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co"># make sure that location is a factor</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>soil.data<span class="sc">$</span>location <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(soil.data<span class="sc">$</span>location)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="multicollinearity" class="level3" data-number="14.3.1">
<h3 data-number="14.3.1" class="anchored" data-anchor-id="multicollinearity"><span class="header-section-number">14.3.1</span> Multicollinearity</h3>
<p>The first thing we need to do is assess the relationships between our explanatory variables (location, P, C, N, avg_pH) and check for <strong>multicollinearity</strong>.</p>
<p>Multicollinearity refers to a situation in which <strong>two or more</strong> explanatory variables are highly linearly related (See <a href="http://web.vu.lt/mif/a.buteikis/wp-content/uploads/PE_Book/4-5-Multiple-collinearity.html">here</a> for a mathematical explanation). This can wreck havoc in an analysis because collinear predictors generate unstable and unreliable regression coefficients that have large standard errors.</p>
<p>So the first step in this analysis is to visualize all our environmental data and assess whether any variables are highly correlated with each other (ie are collinear).</p>
<section id="correlations" class="level4" data-number="14.3.1.1">
<h4 data-number="14.3.1.1" class="anchored" data-anchor-id="correlations"><span class="header-section-number">14.3.1.1</span> Correlations</h4>
<p>There are a couple of ways of assessing collinearity of our predictor variables. One common way you may have heard of is to calculate the correlation between the variables. We do this using <code>cor</code> and can visualize the relationships using <code>pairs</code>. Note that this method only works for numeric data (ie it can’t handle categorical variables).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(soil.data[,<span class="dv">7</span><span class="sc">:</span><span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec11-multivariate-stats_files/figure-html/visualize data-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(soil.data[,<span class="dv">7</span><span class="sc">:</span><span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                P          C          N     avg_pH
P       1.0000000  0.5090295  0.6482000 -0.4091076
C       0.5090295  1.0000000  0.9086144 -0.4722516
N       0.6482000  0.9086144  1.0000000 -0.5681717
avg_pH -0.4091076 -0.4722516 -0.5681717  1.0000000</code></pre>
</div>
</div>
<p>We can see from this that C and N are highly related. When we look at the correlations, this linked relationship is confirmed. As a rule of thumb, ecologists generally report an r&gt;=0.70 as being collinear. Note that this threshold is a bit arbitrary and different sub-disciplines have different traditions.</p>
<p>Based off this output, we could choose to drop either C or N from the analysis as they will explain similar variation in the data. Generally, we would choose to drop whichever variable we were less interested in scientifically.</p>
</section>
<section id="vif-variance-inflation-factor" class="level4" data-number="14.3.1.2">
<h4 data-number="14.3.1.2" class="anchored" data-anchor-id="vif-variance-inflation-factor"><span class="header-section-number">14.3.1.2</span> VIF = Variance Inflation Factor</h4>
<p>The other common way of assessing collinearity is with the VIF. This method can handle both categorical and numeric data.</p>
<p>The VIF is calculated for each predictor by doing a linear regression of that predictor on all the other predictors, and then obtaining the <em>R<sup>2</sup></em> from that regression. The VIF is just 1/(1-<em>R<sup>2</sup></em>). This is why it’s called the variance inflation factor - it estimates how much the variance of a coefficient is “inflated” due to linear dependence with other predictors.</p>
<p>VIF’s start at 1 and have no upper limit. In general:</p>
<ul>
<li><p>VIF = 1 means no correlation between predictor and any other predictor variables in model</p></li>
<li><p>1 &lt; VIF &lt; 5 suggests moderate collinearity</p></li>
<li><p>VIF &gt; 5 means severe collinearity</p></li>
</ul>
<p>Note that there are some cases in which high VIF values can be ignored, for example for polynomial terms (see <a href="https://statisticalhorizons.com/multicollinearity/">here</a> for a detailed discussion).</p>
<p>VIF is used on model rather than on the raw predictors. So let’s fit our model and include location (town vs airport) and see if any of our predictors are collinear. A VIF above 3 (or also commonly 5) indicate high collinearity.</p>
<p>Note that for 28 observations this model is also very over fit (too many predictor terms to get general model parameters), but let’s ignore that for now.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>model_collinear <span class="ot">&lt;-</span> <span class="fu">glm</span>(binary.invaded_status <span class="sc">~</span> location <span class="sc">+</span> P <span class="sc">+</span> C <span class="sc">+</span> N <span class="sc">+</span> avg_pH, </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">data=</span>soil.data, </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">family=</span><span class="st">"binomial"</span>)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>car<span class="sc">::</span><span class="fu">vif</span>(model_collinear)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             GVIF Df GVIF^(1/(2*Df))
location 1.466368  2        1.100426
P        2.441326  1        1.562474
C        5.081174  1        2.254146
N        8.102951  1        2.846568
avg_pH   1.462175  1        1.209204</code></pre>
</div>
</div>
<p>We have multiple levels in our categorical predictor (airport/town/CNSC) and so we are returning GVIF which are produce better estimates when the degrees of freedom &gt; 1. We interpret GVIF the same was as VIF and apply the same rule of thumb.</p>
<p><em>Which variables are collinear? Does this match our conclusion from using cor() above?</em></p>
<p>Just as above, we can see the C and N are collinear and have a high VIF. Generally, best practice is to drop the term with the highest VIF and rerun, and keep going, dropping terms until all VIF are below the threshold for collinearity.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>model_collinear_dropN <span class="ot">&lt;-</span> <span class="fu">glm</span>(binary.invaded_status <span class="sc">~</span> location <span class="sc">+</span> P <span class="sc">+</span> C <span class="sc">+</span> avg_pH, </span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">data=</span>soil.data, </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">family=</span><span class="st">"binomial"</span>)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="fu">vif</span>(model_collinear_dropN) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             GVIF Df GVIF^(1/(2*Df))
location 1.315957  2        1.071052
P        1.964871  1        1.401739
C        1.635915  1        1.279029
avg_pH   1.306870  1        1.143184</code></pre>
</div>
</div>
<p><em>How does VIF change if drop N? How does the output change? What can we conclude from this?</em></p>
<p>Note that dropping variables due to collinearity isn’t p-hacking. In your methods, you would report all the predictors and your method of checking for collinearity, and what threshold you use, and which variables you exclude.</p>
<p>Okay, so what happens though if N is actually our variable of interest and we’ve just gotten rid of it from the model??? Do we panic??</p>
<p>No! Luckily we’ve just learned about multivariate techniques for grouping correlated data into common axes of variation.</p>
</section>
<section id="pca-1" class="level4" data-number="14.3.1.3">
<h4 data-number="14.3.1.3" class="anchored" data-anchor-id="pca-1"><span class="header-section-number">14.3.1.3</span> PCA</h4>
<p><em>Which method might we use on this data?</em></p>
<p>PCA is also useful in this case for exploring how these different soil measurements are organized across sites.</p>
<p><em>Should we use a covariance or correlation matrix?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pca on data</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co"># each row is a site</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(soil.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "sample_id"             "site_number"           "invaded_status"       
 [4] "location"              "latitude"              "longitude"            
 [7] "P"                     "C"                     "N"                    
[10] "avg_pH"                "notes"                 "binary.invaded_status"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make sure to select columns that you want to include in the PCA!</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>pca.soil <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(soil.data[, <span class="fu">c</span>(<span class="dv">7</span><span class="sc">:</span><span class="dv">10</span>)], <span class="at">scale. =</span> <span class="cn">TRUE</span>) <span class="co"># use corr bc multiple units</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s look at the output.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># look at importance of components</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca.soil)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec11-multivariate-stats_files/figure-html/soil loadings and scree plot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><em>How many axes should we keep?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># let's look at loadings</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>pca.soil<span class="sc">$</span>rotation</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              PC1       PC2        PC3        PC4
P      -0.4561476 0.3487729 -0.8033374  0.1579108
C      -0.5326177 0.2372107  0.5269634  0.6183519
N      -0.5713178 0.1657433  0.2459331 -0.7652725
avg_pH  0.4264314 0.8914133  0.1283571 -0.0840417</code></pre>
</div>
</div>
<p><em>So how would we interpret this? Let’s start with PC1.</em></p>
<p>Now let’s visualize, it’ll be much easier to interpret! We’re just going to use <code>autoplot</code> here, but remember the base plot function <code>biplot</code> also works.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(pca.soil, <span class="at">loadings=</span><span class="cn">TRUE</span>, <span class="at">loadings.label=</span><span class="cn">TRUE</span>, </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">loadings.colour =</span> <span class="st">"grey30"</span>,  <span class="co"># changes the colour of the arrows</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">data =</span> soil.data, </span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">colour=</span><span class="st">"location"</span>, <span class="co"># colour obs by location</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">loadings.label.hjust =</span> <span class="fl">1.2</span>) <span class="sc">+</span> <span class="co"># moves loading labels so easier to read</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span>  <span class="co"># make fancier (removes box)</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span> <span class="fu">wes_palette</span>(<span class="st">"FantasticFox1"</span>, <span class="at">n =</span> <span class="dv">3</span>))  <span class="co"># pretty colours!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec11-multivariate-stats_files/figure-html/soil biplot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><em>What jumps out to you about this figure? What are some of the things this tells us?</em></p>
<ol type="1">
<li>PC1 explains most of variation and is doing most of heavy lifting</li>
<li>P, C, and N are all highly correlated together (arrows of similar length and pointing in same direction)</li>
<li>pH is relatively uncorrelated (orthogonal to P, C, N)</li>
<li>P, C, N load strongly on PC1, setting up an axis of more to less concentration</li>
<li>pH loads mostly (but not exclusively) on PC2</li>
<li>If you squint <em>maybe</em> the different locations are grouped in PC space… but more likely it looks like location doesn’t explain where the data (sites) fall out in relation to P, C, N, and pH (ah well, this is what happens when we work with real data!)</li>
</ol>
<p>Let’s run our same logistic model again to test if soil characteristics explain the relationship between if a site and invasive species or not. This time, instead of using all the soil data, we will use PC axes.</p>
<p>But first, some (more) data organization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># see each observation and it's location along PC axes</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>pca.soil<span class="sc">$</span>x</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              PC1         PC2         PC3          PC4
 [1,] -0.23126104  0.77648694 -0.84121494  0.235136523
 [2,]  0.84406257 -0.50103455 -0.34039831  0.114929541
 [3,]  1.88826355 -0.50887992  0.19959381 -0.253290787
 [4,]  1.75535290 -0.76347862  0.04570655 -0.258661129
 [5,]  0.99177246 -0.52426046 -0.15590610  0.098831472
 [6,]  0.53969902  0.08152624 -1.03403889  0.085747726
 [7,]  1.66916888 -0.03427478  0.30698806 -0.171959835
 [8,]  0.80122422 -0.31463880  0.37058135 -0.043269362
 [9,]  1.93351386  0.56837626  0.59677085 -0.005952411
[10,]  1.63224314  0.29434028  0.54303874  0.082469913
[11,] -2.71397334  0.75174581 -0.24500050 -0.299725322
[12,] -0.03236594  0.14566698 -0.31835184  0.157973860
[13,] -0.84681650 -1.12969032 -1.02413728  0.043433577
[14,]  0.63779987 -0.13876400 -0.37249968 -0.141932555
[15,] -2.10740961  0.03617329 -0.86306690  0.051402935
[16,] -2.36369664  0.96422202 -0.37958188 -0.279045169
[17,] -2.97841280 -0.28704561  0.51688116 -0.282026062
[18,] -1.56857576  0.86351100 -1.57655687  0.058062467
[19,] -0.87738490  0.53086067  0.92362157  1.051842622
[20,]  0.13763151  0.16316485  0.27107051  0.304104023
[21,] -4.08292912 -0.12691283  2.11374890 -0.146981087
[22,] -0.03084636 -0.65130088  0.49727778 -0.033064882
[23,]  0.18445264 -0.13967958 -0.06872346  0.014528613
[24,]  1.72798063  0.92802356  0.10157757 -0.193211086
[25,]  0.99564140  0.86534818  0.45856629 -0.223903218
[26,] -0.74210843 -2.72156445 -0.10274029  0.116180707
[27,]  0.75530840  0.17215662 -0.36418195 -0.039898398
[28,]  2.08166538  0.69992209  0.74097578 -0.041722676</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># bind the observations to the data, note that `R` keeps everything in the same order, so this is easy</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>soil.data<span class="sc">$</span>PC1 <span class="ot">&lt;-</span> pca.soil<span class="sc">$</span>x[ ,<span class="dv">1</span>]  <span class="co"># pull PC1</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>soil.data<span class="sc">$</span>PC2 <span class="ot">&lt;-</span> pca.soil<span class="sc">$</span>x[ ,<span class="dv">2</span>]  <span class="co"># pull PC2</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(soil.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 14
  sample_id site_number invaded_status location latitude longitude     P     C
  &lt;chr&gt;           &lt;dbl&gt; &lt;chr&gt;          &lt;fct&gt;       &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 1A                  1 invaded        town         58.8     -94.2 95.2   9.92
2 1B                  1 uninvaded      town         58.8     -94.2 42.4   5.47
3 6A                  2 invaded        town         58.8     -94.2  4.63  2.22
4 6B                  2 uninvaded      town         58.8     -94.2  8.31  1.74
5 7A                  3 invaded        town         58.8     -94.2 33.6   5.49
6 7B                  3 uninvaded      town         58.8     -94.2 77.2   4.93
# ℹ 6 more variables: N &lt;dbl&gt;, avg_pH &lt;dbl&gt;, notes &lt;chr&gt;,
#   binary.invaded_status &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="modelling" class="level4" data-number="14.3.1.4">
<h4 data-number="14.3.1.4" class="anchored" data-anchor-id="modelling"><span class="header-section-number">14.3.1.4</span> Modelling</h4>
<p>Okay, now let’s run the model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>model_PCA <span class="ot">&lt;-</span> <span class="fu">glm</span>(binary.invaded_status <span class="sc">~</span> location <span class="sc">+</span> PC1, </span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">data =</span> soil.data, </span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">family =</span> <span class="st">"binomial"</span>)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="fu">vif</span>(model_PCA) <span class="co"># nice, this is what we want to see</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             GVIF Df GVIF^(1/(2*Df))
location 1.136794  2        1.032572
PC1      1.136794  1        1.066205</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model_PCA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = binary.invaded_status ~ location + PC1, family = "binomial", 
    data = soil.data)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.4490  -1.1062  -0.1187   1.1441   1.5497  

Coefficients:
             Estimate Std. Error z value Pr(&gt;|z|)
(Intercept)   -0.1850     0.6195  -0.299    0.765
locationCNSC   0.4488     1.2323   0.364    0.716
locationtown   0.2988     0.8845   0.338    0.736
PC1           -0.3401     0.2685  -1.267    0.205

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 38.816  on 27  degrees of freedom
Residual deviance: 37.037  on 24  degrees of freedom
AIC: 45.037

Number of Fisher Scoring iterations: 4</code></pre>
</div>
</div>
<p>Ah well, it looks like there’s no relationship between soil characteristics and invasion.</p>
<p>We can actually see this very clearly if we just plot our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(soil.data, <span class="fu">aes</span>(PC1, binary.invaded_status, <span class="at">colour=</span>location)) <span class="sc">+</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">3</span>, <span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span>  </span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span> <span class="fu">wes_palette</span>(<span class="st">"FantasticFox1"</span>, <span class="at">n =</span> <span class="dv">3</span>))  <span class="sc">+</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Invasion success"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec11-multivariate-stats_files/figure-html/final plot-1.png" class="img-fluid" width="672"></p>
</div>
</div>


</section>
</section>
</section>
</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p>This explanation was largely lifted verbatim from Borcard, D., Gillet, F. and Legendre, P. 2011. <em>Numerical Ecology with R.</em> Chapter 5, page 116.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./lec10-model-selection.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Model selection</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./lec12-command-line_git.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Command line &amp; Git(Hub)</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>