<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>EEB313: Quantitative Methods in R for Biology - 2&nbsp; Introduction to R: assignment, vectors, functions, strings, loops</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../lectures/lec03-command_git.html" rel="next">
<link href="../lectures/lec01-r_and_rstudio.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lectures/lec01-r_and_rstudio.html">Lectures</a></li><li class="breadcrumb-item"><a href="../lectures/lec02-basic-r.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to R: assignment, vectors, functions, strings, loops</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">EEB313: Quantitative Methods in R for Biology</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/EEB313/eeb313.github.io" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Syllabus</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../about-us.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024 teaching team</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../downloadingR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Installing R &amp; Ubuntu</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Lectures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec01-r_and_rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to the course</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec02-basic-r.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to R: assignment, vectors, functions, strings, loops</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec03-command_git.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Command line &amp; Git(Hub)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec04-data-wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data wrangling!</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec05-data-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data visualization with ggplot2</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec06-exploratory-data-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploratory data analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec07-self-directed-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Review using Farrell and Davies (2019) dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec08-intro-inference-1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to inference I</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec09-intro-inference-2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to inference II</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec10-linear-models-1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Linear models I</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/lec11-linear-models-2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Linear models II

## Lesson preamble

&gt; ### Learning objectives
&gt;
&gt; - Learn how dependent data can be modeled, and the sitations in which one may encounter such data.
&gt; - Understand the structure, assumptions, and implementation of GLMs.
&gt; - Understand the structure, assumptions, and implementation of LLMs.
&gt; - Learn how to use simulation to inform what data that should be collected, and how.
&gt; 
&gt; ### Lesson outline
&gt; -   Generalized linear models
&gt;     - Structure and assumptions, including interpretation of the effects
&gt;     - Implement logistic regression using dataset of Farrell and Davies (2019)
&gt;     - Other types of GLMs (Poisson, negative binomial, etc.)
&gt; -   Power analysis!
&gt; -   Dealing with dependent data
&gt;     - Splitting the data into groups...
&gt;     - Controlling for phylogeny
&gt;     - Linear mixed models: structure and assumptions
&gt;       - Difference between fixed, random effects; examples
&gt;       - Implement mixed models using sexual size dimorphism data

```{r, message=F, warning=F}
library(tidyverse)
library(broom)
library(lme4)
```

## Generalized linear models: theory and examples

So far we have seen how linear models can be used to describe relationships between a response variable and predictors when the data is normally distributed or can be transformed so that this assumption is not violated. What if we were, say, interested in a *binary* response (e.g., infection status of a host with a particular parasite) and how it changes with a continuous predictor (e.g., age of the host)? In this case, one can use a special kind of linear model called *logistic regression* to estimate the additive effect of predictor(s) on the binary response. Logistic regression is a special kind of **generalized linear model**.

### GLMs: structure and assumptions

Generalized linear models describe how the mean of a response variable changes as a function of the predictors when important assumptions of the linear modeling framework (normality, constant error variance, etc.) are violated. In particular, **GLMs allow us to work with data that are not normal, whose range is restricted, or whose variance depends on the mean.** The latter might be important if, say, larger values of the response were also more variable.

A GLM models the transformed mean of a response variable $Y_i$ as a linear combination of the predictors $x_1,\dots,x_p$. The goal of using a GLM is often to estimate how the predictors (e.g., sex and previous exposure to the disease) affect the response (e.g., infection status). sThe transformation of the response is done using a "link" function $g(\cdot)$ that is specific to the distribution which used to model the data. Written out, a GLM is of the form

$$g(\mu_i) = \beta_0 + \beta_1 x_{1i} + \cdots \beta_p x_{pi}.$$

The link functions for the Normal, Gamma, Exponential, Poisson, and Multinomial distributions are known. In general, GLMs apply when the data are modeled using a member of the [exponential family](https://en.wikipedia.org/wiki/Exponential_family). The distributions will have a mean parameter $\mu$ and, sometimes, a parameter $\tau$ which characterizes the dispersion around the mean. _GLMs are fitted using by maximizing the likelihood function that results from assuming the data arise from a distribution in the exponential family (with a mean and dispersion that depend on the predictors), using using numerical optimization methods._

### Interpretation of the effects

Notice that, in a GLM, because the mean of the response been transformed in a particular way, the coefficients must be interpreted carefully. In particular, $\beta_{j}$ is how a per-unit change in $x_{j}$ increases or decreases the _transformed_ mean $g(\mu_i)$.

### Example: logistic regression

In a previous class, we estimated the probability of death given infection for the wild boar when infected with viruses in the family _Coronaviridae_. In your current homework, you have been tasked with extending that model to all host and parasite family combinations. Excitingly, the dataset which have used includes information about the mean evolutionary isolation of all hosts that are infected with a parasite from all other hosts which are known to be infected.

Farrell and Davies tested if the mean evolutionary isolation affected the probability of death. They used a complex model to control for sampling bias and other confounding aspects of the data. We will ignore those complexities and see if, using a GLM, we can recapitulate their findings. To get started, load the `disease_distance.csv` dataset.

```{r}
disease_distance &lt;- read_csv("data/disease_distance.csv")

disease_distance %&gt;% 
  mutate(AnyDeaths = case_when(Deaths &gt; 0 ~ 1,
                               Deaths == 0 ~ 0)) -&gt; DataBern

DataBern |&gt;
  ggplot(aes(x = EvoIso, y = AnyDeaths)) + 
  geom_point()
```

These are binary data. We CANNOT use a linear model, as this assumes the data are Normal. In fact, these data are distributed according to a Bernoulli($p$) distribution. This is a member of the exponential family and the type of generalized linear model when the data are distributed in this way (i.e., are binary) is called **logistic regression**.

The mean of Bernoulli($p$) data is just $p$, and the link function for $p$ is

$$ \text{logit}(p) = \log\frac{p}{1-p}.$$

$\text{logit}(p)$ is called the *log-odds*, which can be thought of as a likelihood the response takes on the value one. In logistic regression, the log-odds ratio is modeled as a linear combination in the predictors:

$$ \text{logit}(p_i) =  \beta_0 + \beta_1 x_{1i} + \cdots + \beta_p x_{pi}.$$

Notice that increasing $x_j$ by one unit results in change $\beta_j$ to the link-transformed response.This is how the effect sizes are interpreted for GLMs such as this one.


```{r}
model &lt;- glm(AnyDeaths ~ EvoIso, 
             family = "binomial", 
             data = DataBern)
summary(model)
```

#### Challenge

How do we interpret the regression coefficient above?

#### Challenge

What are the log-odds of death if the evolutionary isolation of hosts is $EI = 200$? How much does this quantity change if the evolutionary isolation were to increase by 20 million years?

#### Visualizing the fitted model

```{r}
DataBern |&gt;
  ggplot(aes(x = EvoIso, y = AnyDeaths)) + 
  geom_point() +
  geom_smooth(method = "glm", 
              method.args = list(family = "binomial")
              )

### base R implementation

EvoIso &lt;- seq(0, 200, 0.1)
predicted_prob &lt;- predict(model, list(EvoIso = EvoIso), type = "response")

plot(DataBern$EvoIso, DataBern$AnyDeaths)
lines(EvoIso, predicted_prob)
```

### Other GLMs

Here are some common types of response variables and their corresponding distributions:

-   **Count data**: the **Poisson** distribution
-   **Over-dispersed count data** (when the count data is more spread out than "expected"): the **negative binomial** distribution
-   **Binary data** (two discrete categories): the **binomial** distribution
-   Counts of occurrences of $K$ different types: the **multinomial** distribution
-   **Times** between $r$ events: the **gamma** distribution, which is equivilent to the exponential when $r=1$

You will have the opportunity to implement such models in your homework and on the challenge assignment!

## Power analysis

When designing experiments or how best to collect data, it is best to think about the model you will fit and the kind of experiment you need to design in order to have sufficient power to detect an effect of interest. For example, if we thought that age affected the likelihood that a host dies of a disease, then we would likely fit a logistic regression-type model. By _simulating_ data from such a model, we can determine

- the minimum sample size required to reliably estimate an effect of a certain size
- the minimum sample size required to detect an effect of a certain size at a fixed significance level
- the minimum effect size that can be detected at a fixed significance level and sample size

### Example

Below is an example of a simulation in which binary disease data (death/no death; 0/1) are simulated hosts of various ages, assuming the the log-odds of disease is a linear function of age. (Notice that we assume that host lifetimes are exponentially distributed with rate parameter $\lambda = 1/10$ years. This means that hosts, in this simulation, live for an average of 10 years.) We then fit a logistic regression to this data to determine the effect of age on disease risk.

```{r}
data_generator &lt;- function(sample_size = 100, effect = 0.1){
  age &lt;- rexp(n = sample_size, rate = 1/10)
  linear_predictor &lt;- effect*age
  prob &lt;- 1/(1+exp(-linear_predictor))
  
  disease_status &lt;- c()
  
  for (i in 1:length(prob)){
  disease_status[i] &lt;- rbinom(n = 1, size = 1, prob = prob[i])
  }
  
  return(data.frame(age = age, disease_status = disease_status))
}

data &lt;- data_generator()

data %&gt;% pivot_longer(! age) %&gt;% 
  ggplot(aes(x = age, y = value)) + geom_point() + 
  geom_smooth(method = "glm", method.args = list(family = "binomial")
              ) + labs(y = "prob. of disease (i.e., disease status =1)")

model &lt;- glm(disease_status~scale(age), family = binomial, data = data)
summary(model)
```

Next, we will write a function that performs a **power analysis**. We will use this function determine the sample size that is required so that simulating the age-disease data repeatedly we identify a significant effect of age on disease status (i.e., reject the null hypothesis) at level $\alpha = 0.01$ *at least $95\%$ of the time*.

```{r}
power_analysis_function &lt;- function(sample_size){
  
  sims &lt;- 1000
  pvalues &lt;- c()
  for (i in 1:sims){
  data &lt;- data_generator(sample_size)
  model &lt;- glm(disease_status~scale(age), family = binomial, data = data)
  pvalues[i] &lt;- summary(model)$coefficients[2,4]
  }
  
  power_estimate &lt;- length(which(pvalues &lt; 0.01))/length(pvalues)
  
  return(power_estimate)
}

sample_sizes &lt;- seq(150,200,10); power_estimates &lt;- c()

for (i in 1:length(sample_sizes)){
  power_estimates[i] &lt;- power_analysis_function(sample_size = sample_sizes[i])
}

knitr::kable(cbind(n = sample_sizes, power = power_estimates))
```

## Dealing with dependent data!

To illustrate how mixed models work and what kinds of questions we can answer using them, we will use the sexual size dimorphism data which we analyzed last class. Recall that we did NOT find a significant effect of sex on the average body size. There was no effect of the interaction between Order and sex on body size. Indeed, this matches what you saw in the homework questions where you had to visualize the data --- most of the variation in body size was _between_ orders.

In the models we built, we ignored a pretty important fact about the data: species have a common history (i.e., phylogeny). Thus, observations are not independent! This can make drawing inferences from comparative data difficult. **We will address how to deal with the non-independence of data (due to a common history of species, replication in blocks, etc.) using three approaches.**

But, first, let's download the data we will use in this section!

```{r}
SSDdata &lt;- read_csv("data/SSDinMammals.csv")
```

```{r}
mammal_length &lt;- SSDdata %&gt;%
  select(c("Order", "Scientific_Name", "lengthM", "lengthF")) %&gt;%
  pivot_longer(c(lengthM, lengthF), names_to = "sex", values_to = "length",
               names_pattern = "length(.)")

mammal_mass &lt;- SSDdata %&gt;%
  select(c("Order", "Scientific_Name", "massM", "massF")) %&gt;%
  pivot_longer(c(massM, massF), names_to = "sex", values_to = "mass",
               names_pattern = "mass(.)")

mass_nodup &lt;- mammal_mass %&gt;% 
  group_by(Scientific_Name, sex) %&gt;%
  distinct(Scientific_Name, sex, .keep_all = TRUE)

length_nodup &lt;- mammal_length %&gt;% 
  group_by(Scientific_Name, sex) %&gt;%
  distinct(Scientific_Name, sex, .keep_all = TRUE)

mammal_long &lt;- full_join(mass_nodup, length_nodup, 
                         by = join_by("Scientific_Name", "sex", "Order")) |&gt;
  drop_na()

glimpse(mammal_long)
```

### Group-by-group analyses

**One first way we can handle dependent data is to split observations into groups such that, within each group, observations are independent (or approximately so).** This is what we did last class when we fit order-specific regression coefficients of sex on body size. We saw that ALL order-specific effects had confidence intervals which overlapped zero!

```{r, warning=F}
# run linear model of size on sex for EACH Order

Order_specific_models &lt;- 
  mammal_long |&gt; 
  group_by(Order) |&gt;
  do(model = tidy(lm(log(mass) ~ sex, data = .), conf.int = T))
  
# get coefficients for each Order

Order_specific_models |&gt;
  unnest()

# visualize effects, CIs, and p values

Order_specific_models |&gt;
  unnest() |&gt;
  subset(term == "sexM") |&gt;
  group_by(Order) |&gt;
  ggplot(aes(y = Order, x = estimate, 
             xmin = conf.low,
             xmax = conf.high
             )
         ) +
  geom_pointrange() +
  geom_vline(xintercept = 0, lty = 2)
```

No effect of sex on body size, for any of the orders!

#### Challenge

How would you adjust the previous plot to show the estimated intercepts AND the effects of sex?

```{r}
Order_specific_models |&gt;
  unnest() |&gt;
  # subset(term == "sexM") |&gt;
  group_by(Order) |&gt;
  ggplot(aes(y = Order, x = estimate, 
             xmin = conf.low,
             xmax = conf.high,
             color = term
             )
         ) +
  geom_pointrange() +
  geom_vline(xintercept = 0, lty = 2)
```

#### Challenge

How would you retrieve model fits for each Order in base R?

```{r, include=F}
for (i in 1:nrow(Order_specific_models)){
  print(summary(Order_specific_models$mod[[i]]))
}
```

#### Challenge

Adjust the plot above so that it the size of the estimates depend on the number of observations in an Order? _Hint: determine the number of observations per Order and then use the `merge()` function with the `Order_specific_models` tibble. To adjust the range of point sizes, use `scale_size()`._

```{r, include=F}
mammal_long |&gt; 
  group_by(Order) |&gt;
  summarise(number_obs = n()) |&gt;
  merge(Order_specific_models) |&gt;
  unnest() |&gt;
  ggplot(aes(y = Order, x = estimate, 
             xmin = conf.low,
             xmax = conf.high,
             color = term,
             size = number_obs
             )
         ) +
  geom_pointrange() +
  geom_vline(xintercept = 0, lty = 2) +
  scale_size(range = c(0, 2))
```

#### Pros and cons

There are several advantages to preforming a group-by-group analysis:

- **Fitting more complex models (e.g., those with random effects) can be difficult.** There may be convergence issues, and interpretation of effects and $p$-values can be tricky.
- The group-by-group analysis is **robust in the face of unbalanced data** (i.e., when there are more observations for some groups than others).
- **The conclusions from a group-by-group analysis are _conservative_.**
- The group-by-group analysis is **fairly easy to implement**.

Among the disparages to this approach are the following:

- With fewer samples per group, the analysis **may be under-powered**.
- Splitting the data into groups means there are more coefficients to estimate, and thus confidence intervals to be estimated and hypothesis tests to be performed. This means there is a greater chance that we obtain a **spurious association**.*
- It is **difficult to draw conclusions about the variance between groups.** In some applications, this is of interest. In others, it is not.

*One solution is to adjust the significant level based on the number of tests conducted. If $k$ hypotheses are tested, an [common adjustment](https://en.wikipedia.org/wiki/Bonferroni_correction) is to set $\alpha = 0.05/k$.

### Controlling for phylogeny

A common reason data are dependent in biology is that **species share a common history of descent with modification.** When we assume that the observations are independent, we make implicit assumptions about the degree to which the characters at the tips of a phylogeny have underwent independent evolution. Sometimes, when species are diverged by many millions of years and traits evolve quickly, it is reasonable to ignore the phylogenetic constraints on the data. In other cases, it is essential to consider the role of phylogeny.

Several methods can control for phylogeny (if known). In fact, such methods can _use_ information in the branching pattern of species to draw inferences about the evolution of ecologically-important traits (such as body and brain size, dispersal rate, etc.). We will not discuss how to implement phylogenetic comparative methods, but it is good to know they exist and how, at a surface level, they work.

Intuition for how PCMs work is easiest to grasp when we consider a tree with $n$ species at the tips. If we know the pairwise divergence times for all species, then we can _transform_ the data so that observations are independent by looking at all _differences_ of traits. Not all differences are equally informative; if species have diverged a long time ago, there has been more time for differences to build up. PCMs account for this by specifying how the distribution of trait differences between species $i$ and $j$ depends on the time that has elapsed since these species diverged, In particular, the larger the divergence time, the greater the variance in $Y_i-Y_j$, the difference in trait values between species $i$ and $j$. A simple way to do this is to write $\sigma_{ij} = \sigma^2/T_{ij}$.^[Note: this gives rise to a likelihood function that is functionally VERY similar to the one for the linear model with constant error variance.]

### Random effects!

Another way one can account for dependent data is by including _random effects_. Random effects are often used to control for the fact that observations are clustered (e.g., trait data for species belonging to a higher taxonomic unit, measurements of plant biomass from plots of land).

#### Structure and assumptions

A common way models with random effects are formulated is as follows:

$$Y_{ij} \sim \text{Normal}(\beta_0 + \beta_1 x_{1ij} + \cdots + \beta_{p} x_{pij} + b_{0i} + b_{1i} z_{1ij} + \cdots + b_{qi} z_{qij}, \sigma_{ij}^2),$$

where $ij$ is the $i$th observation from the $j$th cluster and

$$b_{ki} \sim \text{Normal}(0,\tau_k^2).$$

This gives rise to a distribution for $Y$ which depends on the values of the random effects $b_0,\dots,b_q$. Under the hood, methods that fit models of this form numerically maximize a version of the likelihood function that results from these assumptions.

#### Challenge

In each of the following examples, which effects might be reasonably treated as fixed vs. random? Justify your answer.

1. Suppose we are working with rodents that have been infected with an evolved strain of the parasite that causes malaria. Some rodents have been treated with a prospective vaccine and others sham-vaccinated. We are interested if a proxy for virulence (e.g., density of infected red blood cells) depends on vaccination status.

2. Suppose we conduct the same experiment, except rodents are caged are sets of four.

3. Suppose we measure the time between sightings of a raccoon in a Toronto neighborhood using six camera traps (strategically placed in the neighborhood). We do this for year, and want to ask if mean daily temperature predicts the frequency of raccoon occurrence.

4. Suppose that, every year, we go to the Amazon and measure the number of bird calls that come from $n=30$ trees. Assume that the trees are far enough apart we can treat them as independent. We measure the number of birds in a tree, characteristics of the tree (cover, height, width, age), the year in which a measurement was done, and the mean temperature that year. We want to know if tree cover affects the frequency of bird calls, and if it interacts with temperature.

For more on the difference between fixed and random effects, check out the following resources

- [this Cross Validated post](https://stats.stackexchange.com/questions/4700/what-is-the-difference-between-fixed-effect-random-effect-in-mixed-effect-model)
- [this Dynamic Ecology post](https://dynamicecology.wordpress.com/2015/11/04/is-it-a-fixed-or-random-effect/)
- [this book chapter](https://bookdown.org/steve_midway/DAR/random-effects.html)

#### Using `lme4` to fit random and mixed effect models

We will start by fitting a model of log body size on log length where the intercept is random depending on the Order. Based on a quick visualization of the data, such a model may be appropriate.

```{r}
ggplot(data = mammal_long, aes(y = log(mass), 
                               x = log(length)
                               )
       ) + 
    geom_point(aes(color = Order), size = 3) -&gt; p

p 
```

```{r, include=F}
# no Order-dependence:
p + geom_smooth(method = "lm")

# intercept AND slope fixed and specific to the Order:
p + geom_smooth(method = "lm") + facet_wrap(~Order) 
```

To fit regress the response on a set of fixed effects with random _intercepts_ that depends the values assumed by a random effect `x`, we call `lmer` and write a linear model with `(1|x)`.

```{r}
## random intercept for Order

model &lt;- lmer(formula = log(mass) ~ log(length) + (1|Order), data = mammal_long)
summary(model)

fixef(model)
ranef(model)
```

From top to bottom, the output is telling us that the model was fit using a method called REML (restricted maximum likelihood). It returns information about the residuals, random effects, and fixed effects. The output also tells us about the **estimated variance for the random effects** in the model. Here, the variance associated with Order is 4.256. The variance explained by Order is 4.256/(4.256+0.847) _after controlling for the fixed effects_. Note the denominator here is the total variance, including from the residuals. As usual, we also have information about the fixed effects.

To visualize the model, we can predict the overall and Order-specific relationship of log mass on log length for all of the Orders represented in the data.

```{r}
mammal_long |&gt; ungroup() |&gt; select(mass, length, Order) |&gt; 
  mutate(fit.m = predict(model, re.form = NA), # does not include random effects
         fit.c = predict(model, re.form = NULL) # includes random effects
         ) -&gt;
  predicted_values

predicted_values |&gt;
  ggplot(aes(x = log(length), y = log(mass), color = Order)) +
  geom_point(size = 3) +
  geom_line(inherit.aes = F, aes(x = log(length), y = fit.c, color = Order), size = 1) +
  geom_line(inherit.aes = F, aes(x = log(length), y = fit.m), color = "black", size = 2)
```

The thick black line corresponds to the fitted values associated with the fixed-effect component of the model. The colored lines correspond to the fitted values estimated for each Order. 

Perhaps a model with Order-specific random intercepts AND slopes would be better. We fit this model using the code below. The key difference in syntax is that we write `(1+fixed effect|random effect)` to indicate that the random effect has an effect on both the intercept _and_ slope of the response on the fixed effect.

```{r}
## random intercept and slope for Order

model2 &lt;- lmer(formula = log(mass) ~ log(length) + (1+log(length)|Order), data = mammal_long)
summary(model2)

fixef(model2)
ranef(model2)
```

```{r}
mammal_long |&gt; ungroup() |&gt; select(mass, length, Order) |&gt; 
  mutate(fit.m = predict(model2, re.form = NA),
         fit.c = predict(model2, re.form = NULL)
         ) -&gt;
  predicted_values

predicted_values |&gt;
  ggplot(aes(x = log(length), y = log(mass), color = Order)) +
  geom_point(size = 3) +
  geom_line(inherit.aes = F, aes(x = log(length), y = fit.c, color = Order), size = 1) +
  geom_line(inherit.aes = F, aes(x = log(length), y = fit.m), color = "black", size = 2)
```

Why does this look like a mess? As stated above, **mixed models do not work well when datasets are unbalanced.** Indeed, the number of observations in the different Orders are quite variable.

#### Challenge

Regress log body size on sex with Order as a random effect that affects the intercept AND slope of the sex-size relationship.

```{r, include=F}
Model &lt;- lmer(log(mass) ~ sex + (1+sex|Order), mammal_long)

mammal_long |&gt; ungroup() |&gt; select(Scientific_Name, mass, sex, Order) |&gt; 
  mutate(fit.m = predict(Model, re.form = NA),
         fit.c = predict(Model, re.form = NULL)
         ) -&gt;
  predicted_values

predicted_values |&gt;
  ggplot(aes(x = sex, y = log(mass))) +
  geom_line(aes(group = Scientific_Name), color = "lightgrey") +
  geom_point(color = "lightgrey", size = 2) +
  geom_point(inherit.aes = F, aes(x = sex, y = fit.c, color = Order), size = 3) +
  geom_line(inherit.aes = F, aes(x = sex, y = fit.c, color = Order, group = Order), size = 2) +
  geom_point(inherit.aes = F, aes(x = sex, y = fit.m), color = "black", size = 3) +
  geom_line(inherit.aes = F, aes(x = sex, y = fit.m, group = ""), color = "black", size = 3) +
  theme_classic()
```</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Assignments</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-00.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment00: individual interest description</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 01: base R, command line, &amp; Git(Hub)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 02: data wrangling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 03: data visualisation and exploration</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../assignments/assignment-04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignment 04: inference!</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../projects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Project description</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../projects_databases.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Some open-access databases</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Datasets</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/portal_data.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">portal_data.csv</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/pseudo.ara.busco" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pseudo.ara.busco</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/pseudo.LTRs" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pseudo.LTRs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/disease_distance.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">disease_distance.csv</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/SSDinMammals.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SSDinMammals.csv</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lectures/data/meas.csv" class="sidebar-item-text sidebar-link">
 <span class="menu-text">meas.csv</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text">Miscellaneous notes</span></span>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#lesson-preamble" id="toc-lesson-preamble" class="nav-link active" data-scroll-target="#lesson-preamble"><span class="header-section-number">2.1</span> Lesson Preamble</a></li>
  <li><a href="#setting-up-the-r-notebook" id="toc-setting-up-the-r-notebook" class="nav-link" data-scroll-target="#setting-up-the-r-notebook"><span class="header-section-number">2.2</span> Setting up the R Notebook</a></li>
  <li><a href="#creating-objects-in-r" id="toc-creating-objects-in-r" class="nav-link" data-scroll-target="#creating-objects-in-r"><span class="header-section-number">2.3</span> Creating objects in R</a>
  <ul class="collapse">
  <li><a href="#some-tips-on-naming-objects" id="toc-some-tips-on-naming-objects" class="nav-link" data-scroll-target="#some-tips-on-naming-objects"><span class="header-section-number">2.3.1</span> Some tips on naming objects</a></li>
  <li><a href="#preforming-calculations" id="toc-preforming-calculations" class="nav-link" data-scroll-target="#preforming-calculations"><span class="header-section-number">2.3.2</span> Preforming calculations</a></li>
  </ul></li>
  <li><a href="#functions-and-their-arguments" id="toc-functions-and-their-arguments" class="nav-link" data-scroll-target="#functions-and-their-arguments"><span class="header-section-number">2.4</span> Functions and their arguments</a>
  <ul class="collapse">
  <li><a href="#understanding-functions" id="toc-understanding-functions" class="nav-link" data-scroll-target="#understanding-functions"><span class="header-section-number">2.4.1</span> Understanding functions</a></li>
  <li><a href="#tab-completion" id="toc-tab-completion" class="nav-link" data-scroll-target="#tab-completion"><span class="header-section-number">2.4.2</span> Tab-completion</a></li>
  <li><a href="#help-with-defined-functions" id="toc-help-with-defined-functions" class="nav-link" data-scroll-target="#help-with-defined-functions"><span class="header-section-number">2.4.3</span> Help with defined functions</a></li>
  </ul></li>
  <li><a href="#writing-functions" id="toc-writing-functions" class="nav-link" data-scroll-target="#writing-functions"><span class="header-section-number">2.5</span> Writing functions</a></li>
  <li><a href="#vectors-and-data-types" id="toc-vectors-and-data-types" class="nav-link" data-scroll-target="#vectors-and-data-types"><span class="header-section-number">2.6</span> Vectors and data types</a></li>
  <li><a href="#subsetting-vectors" id="toc-subsetting-vectors" class="nav-link" data-scroll-target="#subsetting-vectors"><span class="header-section-number">2.7</span> Subsetting vectors</a>
  <ul class="collapse">
  <li><a href="#conditional-subsetting" id="toc-conditional-subsetting" class="nav-link" data-scroll-target="#conditional-subsetting"><span class="header-section-number">2.7.1</span> Conditional subsetting</a></li>
  <li><a href="#strings-character-vectors" id="toc-strings-character-vectors" class="nav-link" data-scroll-target="#strings-character-vectors"><span class="header-section-number">2.7.2</span> Strings (character vectors)</a></li>
  </ul></li>
  <li><a href="#missing-data" id="toc-missing-data" class="nav-link" data-scroll-target="#missing-data"><span class="header-section-number">2.8</span> Missing data</a></li>
  <li><a href="#loops-and-vectorization" id="toc-loops-and-vectorization" class="nav-link" data-scroll-target="#loops-and-vectorization"><span class="header-section-number">2.9</span> Loops and vectorization</a></li>
  <li><a href="#dataset-background" id="toc-dataset-background" class="nav-link" data-scroll-target="#dataset-background"><span class="header-section-number">2.10</span> Dataset background</a>
  <ul class="collapse">
  <li><a href="#presentation-of-the-survey-data" id="toc-presentation-of-the-survey-data" class="nav-link" data-scroll-target="#presentation-of-the-survey-data"><span class="header-section-number">2.10.1</span> Presentation of the survey data</a></li>
  </ul></li>
  <li><a href="#what-are-data-frames" id="toc-what-are-data-frames" class="nav-link" data-scroll-target="#what-are-data-frames"><span class="header-section-number">2.11</span> What are data frames?</a>
  <ul class="collapse">
  <li><a href="#inspecting-data.frame-objects" id="toc-inspecting-data.frame-objects" class="nav-link" data-scroll-target="#inspecting-data.frame-objects"><span class="header-section-number">2.11.1</span> Inspecting <code>data.frame</code> objects</a></li>
  <li><a href="#indexing-and-subsetting-data-frames" id="toc-indexing-and-subsetting-data-frames" class="nav-link" data-scroll-target="#indexing-and-subsetting-data-frames"><span class="header-section-number">2.11.2</span> Indexing and subsetting data frames</a></li>
  </ul></li>
  <li><a href="#exporting-data" id="toc-exporting-data" class="nav-link" data-scroll-target="#exporting-data"><span class="header-section-number">2.12</span> Exporting data</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lectures/lec01-r_and_rstudio.html">Lectures</a></li><li class="breadcrumb-item"><a href="../lectures/lec02-basic-r.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to R: assignment, vectors, functions, strings, loops</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to R: assignment, vectors, functions, strings, loops</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p><em>Note: This lecture content was originally created by voluntary contributions to <a href="https://datacarpentry.org">Data Carpentry</a> and has been modified to align with the aims of EEB313. Data Carpentry is an organization focused on data literacy, with the objective of teaching skills to researchers to enable them to retrieve, view, manipulate, analyze, and store their and otherâ€™s data in an open and reproducible way.</em></p>
<p>The above paragraph is made explicit since it is one of the core features of working with an open language like R. Many smart people willingly and actively share their material publicly, so that others can modify and build off of the material themselves.</p>
<p>By being open, we can â€œstand on the shoulders of giantsâ€ and continue to contribute for others to then stand on our shoulders. Not only does this help get work done, but it also adds to a feeling of community. In fact, there is a common saying in the open source world:</p>
<blockquote class="blockquote">
<p>I came for the language and stayed for the community.</p>
</blockquote>
<p>This saying captures the spirit, generosity, and fun involved in being a part of these open source projects.</p>
<hr>
<section id="lesson-preamble" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="lesson-preamble"><span class="header-section-number">2.1</span> Lesson Preamble</h2>
<blockquote class="blockquote">
<h3 id="learning-objectives" data-number="2.1.1" class="anchored"><span class="header-section-number">2.1.1</span> Learning Objectives</h3>
<ul>
<li>Define the following terms as they relate to R: call, function, arguments, options.</li>
<li>Use comments within code blocks.</li>
<li>Do simple arithmetic operations in R using values and objects.</li>
<li>Call functions and use arguments to change their default options.</li>
<li>Define our own functions</li>
<li>Inspect the content of vectors and manipulate their content.</li>
<li>Create for-loops</li>
<li>Describe what a data frame is.</li>
<li>Load external data from a .csv file into a data frame in R.</li>
</ul>
<h3 id="lecture-outline" data-number="2.1.2" class="anchored"><span class="header-section-number">2.1.2</span> Lecture outline</h3>
<ul>
<li>Setting up your R Notebook (10 min)</li>
<li>Creating objects/variables in R (10 min)</li>
<li>Using and writing functions (15 min)</li>
<li>Vectors and data types (10 min)</li>
<li>Subsetting vectors (15 min)</li>
<li>Missing data (10 min)</li>
<li>Loops and vectorization (10 min)</li>
<li>Data set background (10 min)</li>
<li>What are data frames? (10 min)</li>
</ul>
</blockquote>
<hr>
</section>
<section id="setting-up-the-r-notebook" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="setting-up-the-r-notebook"><span class="header-section-number">2.2</span> Setting up the R Notebook</h2>
<p>Letâ€™s remove the template RStudio gives us, and add a title of our own.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="pp">---</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span><span class="kw">:</span><span class="at"> Introduction to R</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="pp">---</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This header block is called the YAML header. This is where we specify whether we want to convert this file to a HTML or PDF file. This will be discussed in more detail in another class. For now, we just care about including the lecture title here. If you are interested in playing with other YAML options, check out <a href="https://bookdown.org/yihui/rmarkdown/html-document.html">this guide</a>.</p>
<p>Under this sentence, we will insert our first code chunk. Remember that you insert a code chunk by either clicking the â€œInsertâ€ button or pressing <kbd>Ctrl/Cmd</kbd> + <kbd>Alt</kbd> + <kbd>i</kbd> simultaneously. To run a code chunk, you press the green arrow, or <kbd>Ctrl/Cmd</kbd> + <kbd>Shift</kbd> + <kbd>Enter</kbd>.</p>
</section>
<section id="creating-objects-in-r" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="creating-objects-in-r"><span class="header-section-number">2.3</span> Creating objects in R</h2>
<p>As we saw in our first class, you can get output from R simply by typing math in the console:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> <span class="sc">+</span> <span class="dv">5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="dv">12</span> <span class="sc">/</span> <span class="dv">7</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.714286</code></pre>
</div>
</div>
<p>However, to do useful and interesting things, we need to assign <em>values</em> to <em>objects</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>x <span class="sc">+</span> <span class="dv">5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
</div>
<p>You can name an object in R almost anything you want:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>joel <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>joel <span class="sc">+</span> <span class="dv">5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
</div>
<section id="challenge" class="level4" data-number="2.3.0.1">
<h4 data-number="2.3.0.1" class="anchored" data-anchor-id="challenge"><span class="header-section-number">2.3.0.1</span> Challenge</h4>
<p>So far, we have created two variables, <code>joel</code> and <code>x</code>. What is the sum of these variables?</p>
</section>
<section id="some-tips-on-naming-objects" class="level3" data-number="2.3.1">
<h3 data-number="2.3.1" class="anchored" data-anchor-id="some-tips-on-naming-objects"><span class="header-section-number">2.3.1</span> Some tips on naming objects</h3>
<ul>
<li>Objects can be given any name: <code>x</code>, <code>current_temperature</code>, <code>thing</code>, or <code>subject_id</code>.</li>
<li>You want your object names to be explicit and not too long.</li>
<li>Object names cannot start with a number: <code>x2</code> is valid, but <code>2x</code> is not.</li>
<li>R is also case sensitive: <code>joel</code> is different from <code>Joel</code>.</li>
<li>Avoid using the names of existing functions (e.g.&nbsp;<code>mean</code>, <code>df</code>). You can check whether the name is already in use by using tab completion</li>
<li>Generally good to use underscores (<code>_</code>) to separate words in variable and function names</li>
</ul>
<p>It is also recommended to use nouns for variable names, and verbs for function names. Itâ€™s important to be consistent in the styling of your code (where you put spaces, how you name variables, etc.). Using a consistent coding style<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> makes your code clearer to read for your future self and your collaborators. RStudio will format code for you if you highlight a section of code and press <kbd>Ctrl/Cmd</kbd> + <kbd>Shift</kbd> + <kbd>a</kbd>.</p>
</section>
<section id="preforming-calculations" class="level3" data-number="2.3.2">
<h3 data-number="2.3.2" class="anchored" data-anchor-id="preforming-calculations"><span class="header-section-number">2.3.2</span> Preforming calculations</h3>
<p>When assigning a value to an object, R does not print anything. You can force R to print the value by using parentheses or by typing the object name:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>weight_kg <span class="ot">&lt;-</span> <span class="dv">55</span>    <span class="co"># doesn't print anything</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>(weight_kg <span class="ot">&lt;-</span> <span class="dv">55</span>)  <span class="co"># but putting parentheses around the call prints the value of `weight_kg`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 55</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>weight_kg          <span class="co"># and so does typing the name of the object</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 55</code></pre>
</div>
</div>
<p>The variable <code>weight_kg</code> is stored in the computerâ€™s memory where R can access it, and we can start doing arithmetic with it efficiently. For instance, we may want to convert this weight into pounds (weight in pounds is 2.2 times the weight in kg):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fl">2.2</span> <span class="sc">*</span> weight_kg</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 121</code></pre>
</div>
</div>
<p>We can also change a variableâ€™s value by assigning it a new one:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>weight_kg <span class="ot">&lt;-</span> <span class="fl">57.5</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fl">2.2</span> <span class="sc">*</span> weight_kg</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 126.5</code></pre>
</div>
</div>
<p>This means that assigning a value to one variable does not change the values of other variables. For example, letâ€™s store the animalâ€™s weight in pounds in a new variable, <code>weight_lb</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>weight_lb <span class="ot">&lt;-</span> <span class="fl">2.2</span> <span class="sc">*</span> weight_kg <span class="co"># Actually, 1 kg = 2.204623 lbs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and then change <code>weight_kg</code> to 100.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>weight_kg <span class="ot">&lt;-</span> <span class="dv">100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="challenge-1" class="level4" data-number="2.3.2.1">
<h4 data-number="2.3.2.1" class="anchored" data-anchor-id="challenge-1"><span class="header-section-number">2.3.2.1</span> Challenge</h4>
<p>What do you think is the current content of the object <code>weight_lb</code>? 126.5 or 220?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>weight_lb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="challenge-2" class="level4" data-number="2.3.2.2">
<h4 data-number="2.3.2.2" class="anchored" data-anchor-id="challenge-2"><span class="header-section-number">2.3.2.2</span> Challenge</h4>
<p>What are the values after each statement in the following?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>mass <span class="ot">&lt;-</span> <span class="fl">47.5</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>age  <span class="ot">&lt;-</span> <span class="dv">122</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>mass <span class="ot">&lt;-</span> mass <span class="sc">*</span> <span class="fl">2.0</span>      <span class="co"># mass?</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>age  <span class="ot">&lt;-</span> age <span class="sc">-</span> <span class="dv">20</span>        <span class="co"># age?</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>mass_index <span class="ot">&lt;-</span> mass<span class="sc">/</span>age  <span class="co"># mass_index?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="functions-and-their-arguments" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="functions-and-their-arguments"><span class="header-section-number">2.4</span> Functions and their arguments</h2>
<section id="understanding-functions" class="level3" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="understanding-functions"><span class="header-section-number">2.4.1</span> Understanding functions</h3>
<p>Functions can be thought of as recipes. You give a few ingredients as input to a function, and it will generate an output based on these ingredients. Just as with baking, both the ingredients and the actual recipe will influence what comes out of the recipe in the end: will it be a cake or a loaf of bread? In R, the inputs to a function are not called ingredients, but rather <em>arguments</em>, and the output is called the <em>return value</em> of the function. A function does not technically have to return a value, but often does so. Functions are used to automate more complicated sets of commands and many of them are already predefined in R. A typical example would be the function <code>sqrt()</code>. The input (the argument) must be a number, and the return value (in fact, the output) is the square root of that number. Executing a function (â€˜running itâ€™) is called <em>calling</em> the function. An example of a function call is:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="dv">9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
<p>Which is the same as assigning the value to a variable and then passing that variable to the function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">9</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(a)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>b</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
<p>Here, the value of <code>a</code> is given to the <code>sqrt()</code> function, the <code>sqrt()</code> function calculates the square root, and returns the value which is then assigned to variable <code>b</code>. This function is very simple, because it takes just one argument.</p>
<p>The return â€˜valueâ€™ of a function need not be numerical (like that of <code>sqrt()</code>), and it also does not need to be a single item: it can be a set of things, or even a dataset, as we will see later on.</p>
<p>Arguments can be anything, not only numbers or filenames, but also other objects. Exactly what each argument means differs per function, and must be looked up in the documentation (see below). Some functions take arguments which may either be specified by the user, or, if left out, take on a <em>default</em> value: these are called <em>options</em>. Options are typically used to alter the way the function operates, such as whether it ignores â€˜bad valuesâ€™, or what symbol to use in a plot. However, if you want something specific, you can specify a value of your choice which will be used instead of the default.</p>
</section>
<section id="tab-completion" class="level3" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="tab-completion"><span class="header-section-number">2.4.2</span> Tab-completion</h3>
<p>To access help about <code>sqrt</code>, we are first going to learn about tab-completion. Type <code>s</code> and press <kbd>Tab</kbd>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>s<span class="sc">&lt;</span>tab<span class="sc">&gt;</span>q</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can see that R gives you suggestions of what functions and variables are available that start with the letter <code>s</code>, and thanks to RStudio they are formatted in this nice list. There are <em>many</em> suggestions here, so letâ€™s be a bit more specific and append a <code>q</code>, to find what we want. If we press enter or tab again, R will insert the selected option.</p>
<p>You can see that R inserts a pair of parentheses together with the name of the function. This is how the function syntax looks for R and many other programming languages, and it means that within these parentheses, we will specify all the arguments (the ingredients) that we want to pass to this function.</p>
<p>If we press tab again, R will helpfully display all the available parameters for this function that we can pass an argument to. The word <em>parameter</em> is used to describe the name that the argument can be passed to. More on that later.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="sc">&lt;</span>tab<span class="sc">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are many things in this list, but only one of them is marked in purple. Purple here means that this list item is a parameter we can use for the function, while yellow means that it is a variable that we defined earlier.<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a></p>
</section>
<section id="help-with-defined-functions" class="level3" data-number="2.4.3">
<h3 data-number="2.4.3" class="anchored" data-anchor-id="help-with-defined-functions"><span class="header-section-number">2.4.3</span> Help with defined functions</h3>
<p>To read the full help about <code>sqrt</code>, we can use the question mark, or type it directly into the help document browser.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>?sqrt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As you can see, <code>sqrt()</code> takes only one argument, <code>x</code>, which needs to be a <em>numerical vector</em>. Donâ€™t worry too much about the fact that it says <em>vector</em> here; we will talk more about that later. Briefly, a numerical vector is one or more numbers. In R, every number is a vector, so you donâ€™t have to do anything special to create a vector. More on vectors later.</p>
<p>Letâ€™s try a function that can take multiple arguments: <code>round()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="sc">&lt;</span>tab<span class="sc">&gt;</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>?round</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If we try round with a value:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fl">3.14159</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
<p>Here, weâ€™ve called <code>round()</code> with just one argument, <code>3.14159</code>, and it has returned the value <code>3</code>. Thatâ€™s because the default is to round to the nearest whole number, or integer. If we want more digits we can pass an argument to the <code>digits</code> parameter, to specify how many decimals we want to round to.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fl">3.14159</span>, <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3.14</code></pre>
</div>
</div>
<p>So, above we pass the <em>argument</em> <code>2</code>, to the <em>parameter</em> <code>digits</code>. Knowing this nomenclature is not essential for doing your own data analysis, but it will be very helpful when you are reading through help documents online and in RStudio.</p>
<p>We can leave out the word <code>digits</code> since we know it comes as the second parameter, after <code>x</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fl">3.14159</span>, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3.14</code></pre>
</div>
</div>
<p>As you notice, we have been leaving out <code>x</code> from the beginning. If you provide the names for both the arguments, we can switch their order:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>, <span class="at">x =</span> <span class="fl">3.14159</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3.14</code></pre>
</div>
</div>
<p>Itâ€™s good practice to put the non-optional arguments (like the number youâ€™re rounding) first in your function call, and to specify the names of all optional arguments. If you donâ€™t, someone reading your code might have to look up the definition of a function with unfamiliar arguments to understand what youâ€™re doing.</p>
</section>
</section>
<section id="writing-functions" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="writing-functions"><span class="header-section-number">2.5</span> Writing functions</h2>
<p>In this class, you will be working a lot with functions, especially those that someone else has already written. When you type <code>sum</code>, <code>c()</code>, or <code>mean()</code>, you are using a function that has been made previously and built into R. To remove some of the magic around these functions, we will go through how to make a basic function of our own. Letâ€™s start with a simple example where we add two numbers together:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>add_two_numbers <span class="ot">&lt;-</span> <span class="cf">function</span>(num1, num2) {</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(num1 <span class="sc">+</span> num2)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="fu">add_two_numbers</span>(<span class="dv">4</span>, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9</code></pre>
</div>
</div>
<p>As you can see, running this function on two numbers returns their sum. We could also assign to a variable in the function and return the function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>add_two_numbers <span class="ot">&lt;-</span> <span class="cf">function</span>(num1, num2) {</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>    my_sum <span class="ot">&lt;-</span> num1 <span class="sc">+</span> num2</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(my_sum)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="fu">add_two_numbers</span>(<span class="dv">4</span>, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9</code></pre>
</div>
</div>
<section id="challenge-3" class="level4" data-number="2.5.0.1">
<h4 data-number="2.5.0.1" class="anchored" data-anchor-id="challenge-3"><span class="header-section-number">2.5.0.1</span> Challenge</h4>
<p>Can you write a function that calculates the mean of 3 numbers?</p>
</section>
</section>
<section id="vectors-and-data-types" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="vectors-and-data-types"><span class="header-section-number">2.6</span> Vectors and data types</h2>
<p>A vector is the most common and basic data type in R, and is pretty much the workhorse of R. A vector is composed by a series of values, which can be either numbers or characters. We can assign a series of values to a vector using the <code>c()</code> function, which stands for â€œconcatenate (combine/connect one after another) values into a vectorâ€ For example we can create a vector of animal weights and assign it to a new object <code>weight_g</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>weight_g <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">50</span>, <span class="dv">60</span>, <span class="dv">65</span>, <span class="dv">82</span>) <span class="co"># Concatenate/Combine values into a vector</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>weight_g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 50 60 65 82</code></pre>
</div>
</div>
<p>You can also use the built-in command <code>seq</code>, to create a sequence of numbers without typing all of them in manually.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">30</span>) <span class="co"># This is the same as just `0:30`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
[26] 25 26 27 28 29 30</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">30</span>, <span class="dv">3</span>) <span class="co"># Every third number</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  0  3  6  9 12 15 18 21 24 27 30</code></pre>
</div>
</div>
<p>A vector can also contain characters:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>animals <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'mouse'</span>, <span class="st">'rat'</span>, <span class="st">'dog'</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>animals</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "mouse" "rat"   "dog"  </code></pre>
</div>
</div>
<p>The quotes around â€œmouseâ€, â€œratâ€, etc. are essential here and can be either single or double quotes. Without the quotes R will assume there are objects called <code>mouse</code>, <code>rat</code> and <code>dog</code>. As these objects donâ€™t exist in Râ€™s memory, there will be an error message.</p>
<p>There are many functions that allow you to inspect the content of a vector. <code>length()</code> tells you how many elements are in a particular vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(weight_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(animals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
<p>An important feature of a vector is that all of the elements are the same type of data. The function <code>class()</code> indicates the class (the type of element) of an object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(weight_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numeric"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(animals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "character"</code></pre>
</div>
</div>
<p>The function <code>str()</code> provides an overview of the <strong>str</strong>ucture of an object and its elements. It is a useful function when working with large and complex objects:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(weight_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> num [1:4] 50 60 65 82</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(animals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> chr [1:3] "mouse" "rat" "dog"</code></pre>
</div>
</div>
<p>You can use the <code>c()</code> function to add other elements to your vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>weight_g <span class="ot">&lt;-</span> <span class="fu">c</span>(weight_g, <span class="dv">90</span>) <span class="co"># add to the end of the vector</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>weight_g <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">30</span>, weight_g) <span class="co"># add to the beginning of the vector</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>weight_g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 30 50 60 65 82 90</code></pre>
</div>
</div>
<p>In the first line, we take the original vector <code>weight_g</code>, add the value <code>90</code> to the end of it, and save the result back into <code>weight_g</code>. Then we add the value <code>30</code> to the beginning, again saving the result back into <code>weight_g</code>.</p>
<p>We can do this over and over again to grow a vector, or assemble a dataset. As we program, this may be useful to add results that we are collecting or calculating.</p>
<p>An <strong>atomic vector</strong> is the simplest R <strong>data type</strong> and it is a linear vector of a single type, e.g.&nbsp;all numbers. Above, we saw 2 of the 6 main <strong>atomic vector</strong> types that R uses: <code>"character"</code> and <code>"numeric"</code> (or <code>"double"</code>). These are the basic building blocks that all R objects are built from.</p>
<p>Vectors are one of the many <strong>data structures</strong> that R uses. Other important ones are lists (<code>list</code>), matrices (<code>matrix</code>), data frames (<code>data.frame</code>), factors (<code>factor</code>) and arrays (<code>array</code>). In this class, we will focus on data frames, which is most commonly used one for data analyses.</p>
<section id="challenge-4" class="level4" data-number="2.6.0.1">
<h4 data-number="2.6.0.1" class="anchored" data-anchor-id="challenge-4"><span class="header-section-number">2.6.0.1</span> Challenge</h4>
<p>Weâ€™ve seen that atomic vectors can be of type character, numeric (or double), integer, and logical. But what happens if we try to mix these types in a single vector? Find out by using <code>class</code> to test these examples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>num_char <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="st">'a'</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>num_logical <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="cn">TRUE</span>)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>char_logical <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'a'</span>, <span class="st">'b'</span>, <span class="st">'c'</span>, <span class="cn">TRUE</span>)</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>tricky <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="st">'4'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This happens because vectors can be of only one data type. Instead of throwing an error and saying that you are trying to mix different types in the same vector, R tries to convert (coerce) the content of this vector to find a â€œcommon denominatorâ€. A logical can be turn into 1 or 0, and a number can be turned into a string/character representation. It would be difficult to do it the other way around: would 5 be TRUE or FALSE? What number would â€˜tâ€™ be?</p>
<p>In R, we call converting objects from one class into another class <em>coercion</em>. These conversions happen according to a hierarchy, whereby some types get preferentially coerced into other types. Can you draw a diagram that represents the hierarchy of how these data types are coerced?</p>
<p>This can be important to watch for in data sets that you import.</p>
</section>
</section>
<section id="subsetting-vectors" class="level2" data-number="2.7">
<h2 data-number="2.7" class="anchored" data-anchor-id="subsetting-vectors"><span class="header-section-number">2.7</span> Subsetting vectors</h2>
<p>If we want to extract one or several values from a vector, we must provide one or several indices in square brackets. For instance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>animals <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"mouse"</span>, <span class="st">"rat"</span>, <span class="st">"dog"</span>, <span class="st">"cat"</span>)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>animals[<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "rat"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>animals[<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "dog" "rat"</code></pre>
</div>
</div>
<p>We can also repeat the indices to create an object with more elements than the original one:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>more_animals <span class="ot">&lt;-</span> animals[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">4</span>)]</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>more_animals</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "mouse" "rat"   "dog"   "rat"   "mouse" "cat"  </code></pre>
</div>
</div>
<p>R indices start at 1. Programming languages like Fortran, MATLAB, Julia, and R start counting at 1, because thatâ€™s what human beings typically do. Languages in the C family (including C++, Java, Perl, and Python) count from 0 because that was historically simpler for computers and can allow for more elegant code.</p>
<section id="conditional-subsetting" class="level3" data-number="2.7.1">
<h3 data-number="2.7.1" class="anchored" data-anchor-id="conditional-subsetting"><span class="header-section-number">2.7.1</span> Conditional subsetting</h3>
<p>Another common way of subsetting is by using a logical vector. <code>TRUE</code> will select the element with the same index, while <code>FALSE</code> will not:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>weight_g <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">34</span>, <span class="dv">39</span>, <span class="dv">54</span>, <span class="dv">55</span>)</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>weight_g[<span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 21 39 54</code></pre>
</div>
</div>
<p>Typically, these logical vectors are not typed by hand, but are the output of other functions or logical tests. For instance, if you wanted to select only the values above 50:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>weight_g <span class="sc">&gt;</span> <span class="dv">50</span>    <span class="co"># will return logicals with TRUE for the indices that meet the condition</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE FALSE FALSE  TRUE  TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="do">## so we can use this to select only the values above 50</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>weight_g[weight_g <span class="sc">&gt;</span> <span class="dv">50</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 54 55</code></pre>
</div>
</div>
<p>We will consider conditions in more detail in the next few lectures.</p>
</section>
<section id="strings-character-vectors" class="level3" data-number="2.7.2">
<h3 data-number="2.7.2" class="anchored" data-anchor-id="strings-character-vectors"><span class="header-section-number">2.7.2</span> Strings (character vectors)</h3>
<p>Just a small note about character vectors, also called <code>strings</code>. There are built-in packages for subsetting them that weâ€™ll learn about later. They can be particularly relevant for ecological and genomic data because important data can be nested in complicated strings of text (ex: extracting only the observations that occurred in wet habitats from a column of habitat descriptions or only genes with functions related to drought tolerance).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>string1 <span class="ot">&lt;-</span> <span class="st">"This is a string"</span> <span class="co"># you can include spaces between your quotes</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>string2 <span class="ot">&lt;-</span> <span class="fu">c</span>(string1, <span class="st">"so is this"</span>) <span class="co"># concatenate with another string</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>string2[<span class="dv">2</span>] <span class="co"># can access the second string via subsetting</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "so is this"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Playing a bit with declaring variables</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="st">"You can include 'quotes' in a string"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "You can include 'quotes' in a string"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>string3 <span class="ot">&lt;-</span> <span class="st">'You can include "quotes" in a string'</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>string3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "You can include \"quotes\" in a string"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="st">"You can include </span><span class="sc">\"</span><span class="st">matching quotes</span><span class="sc">\"</span><span class="st"> if you 'escape' them with a backslash (</span><span class="sc">\\</span><span class="st">)"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "You can include \"matching quotes\" if you 'escape' them with a backslash (\\)"</code></pre>
</div>
</div>
</section>
</section>
<section id="missing-data" class="level2" data-number="2.8">
<h2 data-number="2.8" class="anchored" data-anchor-id="missing-data"><span class="header-section-number">2.8</span> Missing data</h2>
<p>As R was designed to analyze datasets, it includes the concept of missing data (which is uncommon in other programming languages). Missing data are represented in vectors as <code>NA</code>.</p>
<p>When doing operations on numbers, most functions will return <code>NA</code> if the data you are working with include missing values. This feature makes it harder to overlook the cases where you are dealing with missing data. You can add the argument <code>na.rm = TRUE</code> to calculate the result while ignoring the missing values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>heights <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">4</span>, <span class="cn">NA</span>, <span class="dv">6</span>)</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(heights)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(heights)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(heights, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(heights, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract those elements which are not missing values.</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>heights[<span class="sc">!</span><span class="fu">is.na</span>(heights)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2 4 4 6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Returns the object with incomplete cases removed. The returned object is an atomic vector of type `"numeric"` (or `"double"`).</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a><span class="fu">na.omit</span>(heights)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2 4 4 6
attr(,"na.action")
[1] 4
attr(,"class")
[1] "omit"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract those elements which are complete cases. The returned object is an atomic vector of type `"numeric"` (or `"double"`).</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>heights[<span class="fu">complete.cases</span>(heights)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2 4 4 6</code></pre>
</div>
</div>
<p>Recall that you can use the <code>class()</code> function to find the type of your atomic vector.</p>
<section id="challenge-5" class="level4" data-number="2.8.0.1">
<h4 data-number="2.8.0.1" class="anchored" data-anchor-id="challenge-5"><span class="header-section-number">2.8.0.1</span> Challenge</h4>
<ol type="1">
<li>Using this vector of length measurements, create a new vector with the NAs removed.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>lengths <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">24</span>, <span class="cn">NA</span>, <span class="dv">18</span>, <span class="cn">NA</span>, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Use the function <code>median()</code> to calculate the median of the <code>lengths</code> vector.</li>
</ol>
</section>
</section>
<section id="loops-and-vectorization" class="level2" data-number="2.9">
<h2 data-number="2.9" class="anchored" data-anchor-id="loops-and-vectorization"><span class="header-section-number">2.9</span> Loops and vectorization</h2>
<p>Loops, specifically for-loops, are essential to programming in general. However, in R, you should avoid them as often as possible because there are more efficient ways of doing things that you should use instead. It is still important that you understand the concept of loops and you might also use them in some of your own functions if there is no vectorized way of going about what you want to do.</p>
<p>You can think of a for-loop as: â€œfor each number contained in a list/vector, perform this operationâ€ and the syntax basically says the same thing:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>v <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>)</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (num <span class="cf">in</span> v) {</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(num)</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2
[1] 4
[1] 6</code></pre>
</div>
</div>
<p>Instead of printing out every number to the console, we could also add numbers cumulatively, to calculate the sum of all the numbers in the vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To increment `w` each time, we must first create the variable,</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="co"># which we do by setting `w &lt;- 0`, referred to as initializing.</span></span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a><span class="co"># This also ensures that `w` is zero at the start of the loop and</span></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a><span class="co"># doesn't retain the value from last time we ran this code.</span></span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>w <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (num <span class="cf">in</span> v) {</span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a>    w <span class="ot">&lt;-</span> w <span class="sc">+</span> num</span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a>w</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12</code></pre>
</div>
</div>
<p>If we put what we just did inside a function, we have essentially recreated the <code>sum</code> function in R.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>my_sum <span class="ot">&lt;-</span> <span class="cf">function</span>(input_vector) {</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>    vector_sum <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (num <span class="cf">in</span> input_vector){</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>        vector_sum <span class="ot">&lt;-</span> vector_sum <span class="sc">+</span> num</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(vector_sum)</span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-9"><a href="#cb104-9" aria-hidden="true" tabindex="-1"></a><span class="fu">my_sum</span>(v)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12</code></pre>
</div>
</div>
<p>Although this gives us the same output as the built-in function <code>sum</code>, the built-in function has many more optimizations so it is much faster than our function. In R, it is always faster to try to find a way of doing things without writing a loop yourself. When you are reading about R, you might see suggestions that you should try to <em>vectorize</em> your code to make it faster. What people are referring to, is that you should not write for loops in R and instead use the ready-made functions that are much more efficient in working with vectors and essentially performs operations on entire vector at once instead of one number at a time. Conceptually, loops operate on one element at a time while vectorized code operates on all elements of a vector at once.</p>
</section>
<section id="dataset-background" class="level2" data-number="2.10">
<h2 data-number="2.10" class="anchored" data-anchor-id="dataset-background"><span class="header-section-number">2.10</span> Dataset background</h2>
<p>Today, we will be working with real data from a longitudinal study of the species abundance in the Chihuahuan desert ecosystem near Portal, Arizona, USA. This study includes observations of plants, ants, and rodents from 1977 - 2002, and has been used in over 100 publications. More information is available in <a href="https://onlinelibrary.wiley.com/doi/10.1890/08-1222.1/full">the abstract of this paper from 2009</a>. There are several datasets available related to this study, and we will be working with datasets that have been preprocessed by the <a href="https://www.datacarpentry.org">Data Carpentry</a> to facilitate teaching. These are made available online as <em>The Portal Project Teaching Database</em>, both at the <a href="https://datacarpentry.org/ecology-workshop/data.html">Data Carpentry website</a>, and on <a href="https://figshare.com/articles/Portal_Project_Teaching_Database/1314459/6">Figshare</a>. Figshare is a great place to publish data, code, figures, and more openly to make them available for other researchers and to communicate findings that are not part of a longer paper.</p>
<section id="presentation-of-the-survey-data" class="level3" data-number="2.10.1">
<h3 data-number="2.10.1" class="anchored" data-anchor-id="presentation-of-the-survey-data"><span class="header-section-number">2.10.1</span> Presentation of the survey data</h3>
<p>We are studying the species and weight of animals caught in plots in our study area. The dataset is stored as a comma separated value (CSV) file. Each row holds information for a single animal, and the columns represent:</p>
<table class="table">
<thead>
<tr class="header">
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>record_id</td>
<td>unique id for the observation</td>
</tr>
<tr class="even">
<td>month</td>
<td>month of observation</td>
</tr>
<tr class="odd">
<td>day</td>
<td>day of observation</td>
</tr>
<tr class="even">
<td>year</td>
<td>year of observation</td>
</tr>
<tr class="odd">
<td>plot_id</td>
<td>ID of a particular plot</td>
</tr>
<tr class="even">
<td>species_id</td>
<td>2-letter code</td>
</tr>
<tr class="odd">
<td>sex</td>
<td>sex of animal (â€œMâ€, â€œFâ€)</td>
</tr>
<tr class="even">
<td>hindfoot_length</td>
<td>length of the hindfoot in mm</td>
</tr>
<tr class="odd">
<td>weight</td>
<td>weight of the animal in grams</td>
</tr>
<tr class="even">
<td>genus</td>
<td>genus of animal</td>
</tr>
<tr class="odd">
<td>species</td>
<td>species of animal</td>
</tr>
<tr class="even">
<td>taxa</td>
<td>e.g.&nbsp;rodent, reptile, bird, rabbit</td>
</tr>
<tr class="odd">
<td>plot_type</td>
<td>type of plot</td>
</tr>
</tbody>
</table>
<p>To read the data into R, we are going to use a function called <code>read_csv</code>. This function is contained in an R-package called <a href="https://readr.tidyverse.org/"><code>readr</code></a>. R-packages are a bit like browser extensions; they are not essential, but can provide nifty functionality. We will go through R-packages in general and which ones are good for data analyses. One useful option that <code>read_csv</code> includes, is the ability to read a CSV file directly from a URL, without downloading it in a separate step:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>surveys <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="st">'https://ndownloader.figshare.com/files/2292169'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>However, it is often a good idea to download the data first, so you have a copy stored locally on your computer in case you want to do some offline analyses, or the online version of the file changes or the file is taken down. You can either download the data manually or from within R:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://ndownloader.figshare.com/files/2292169"</span>,</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>              <span class="st">"data/portal_data.csv"</span>) <span class="co"># Saves to current directory with this name</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The data is read in by specifying its local path.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>surveys <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="st">'data/portal_data.csv'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 34786 Columns: 13
â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
Delimiter: ","
chr (6): species_id, sex, genus, species, taxa, plot_type
dbl (7): record_id, month, day, year, plot_id, hindfoot_length, weight

â„¹ Use `spec()` to retrieve the full column specification for this data.
â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>This statement produces some output regarding which data type it found in each column. If we want to check this in more detail, we can print the variableâ€™s value: <code>surveys</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>surveys</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 34,786 Ã— 13
   record_id month   day  year plot_id species_id sex   hindfoot_length weight
       &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt;
 1         1     7    16  1977       2 NL         M                  32     NA
 2        72     8    19  1977       2 NL         M                  31     NA
 3       224     9    13  1977       2 NL         &lt;NA&gt;               NA     NA
 4       266    10    16  1977       2 NL         &lt;NA&gt;               NA     NA
 5       349    11    12  1977       2 NL         &lt;NA&gt;               NA     NA
 6       363    11    12  1977       2 NL         &lt;NA&gt;               NA     NA
 7       435    12    10  1977       2 NL         &lt;NA&gt;               NA     NA
 8       506     1     8  1978       2 NL         &lt;NA&gt;               NA     NA
 9       588     2    18  1978       2 NL         M                  NA    218
10       661     3    11  1978       2 NL         &lt;NA&gt;               NA     NA
# â„¹ 34,776 more rows
# â„¹ 4 more variables: genus &lt;chr&gt;, species &lt;chr&gt;, taxa &lt;chr&gt;, plot_type &lt;chr&gt;</code></pre>
</div>
</div>
<p>This displays a nice tabular view of the data, which also includes pagination when there are many rows and we can click the arrow to view all the columns. Technically, this object is actually a <code>tibble</code> rather than a data frame, as indicated in the output. The reason for this is that <code>read_csv</code> automatically converts the data into to a <code>tibble</code> when loading it. Since a <code>tibble</code> is just a data frame with some convenient extra functionality, we will use these words interchangeably from now on.</p>
<p>If we just want to glance at how the data frame looks, it is sufficient to display only the top (the first 6 lines) using the function <code>head()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(surveys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 Ã— 13
  record_id month   day  year plot_id species_id sex   hindfoot_length weight
      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt;
1         1     7    16  1977       2 NL         M                  32     NA
2        72     8    19  1977       2 NL         M                  31     NA
3       224     9    13  1977       2 NL         &lt;NA&gt;               NA     NA
4       266    10    16  1977       2 NL         &lt;NA&gt;               NA     NA
5       349    11    12  1977       2 NL         &lt;NA&gt;               NA     NA
6       363    11    12  1977       2 NL         &lt;NA&gt;               NA     NA
# â„¹ 4 more variables: genus &lt;chr&gt;, species &lt;chr&gt;, taxa &lt;chr&gt;, plot_type &lt;chr&gt;</code></pre>
</div>
</div>
</section>
</section>
<section id="what-are-data-frames" class="level2" data-number="2.11">
<h2 data-number="2.11" class="anchored" data-anchor-id="what-are-data-frames"><span class="header-section-number">2.11</span> What are data frames?</h2>
<p>Data frames are the <em>de facto</em> data structure for most tabular data, and what we use for statistics and plotting. A data frame can be created by hand, but most commonly they are generated by the function <code>read_csv()</code>; in other words, when importing spreadsheets from your hard drive (or the web).</p>
<p>A data frame is a representation of data in the format of a table where the columns are vectors that all have the same length. Because the columns are vectors, they all contain the same type of data as we discussed in last class (e.g., characters, integers, factors). We can see this when inspecting the structure of a data frame with the function <code>str()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(surveys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>spc_tbl_ [34,786 Ã— 13] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ record_id      : num [1:34786] 1 72 224 266 349 363 435 506 588 661 ...
 $ month          : num [1:34786] 7 8 9 10 11 11 12 1 2 3 ...
 $ day            : num [1:34786] 16 19 13 16 12 12 10 8 18 11 ...
 $ year           : num [1:34786] 1977 1977 1977 1977 1977 ...
 $ plot_id        : num [1:34786] 2 2 2 2 2 2 2 2 2 2 ...
 $ species_id     : chr [1:34786] "NL" "NL" "NL" "NL" ...
 $ sex            : chr [1:34786] "M" "M" NA NA ...
 $ hindfoot_length: num [1:34786] 32 31 NA NA NA NA NA NA NA NA ...
 $ weight         : num [1:34786] NA NA NA NA NA NA NA NA 218 NA ...
 $ genus          : chr [1:34786] "Neotoma" "Neotoma" "Neotoma" "Neotoma" ...
 $ species        : chr [1:34786] "albigula" "albigula" "albigula" "albigula" ...
 $ taxa           : chr [1:34786] "Rodent" "Rodent" "Rodent" "Rodent" ...
 $ plot_type      : chr [1:34786] "Control" "Control" "Control" "Control" ...
 - attr(*, "spec")=
  .. cols(
  ..   record_id = col_double(),
  ..   month = col_double(),
  ..   day = col_double(),
  ..   year = col_double(),
  ..   plot_id = col_double(),
  ..   species_id = col_character(),
  ..   sex = col_character(),
  ..   hindfoot_length = col_double(),
  ..   weight = col_double(),
  ..   genus = col_character(),
  ..   species = col_character(),
  ..   taxa = col_character(),
  ..   plot_type = col_character()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
</div>
<p>Integer refers to a whole number, such as 1, 2, 3, 4, etc. Numbers with decimals, 1.0, 2.4, 3.333, are referred to as floats. Factors are used to represent categorical data. Factors can be ordered or unordered, and understanding them is necessary for statistical analysis and for plotting. Factors are stored as integers, and have labels (text) associated with these unique integers. While factors look (and often behave) like character vectors, they are actually integers under the hood, and you need to be careful when treating them like strings.</p>
<section id="inspecting-data.frame-objects" class="level3" data-number="2.11.1">
<h3 data-number="2.11.1" class="anchored" data-anchor-id="inspecting-data.frame-objects"><span class="header-section-number">2.11.1</span> Inspecting <code>data.frame</code> objects</h3>
<p>We already saw how the functions <code>head()</code> and <code>str()</code> can be useful to check the content and the structure of a data frame. Here is a non-exhaustive list of functions to get a sense of the content/structure of the data. Letâ€™s try them out!</p>
<ul>
<li>Size:
<ul>
<li><code>dim(surveys)</code> - returns a vector with the number of rows in the first element and the number of columns as the second element (the dimensions of the object)</li>
<li><code>nrow(surveys)</code> - returns the number of rows</li>
<li><code>ncol(surveys)</code> - returns the number of columns</li>
</ul></li>
<li>Content:
<ul>
<li><code>head(surveys)</code> - shows the first 6 rows</li>
<li><code>tail(surveys)</code> - shows the last 6 rows</li>
</ul></li>
<li>Names:
<ul>
<li><code>names(surveys)</code> - returns the column names (synonym of <code>colnames()</code> for <code>data.frame</code> objects)</li>
<li><code>rownames(surveys)</code> - returns the row names</li>
</ul></li>
<li>Summary:
<ul>
<li><code>str(surveys)</code> - structure of the object and information about the class, length, and content of each column</li>
<li><code>summary(surveys)</code> - summary statistics for each column</li>
</ul></li>
</ul>
<p>Note: most of these functions are â€œgenericâ€, they can be used on other types of objects besides <code>data.frame</code>.</p>
<section id="challenge-6" class="level4" data-number="2.11.1.1">
<h4 data-number="2.11.1.1" class="anchored" data-anchor-id="challenge-6"><span class="header-section-number">2.11.1.1</span> Challenge</h4>
<p>Based on the output of <code>str(surveys)</code>, can you answer the following questions?</p>
<ul>
<li>What is the class of the object <code>surveys</code>?</li>
<li>How many rows and how many columns are in this object?</li>
<li>How many species have been recorded during these surveys?</li>
</ul>
</section>
</section>
<section id="indexing-and-subsetting-data-frames" class="level3" data-number="2.11.2">
<h3 data-number="2.11.2" class="anchored" data-anchor-id="indexing-and-subsetting-data-frames"><span class="header-section-number">2.11.2</span> Indexing and subsetting data frames</h3>
<p>Our survey data frame has rows and columns (it has 2 dimensions). If we want to extract some specific data from it, we need to specify the â€œcoordinatesâ€ we want from it. Row numbers come first, followed by column numbers. When indexing, base R data frames return a different format depending on how we index the data (i.e.&nbsp;either a vector or a data frame), but with enhanced data frames, <code>tibbles</code>, the returned object is almost always a data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>surveys[<span class="dv">1</span>, <span class="dv">1</span>]   <span class="co"># first element in the first column of the data frame</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 Ã— 1
  record_id
      &lt;dbl&gt;
1         1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>surveys[<span class="dv">1</span>, <span class="dv">6</span>]   <span class="co"># first element in the 6th column</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 Ã— 1
  species_id
  &lt;chr&gt;     
1 NL        </code></pre>
</div>
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>surveys[, <span class="dv">1</span>]    <span class="co"># first column in the data frame</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 34,786 Ã— 1
   record_id
       &lt;dbl&gt;
 1         1
 2        72
 3       224
 4       266
 5       349
 6       363
 7       435
 8       506
 9       588
10       661
# â„¹ 34,776 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>surveys[<span class="dv">1</span>]      <span class="co"># first column in the data frame</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 34,786 Ã— 1
   record_id
       &lt;dbl&gt;
 1         1
 2        72
 3       224
 4       266
 5       349
 6       363
 7       435
 8       506
 9       588
10       661
# â„¹ 34,776 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>surveys[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">7</span>] <span class="co"># first three elements in the 7th column</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 Ã— 1
  sex  
  &lt;chr&gt;
1 M    
2 M    
3 &lt;NA&gt; </code></pre>
</div>
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>surveys[<span class="dv">3</span>, ]    <span class="co"># the 3rd element for all columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 Ã— 13
  record_id month   day  year plot_id species_id sex   hindfoot_length weight
      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt;
1       224     9    13  1977       2 NL         &lt;NA&gt;               NA     NA
# â„¹ 4 more variables: genus &lt;chr&gt;, species &lt;chr&gt;, taxa &lt;chr&gt;, plot_type &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>surveys[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, ]  <span class="co"># equivalent to head(surveys)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 Ã— 13
  record_id month   day  year plot_id species_id sex   hindfoot_length weight
      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt;
1         1     7    16  1977       2 NL         M                  32     NA
2        72     8    19  1977       2 NL         M                  31     NA
3       224     9    13  1977       2 NL         &lt;NA&gt;               NA     NA
4       266    10    16  1977       2 NL         &lt;NA&gt;               NA     NA
5       349    11    12  1977       2 NL         &lt;NA&gt;               NA     NA
6       363    11    12  1977       2 NL         &lt;NA&gt;               NA     NA
# â„¹ 4 more variables: genus &lt;chr&gt;, species &lt;chr&gt;, taxa &lt;chr&gt;, plot_type &lt;chr&gt;</code></pre>
</div>
</div>
<p><code>:</code> is a special operator that creates numeric vectors of integers in increasing or decreasing order; test <code>1:10</code> and <code>10:1</code> for instance. This works similarly to <code>seq</code>, which we looked at earlier in class:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="dv">0</span><span class="sc">:</span><span class="dv">10</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  0  1  2  3  4  5  6  7  8  9 10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  0  1  2  3  4  5  6  7  8  9 10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can test if all elements are the same</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a><span class="dv">0</span><span class="sc">:</span><span class="dv">10</span> <span class="sc">==</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">10</span> <span class="sc">==</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">10</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>You can also exclude certain parts of a data frame using the â€œ<code>-</code>â€ sign:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>surveys[,<span class="sc">-</span><span class="dv">1</span>]    <span class="co"># All columns, except the first</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 34,786 Ã— 12
   month   day  year plot_id species_id sex   hindfoot_length weight genus  
   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  
 1     7    16  1977       2 NL         M                  32     NA Neotoma
 2     8    19  1977       2 NL         M                  31     NA Neotoma
 3     9    13  1977       2 NL         &lt;NA&gt;               NA     NA Neotoma
 4    10    16  1977       2 NL         &lt;NA&gt;               NA     NA Neotoma
 5    11    12  1977       2 NL         &lt;NA&gt;               NA     NA Neotoma
 6    11    12  1977       2 NL         &lt;NA&gt;               NA     NA Neotoma
 7    12    10  1977       2 NL         &lt;NA&gt;               NA     NA Neotoma
 8     1     8  1978       2 NL         &lt;NA&gt;               NA     NA Neotoma
 9     2    18  1978       2 NL         M                  NA    218 Neotoma
10     3    11  1978       2 NL         &lt;NA&gt;               NA     NA Neotoma
# â„¹ 34,776 more rows
# â„¹ 3 more variables: species &lt;chr&gt;, taxa &lt;chr&gt;, plot_type &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a>surveys[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">7</span><span class="sc">:</span><span class="dv">34786</span>),] <span class="co"># Equivalent to head(surveys)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 Ã— 13
  record_id month   day  year plot_id species_id sex   hindfoot_length weight
      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt;
1         1     7    16  1977       2 NL         M                  32     NA
2        72     8    19  1977       2 NL         M                  31     NA
3       224     9    13  1977       2 NL         &lt;NA&gt;               NA     NA
4       266    10    16  1977       2 NL         &lt;NA&gt;               NA     NA
5       349    11    12  1977       2 NL         &lt;NA&gt;               NA     NA
6       363    11    12  1977       2 NL         &lt;NA&gt;               NA     NA
# â„¹ 4 more variables: genus &lt;chr&gt;, species &lt;chr&gt;, taxa &lt;chr&gt;, plot_type &lt;chr&gt;</code></pre>
</div>
</div>
<p>As well as using numeric values to subset a <code>data.frame</code> (or <code>matrix</code>), columns can be called by name, using one of the four following notations: <!-- Not sure how important it is to learn the difference vs just teaching the preferred way with the footnote that there are other ways also. --></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>surveys[<span class="st">"species_id"</span>]       <span class="co"># Result is a data.frame</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 34,786 Ã— 1
   species_id
   &lt;chr&gt;     
 1 NL        
 2 NL        
 3 NL        
 4 NL        
 5 NL        
 6 NL        
 7 NL        
 8 NL        
 9 NL        
10 NL        
# â„¹ 34,776 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>surveys[, <span class="st">"species_id"</span>]     <span class="co"># Result is a data.frame</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 34,786 Ã— 1
   species_id
   &lt;chr&gt;     
 1 NL        
 2 NL        
 3 NL        
 4 NL        
 5 NL        
 6 NL        
 7 NL        
 8 NL        
 9 NL        
10 NL        
# â„¹ 34,776 more rows</code></pre>
</div>
</div>
<p>For our purposes, these notations are equivalent. RStudio knows about the columns in your data frame, so you can take advantage of the autocompletion feature to get the full and correct column name.</p>
<p>Another syntax that is often used to specify column names is <code>$</code>. In this case, the returned object is actually a vector. We will not go into detail about this, but since it is such common usage, it is good to be aware of this.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We use `head()` since the output from vectors are not automatically cut off</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and we don't want to clutter the screen with all the `species_id` values</span></span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(surveys<span class="sc">$</span>species_id)          <span class="co"># Result is a vector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "NL" "NL" "NL" "NL" "NL" "NL"</code></pre>
</div>
</div>
<section id="challenge-7" class="level4" data-number="2.11.2.1">
<h4 data-number="2.11.2.1" class="anchored" data-anchor-id="challenge-7"><span class="header-section-number">2.11.2.1</span> Challenge</h4>
<ol type="1">
<li><p>Create a <code>data.frame</code> (<code>surveys_200</code>) containing only the observations from row 200 of the <code>surveys</code> dataset.</p></li>
<li><p>Notice how <code>nrow()</code> gave you the number of rows in a <code>data.frame</code>?</p>
<ul>
<li>Use that number to pull out just that last row in the data frame.</li>
<li>Compare that with what you see as the last row using <code>tail()</code> to make sure itâ€™s meeting expectations.</li>
<li>Pull out that last row using <code>nrow()</code> instead of the row number.</li>
<li>Create a new data frame object (<code>surveys_last</code>) from that last row.</li>
</ul></li>
<li><p>Use <code>nrow()</code> to extract the row that is in the middle of the data frame. Store the content of this row in an object named <code>surveys_middle</code>.</p></li>
<li><p>Combine <code>nrow()</code> with the <code>-</code> notation above to reproduce the behavior of <code>head(surveys)</code> keeping just the first through 6th rows of the surveys dataset.</p></li>
</ol>
</section>
</section>
</section>
<section id="exporting-data" class="level2" data-number="2.12">
<h2 data-number="2.12" class="anchored" data-anchor-id="exporting-data"><span class="header-section-number">2.12</span> Exporting data</h2>
<p>As you begin to play with your raw data, you may want to export these new, processed, datasets to share them with your collaborators or for archival. Similar to the <code>read_csv()</code> function used for reading CSV files into R, there is a <code>write_csv()</code> function that generates CSV files from data frames.</p>
<p>Manually create a new folder called â€œdata-processedâ€ in your directory. Alternatively, get R to help you with it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"Processed data"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are going to prepare a cleaned up version of the data without NAs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that this omits observations with NA in *any* column.</span></span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a><span class="co"># There is no way to control which columns to use.</span></span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a>surveys_complete_naomit <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(surveys)</span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb149-5"><a href="#cb149-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare the dimensions of the original and the cleaned data frame</span></span>
<span id="cb149-6"><a href="#cb149-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(surveys)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 34786    13</code></pre>
</div>
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(surveys_complete_naomit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 30676    13</code></pre>
</div>
</div>
<p>Now that our dataset is ready, we can save it as a CSV file in our <code>Processed data</code> folder.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To save to current directory</span></span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(surveys_complete_naomit, <span class="st">"surveys_complete_naomit.csv"</span>)</span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb153-4"><a href="#cb153-4" aria-hidden="true" tabindex="-1"></a><span class="co"># To save to newly created directory</span></span>
<span id="cb153-5"><a href="#cb153-5" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(surveys_complete_naomit, </span>
<span id="cb153-6"><a href="#cb153-6" aria-hidden="true" tabindex="-1"></a>          <span class="fu">file.path</span>(<span class="st">"~/Processed data"</span>, <span class="st">"surveys_complete_naomit.csv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next lecture, weâ€™re going to discuss collaboration with GitHub and go over an intro to the command line.</p>


</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p>Refer to the <a href="https://style.tidyverse.org/index.html">tidy style guide</a> for which style to adhere to.<a href="#fnref1" class="footnote-back" role="doc-backlink">â†©ï¸Ž</a></p></li>
<li id="fn2"><p>There are a few other symbols as well, all of which can be viewed at the end of <a href="https://support.rstudio.com/hc/en-us/articles/205273297-Code-Completion">this post about RStudio code completion</a>.<a href="#fnref2" class="footnote-back" role="doc-backlink">â†©ï¸Ž</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../lectures/lec01-r_and_rstudio.html" class="pagination-link" aria-label="Introduction to the course">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to the course</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../lectures/lec03-command_git.html" class="pagination-link" aria-label="Command line &amp; Git(Hub)">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Command line &amp; Git(Hub)</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>